---
output: html_document
editor_options: 
  chunk_output_type: console
---

# Fine-mapping with SuSiE

```{r setup, include=FALSE}
WORDS_TO_IGNORE <- c(
  "Alesha", "bigreadr", "bioavailable", "dplyr", "Gao" , "GCST", "ggplot",
  "ggrepel", "Hatton", "PIPs", "pre", "rss", "SuSiE", "th", "COJO")
source("knitr-options.R")
source("spelling-check.R")
```

This tutorial was created largely by modifying pre-existing tutorials written by Alesha Hatton (https://cnsgenomics.com/data/teaching/GNGWS23/module1/11_fine-mapping.html) and Gao Wang (https://stephenslab.github.io/susie-paper/manuscript_results/pedagogical_example.html). Thank you Alesha and Gao for your excellent tutorials!

## Objective 

SuSiE [@wang2020simple; @zou2022fine] is a fine-mapping method used to determine which variants are most likely to be causal for a trait. It uses a Bayesian iterative approach to create "credible sets" of SNPs that are likely to each contain a causal variant.

In this tutorial we will first use SuSiE on simulated data where we know the true causal effects. We will then apply it to real data from [@ruth2020using]. 

### Requirements:

- 1000 Genomes European reference panel in PLINK1 format (bed/bim/fam)  
- GWAS summary statistics from [@ruth2020using] (GWAS catalog accession: GCST90012102)
- PLINK 1.9
- R packages susieR, dplyr, ggplot2, ggrepel, bigsnpr, and bigreadr.  


### Load libraries
```{r, warning=FALSE, message=FALSE}
library(susieR)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(bigsnpr)
library(bigreadr)
set.seed(7236)
```

## Simulated Example 
This example uses simulated data to illustrate the use of SuSiE for fine-mapping over stage-wise selection.


### Load data
```{r}
# This data is included in the susieR package
dat <- N2finemapping
str(dat)
# This simulated data-set has two replicates. Let's focus on the first replicate:
y <- dat$Y[, 1]
```

### Fit SuSiE to the data 

Also perform univariate regression so that PIPs can be compared with p-values 
```{r}
# Fit SuSiE with L=5 (maximum number of causal variants per replicate)
fitted <- susie(dat$X, y, L = 5,
                estimate_residual_variance = TRUE,
                scaled_prior_variance = 0.2,
                tol = 1e-3, track_fit = TRUE,
                compute_univariate_zscore = TRUE,
                coverage = 0.95, min_abs_corr = 0.1)
str(fitted, max.level = 1)

# Let's have a look at the first iteration (by setting max_iter=1 this will give us the first iteration of fitted)
fitted.one.iter <- susie(dat$X, y, L = 5, max_iter = 1,
                         estimate_residual_variance = TRUE,
                         scaled_prior_variance = 0.2,
                         tol = 1e-3,
                         coverage = 0.95, min_abs_corr = 0.1)
str(fitted.one.iter)
```


### Plot the SuSiE results
Plot both PIPs from SuSiE as well as the p-values from the regressions.
```{r, out.width="95%"}
b <- dat$true_coef[, 1]
b[which(b != 0)] <- 1

# Run this code all at once to get side-by-side plots
par_saved <- par(mfrow = c(1, 3), cex.axis = 0.9)
# Plot the marginal associations 
susie_plot(fitted, y = "z", b = b, max_cs = 1, main = "Marginal associations", 
           xlab = "variable (SNP)", col = "#767676")
# Plot PIPs after the first iteration
susie_plot(fitted.one.iter, y = "PIP", b = b, max_cs = 0.4, 
           main = "IBSS after 1 iteration", add_legend = FALSE,
           ylim = c(0, 1), xlab = "variable (SNP)", col = "#767676") 
# Plot PIPs after convergence 
susie_plot(fitted, y = "PIP", b = b, max_cs = 0.4, 
           main = "IBSS after 10 iterations", add_legend = FALSE, 
           ylim = c(0, 1), xlab = "variable (SNP)", col = "#767676")
par(par_saved)  # back to as before
```

The "true" effects are highlighted in red. The strongest signal by p-value does not contain the causal variant, but is being tagged by two causal variants. The first iteration of SuSiE identifies the strongest signal by p-value, but by the 10th iteration the true causal variants are identified within two credible sets.

### Let's take a closer look at which variants are in the credible sets 
```{r}
fitted.one.iter$sets$cs
fitted$sets$cs
```

### How correlated are the variants in the credible sets?
```{r}
fitted$sets$purity
```


## Real data

In the simulation we used individual level data (genotypes and phenotypes). Often this is not available due to data access restrictions. Here we will use GWAS summary statistics and an LD reference panel. 
Let's look at bioavailable testosterone in females from [@ruth2020using].

### Import GWAS data

Here we use (a subset of) the GWAS summary statistics originally accessible from [the GWAS Catalog](https://www.ebi.ac.uk/gwas/studies/GCST90012102) (accession number: GCST90012102). 

```{r}
# original data
# tgz <- runonce::download_file(
#   "https://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics/GCST90012001-GCST90013000/GCST90012102/GCST90012102_buildGRCh37.tsv.gz",
#   dir = "tmp-data")  # 419 MB

# small subset, for the tutorial
tgz <- runonce::download_file(
  "https://figshare.com/ndownloader/files/55262768", 
  dir = "tmp-data", fname = "FT_sumstats_small.tsv.gz")

readLines(tgz, n = 5)
```

And some reference panel (also subsetted for this tutorial) to compute LD from:

```{r, warning=FALSE}
# original data
# gzip <- runonce::download_file(
#   "https://vu.data.surfsara.nl/index.php/s/VZNByNwpD8qqINe/download",
#   fname = "g1000_eur.zip", dir = "tmp-data")  # 488 MB

# small subset, for the tutorial
gzip <- runonce::download_file(
  "https://figshare.com/ndownloader/files/55263098",
  fname = "g1000_eur_small.zip", dir = "tmp-data")

unzip(gzip, exdir = "tmp-data", overwrite = FALSE)
g1000_map <- fread2("tmp-data/g1000_eur_small.bim", select = c(1:2, 4:6),
                    col.names = c("chr", "rsid", "pos", "a1", "a0"))
```


```{r}
# import GWAS summary statistics of bioavailable testosterone in females
FT <- fread2(tgz)
head(FT)

FT_cleaned <- FT %>%
  # QC on MAF >= 0.01
  filter(effect_allele_frequency >= 0.01 & effect_allele_frequency <= 0.99) %>%
  # rename some columns for compatibility with bigsnpr::snp_match()
  rename(chr = chromosome, pos = base_pair_location, 
         a1 = effect_allele, a0 = other_allele) %>%
  # align alleles to reference panel
  snp_match(g1000_map)  # match on chr/pos but can also match on chr/rsid
```

### Find window around loci of interest

Let's look at two loci:

- Locus 1: rs1989147.

- Locus 2: rs34954997, rs11879227, rs34255979. These variants in locus 2 are nearby so will be considered as one locus for the purpose of running SuSiE.

```{r}
# find coordinates
filter(FT_cleaned, rsid == "rs1989147")  # on chr1

# Extract 1 Mb locus surrounding rs1989147
locus1 <- filter(FT_cleaned, chr == 1, pos > 7909373 - 5e5, pos < 7909373 + 5e5)

# Plot locus
ggplot(locus1, aes(x = pos, y = -log10(p_value))) +
  theme_bw(13) +
  geom_point(alpha = 0.8, size = 1.3) +
  geom_point(aes(x = pos, y = -log10(p_value)), color = "red", size = 2,
             data = filter(locus1, rsid == "rs1989147")) +
  geom_label_repel(aes(label = ifelse(rsid == "rs1989147", rsid, NA)), 
                   size = 4, min.segment.length = 0)
```


### Create an LD matrix {#ldplink}

As we are using GWAS summary statistics, we need information of the correlation between variants. Here we are using 1000 Genomes European for convenience. If you are performing this analysis yourself, it is advised to use a larger sample. For fine-mapping, it is often recommended to use in-sample LD (LD computed from the original GWAS data).
Here we use PLINK, but remember that you can also use functions from bigsnpr for computing LD and actually also from bigstatsr when you need all correlation values for a set of variants (see `big_cor()`).

```{r}
# Export the SNPs in this locus
locus1_snps_txt <- tempfile(fileext = ".txt")
write(locus1$rsid, file = locus1_snps_txt)

# Compute correlation matrix (LD) with PLINK
plink <- download_plink("tmp-data")
system(glue::glue(
  "{plink} --bfile tmp-data/g1000_eur_small",
  " --extract {locus1_snps_txt}",
  " --keep-allele-order",
  " --r square",  # correlations as a square matrix, not squared correlations
  " --out tmp-data/ld_locus1"
))

# Read in LD matrix
ld_mat <- as.matrix(fread2("tmp-data/ld_locus1.ld"))
```


### Read in data and fit SuSiE

`susie_rss()` is the function to fit SuSiE using summary statistics. Earlier we used `susie()` function which requires individual level data.
```{r}
# Fit SuSiE using sample size from publication (n=188507)
fitted_rss_1 = susie_rss(bhat = locus1$beta, shat = locus1$standard_error, 
                         n = 188507, R = ld_mat)
```

### Examine results
```{r}
# Examine results
summary(fitted_rss_1)
locus1_susie <- arrange(summary(fitted_rss_1)$vars, variable)
locus1 <- cbind(locus1, locus1_susie[-1])

# Plot
ggplot(locus1, aes(x = pos, y = -log10(p_value))) +
  geom_point(alpha = 0.8, size = 1.3) +
  geom_point(color = "red", size = 2, data = filter(locus1, cs == 1)) +
  geom_label_repel(aes(label = rsid), size = 2, max.overlaps = 45,
                   data = filter(locus1, cs == 1))
```

:::: {.infobox .exo}
How do the PIPs compare to the p-values? Generate a plot to show this.
::::


### Locus 2

```{r}
locus2 <- filter(FT_cleaned, chr == 19, pos > 45417638 - 5e5, pos < 46384830 + 5e5)
```

:::: {.infobox .exo}
Now perform similar analyses on locus2.
::::

:::: {.infobox .exo}
How do the PIPs compare to the p-values? Generate a plot to show this.
:::: 

:::: {.infobox .exo}
What if we changed to a 75% credible set for locus 2? Write some code to do this and plot. Can we trust this result?
:::: 

:::: {.infobox .question}
Why might there be no result in fitted_rss_2 for L=2?
:::: 

### Reflections

- What can you do next with these results?

- What is the advantages of using SuSiE over conditional approaches (such as LD clumping or COJO)?

- What assumptions are we making when using SuSiE?
