# Introduction

```{r setup, include=FALSE}
WORDS_TO_IGNORE <- c("EWAS", "methylation", "futureverse", "tidyverse", "epigenetic", "omics", "reimplementing", "transcriptomic")
source("knitr-options.R")
source("spelling-check.R")
```

## Main motivation for developing {bigstatsr} and {bigsnpr}

At the time, there was a notable lack of user-friendly and efficient R packages for genetic analyses, which posed challenges for researchers. 
The existing workflows often required the use of disparate software tools with inconsistent input formats, reliance on text files for parameter settings, and limited compatibility with exploratory data analysis and familiar R packages. 
Additionally, the development of new methods was hindered by the absence of tools supporting a simple matrix-like data structure. 
To address these challenges, I initiated the development of the R package {bigsnpr} in 2016, reimplementing the statistical methods commonly used in genetic analyses within a cohesive and accessible R framework.

At some point, I realized that many functions (e.g. to perform genome-wide association studies (GWAS), principal component analysis (PCA), summary statistics, etc.) were not specific to genotype data. Indeed, both association studies and PCA are applicable to other omics data, such as transcriptomic or epigenetic datasets. Therefore I decided to move all these functions that could be used on any data stored as a matrix into a new R package, {bigstatsr}. This is why there are two packages, where {bigstatsr} can basically be used by any field using data stored as large numeric matrices, while {bigsnpr} provides some tools more specific to genotype data, largely building on top of {bigstatsr}. The initial description of the two packages is available in @prive2017efficient.

:::: {.infobox .info}
Functions starting with `big_` are part of {bigstatsr}, while functions starting with `snp_` or `bed_` are part of {bigsnpr}.
::::

## Features

There are now many functions implemented in the packages. You can find a comprehensive list of available functions on the package website [of {bigstatsr}](https://privefl.github.io/bigstatsr/reference/index.html) and [of {bigsnpr}](https://privefl.github.io/bigsnpr/reference/index.html).

The next table presents an overview of common genetic analyses that are already implemented in {bigstatsr} and {bigsnpr}. This listing is inspired from table 1 of @visscher201710.

| Analysis                          | Available in {bigstatsr} and {bigsnpr} | Still missing               | Citations              |
|-----------------------------------|----------------------------------------|-----------------------------|----------------------------------|
| Polygenic risk scores             | - penalized regressions on individual-level data <br> - (stacked) C+T <br> - LDpred2 <br> - lassosum2        | multi-ancestry training          | @prive2019efficient <br> @prive2019making <br> @prive2020ldpred2 <br> @prive2021identifying  |
| Population structure              | - principal component analysis (with automatic removal of LD) <br> - ancestry inference <br> - fixation index ($F_{ST}$) <br> - local adaptation |           | @prive2017efficient <br> @prive2020efficient <br> @prive2021high <br> @prive2021using <br> @prive2020performing |
| GWAS                              | linear and logistic                    | - mixed models <br> - rare variant association                 | @prive2017efficient |
| Genome-wide assessment of LD      | - sparse correlation matrix <br> - optimal LD splitting    |          | @prive2017efficient <br> @prive2021optimal |
| Estimation of SNP heritability    | - LD score regression <br> - LDpred2-auto        |     | @prive2020ldpred2 <br> @prive2022inferring |
| Estimation of polygenicity        | LDpred2-auto                           |      | @prive2022inferring |
| Estimation of genetic correlation |                                        | need to extend LDpred2-auto |                            |
| Fine-mapping                      | LDpred2-auto                           |  - using millions of variants <br> - integrating functional annotations    | @prive2022inferring |
| Imputation of GWAS summary statistics           |                                        | in development |                            |
| Mendelian randomization           |                                        | completely missing          |                          |
| Miscellaneous                     | - integration with PLINK <br> - format conversion <br> - imputation of genotyped variants <br> - matrix operations <br> - summaries       |              | @prive2017efficient |


## Example code

- When you want to use a function for the first time, check the documentation and the examples in there (usually they are very short).

- There are also many (longer) tutorials available (usually one with each paper), which will be linked from here or are available at the packages' websites.

- Some other examples are provided in this extended documentation (i.e. in the next chapters).

- All the code used in all my papers is available on GitHub. It mostly consists of R scripts based on {bigsnpr}, {bigstatsr}, the tidyverse, and the futureverse [@prive2017efficient; @wickham2019welcome; @bengtsson2020unifying].

## Installation

Both packages are available on CRAN, so you can use `install.packages()`:

```{r, eval=FALSE}
install.packages("bigstatsr")
install.packages("bigsnpr")
```

To install the latest versions (from GitHub), you can use {remotes}:

```{r, eval=FALSE}
# install.packages("remotes")
remotes::install_github("privefl/bigstatsr")
remotes::install_github("privefl/bigsnpr")
```


## Correct spellings

<blockquote class="twitter-tweet" data-lang="en" align="center" width="50%"><p lang="en" dir="ltr">A friendly reminder:<br>The correct spelling is<br> - bigstatsr -- not bigstatr / BIGstatsR<br> - bigsnpr -- not BIGsnpR / bigSNPr<br> - pcadapt -- not PCAdapt<br> - LDpred -- not LDPred<br><br>Thank you<br>The kittens thank you too <a href="https://t.co/S8wyE4G6BG">pic.twitter.com/S8wyE4G6BG</a></p>&mdash; Florian Priv√© (@privefl) <a href="https://twitter.com/privefl/status/1329109034218237953?ref_src=twsrc%5Etfw">November 18, 2020</a></blockquote> <script async src="https://platform.twitter.com/widgets.js" charset="utf-8"></script> 
