<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Priv√© &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="population-structure.html"/>
<link rel="next" href="fine-mapping-with-susie.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.2</b> License</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i><b>1.3</b> Author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> The bigSNP format from {bigsnpr}</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.3</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.4</b> The FBM format from {bigstatsr}</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.5</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.5.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.5.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.5.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.5.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.5.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.5.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.5.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry-inference"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>6</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="6.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>6.1</b> Objective</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>6.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="6.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>6.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>6.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>6.2.1</b> Load data</a></li>
<li class="chapter" data-level="6.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>6.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="6.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>6.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="6.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>6.2.4</b> Let‚Äôs take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="6.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>6.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>6.3</b> Real data</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>6.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="6.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>6.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="6.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#create-ld-matrix"><i class="fa fa-check"></i><b>6.3.3</b> Create LD matrix</a></li>
<li class="chapter" data-level="6.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>6.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="6.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>6.3.5</b> Examine results</a></li>
<li class="chapter" data-level="6.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>6.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="6.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>6.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="genome-wide-association-study-gwas" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Genome-Wide Association Study (GWAS)<a href="genome-wide-association-study-gwas.html#genome-wide-association-study-gwas" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In {bigstatsr}, you can perform both standard linear and logistic regressions GWAS, using either <code>big_univLinReg()</code> or <code>big_univLogReg()</code>.
Function <code>big_univLinReg()</code> should be very fast, while <code>big_univLogReg()</code> is slower.</p>
<div class="infobox info">
<p>This type of association, where each variable is considered independently, can be performed for any type of FBM (i.e.¬†it does not have to be a genotype matrix). This is why these two functions are in package {bigstatsr}, and not {bigsnpr}.</p>
</div>
<div id="exo-gwas" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Example<a href="genome-wide-association-study-gwas.html#exo-gwas" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let us reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">3.3</a> and the PCs in <a href="population-structure.html#exo-pca">4.4</a>.</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="genome-wide-association-study-gwas.html#cb118-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="genome-wide-association-study-gwas.html#cb120-1" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb120-2"><a href="genome-wide-association-study-gwas.html#cb120-2" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span></code></pre></div>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="genome-wide-association-study-gwas.html#cb121-1" tabindex="-1"></a>obj.svd <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;tmp-data/PCA_GWAS_data.rds&quot;</span>)</span>
<span id="cb121-2"><a href="genome-wide-association-study-gwas.html#cb121-2" tabindex="-1"></a>PC <span class="ot">&lt;-</span> <span class="fu">predict</span>(obj.svd)</span></code></pre></div>
<p>The clinical data includes age, sex, high-density lipoprotein (HDL)-cholesterol (<code>hdl</code>), low-density lipoprotein (LDL)-cholesterol (<code>ldl</code>), triglycerides (<code>tg</code>) and coronary artery disease status (<code>CAD</code>).</p>
<p>For the set of covariates, we will use sex, age, and the first 6 PCs:</p>
<div class="sourceCode" id="cb122"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="genome-wide-association-study-gwas.html#cb122-1" tabindex="-1"></a>covar <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.matrix</span>(obj.bigsnp<span class="sc">$</span>fam[<span class="fu">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>)]), PC[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code></pre></div>
<div class="infobox caution">
<p>You probably should not account for other information such as cholesterol as they are heritable covariates, which can lead to collider bias <span class="citation">(<a href="#ref-aschard2015adjusting">Aschard, Vilhj√°lmsson, Joshi, Price, &amp; Kraft, 2015</a>; <a href="#ref-day2016robust">Day, Loh, Scott, Ong, &amp; Perry, 2016</a>)</span>.</p>
</div>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="genome-wide-association-study-gwas.html#cb123-1" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb123-2"><a href="genome-wide-association-study-gwas.html#cb123-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>CAD</span>
<span id="cb123-3"><a href="genome-wide-association-study-gwas.html#cb123-3" tabindex="-1"></a>ind.gwas <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(y) <span class="sc">&amp;</span> <span class="fu">complete.cases</span>(covar))</span></code></pre></div>
<div class="infobox caution">
<p>To only use a subset of the data stored as an FBM (<code>G</code> here), you should almost never make a copy of the data; instead, use parameters <code>ind.row</code> (or <code>ind.train</code>) and <code>ind.col</code> to apply functions to a subset of the data.</p>
</div>
<p>Let us perform a case-control GWAS for CAD:</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb124-1"><a href="genome-wide-association-study-gwas.html#cb124-1" tabindex="-1"></a>gwas <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb124-2"><a href="genome-wide-association-study-gwas.html#cb124-2" tabindex="-1"></a>  <span class="fu">big_univLogReg</span>(G, y[ind.gwas], <span class="at">ind.train =</span> ind.gwas,</span>
<span id="cb124-3"><a href="genome-wide-association-study-gwas.html#cb124-3" tabindex="-1"></a>                 <span class="at">covar.train =</span> covar[ind.gwas, ], <span class="at">ncores =</span> NCORES),</span>
<span id="cb124-4"><a href="genome-wide-association-study-gwas.html#cb124-4" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/GWAS_CAD.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;    0.17    0.14  157.05 
#&gt; Code finished running at 2025-06-10 13:44:27 CEST</code></pre>
<div class="infobox info">
<p>This takes about two minutes with 4 cores on my laptop. Note that <code>big_univLinReg()</code> takes two seconds only, and should give very similar p-values, if you just need something quick.</p>
</div>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="genome-wide-association-study-gwas.html#cb126-1" tabindex="-1"></a><span class="fu">plot</span>(gwas)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-6-1.png" width="95%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="genome-wide-association-study-gwas.html#cb127-1" tabindex="-1"></a>CHR <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>map<span class="sc">$</span>chromosome</span>
<span id="cb127-2"><a href="genome-wide-association-study-gwas.html#cb127-2" tabindex="-1"></a>POS <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>map<span class="sc">$</span>physical.pos</span>
<span id="cb127-3"><a href="genome-wide-association-study-gwas.html#cb127-3" tabindex="-1"></a><span class="fu">snp_manhattan</span>(gwas, CHR, POS, <span class="at">npoints =</span> <span class="fl">50e3</span>) <span class="sc">+</span></span>
<span id="cb127-4"><a href="genome-wide-association-study-gwas.html#cb127-4" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-6-2.png" width="95%" style="display: block; margin: auto;" />
Here, nothing is genome-wide significant because of the small sample size.</p>
<div class="infobox exo">
<p>Perform a GWAS for the other phenotypes, and look at the histogram, Q-Q plot and Manhattan plot.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="genome-wide-association-study-gwas.html#cb128-1" tabindex="-1"></a>y2 <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>hdl</span>
<span id="cb128-2"><a href="genome-wide-association-study-gwas.html#cb128-2" tabindex="-1"></a>ind.gwas2 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(y2) <span class="sc">&amp;</span> <span class="fu">complete.cases</span>(covar))</span>
<span id="cb128-3"><a href="genome-wide-association-study-gwas.html#cb128-3" tabindex="-1"></a>gwas2 <span class="ot">&lt;-</span> <span class="fu">big_univLinReg</span>(G, y2[ind.gwas2], <span class="at">ind.train =</span> ind.gwas2,</span>
<span id="cb128-4"><a href="genome-wide-association-study-gwas.html#cb128-4" tabindex="-1"></a>                        <span class="at">covar.train =</span> covar[ind.gwas2, ], <span class="at">ncores =</span> NCORES)</span>
<span id="cb128-5"><a href="genome-wide-association-study-gwas.html#cb128-5" tabindex="-1"></a><span class="fu">snp_manhattan</span>(gwas2, CHR, POS, <span class="at">npoints =</span> <span class="fl">50e3</span>) <span class="sc">+</span></span>
<span id="cb128-6"><a href="genome-wide-association-study-gwas.html#cb128-6" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-7-1.png" width="95%" style="display: block; margin: auto;" /></p>
<details>
<hr />
<p>Some other example code:</p>
<ul>
<li><a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/3-GWAS.R">GWAS in iPSYCH</a>; you can perform the GWAS on multiple nodes in parallel that would each process a chunk of the variants only</li>
<li><a href="https://github.com/privefl/UKBB-PGS/blob/main/code/run-GWAS-large.R">GWAS for very large data and multiple phenotypes</a>; you should perform the GWAS for all phenotypes for a ‚Äúsmall‚Äù chunk of columns to avoid repeated access from disk, and can process these chunks on multiple nodes in parallel</li>
<li><a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/batchtools.slurm.tmpl">some template for {future.batchtools} when using Slurm</a></li>
</ul>
</div>
<div id="regenie" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> REGENIE<a href="genome-wide-association-study-gwas.html#regenie" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://rgcgithub.github.io/regenie/">regenie</a> <span class="citation">(<a href="#ref-mbatchou2021computationally">Mbatchou et al., 2021</a>)</span> is a C++ program for whole genome regression modeling of large genome-wide association studies.</p>
<p>It is developed and supported by a team of scientists at the Regeneron Genetics Center.</p>
<p>The method has the following properties:</p>
<ul>
<li><p>It works on quantitative and binary traits, including binary traits with unbalanced case-control ratios</p></li>
<li><p>It can handle population structure and relatedness</p></li>
<li><p>It can process multiple phenotypes at once efficiently</p></li>
<li><p>For binary traits, it supports Firth logistic regression and an SPA test</p></li>
<li><p>It can perform gene/region-based tests (Burden, SBAT, SKAT/SKATO, ACATV/ACATO)</p></li>
<li><p>It can perform interaction tests (GxE, GxG) as well as conditional analyses</p></li>
<li><p>Meta-analysis of REGENIE summary statistics can be performed using REMETA</p></li>
<li><p>It is fast and memory efficient üî•</p></li>
<li><p>It can provide some power boost</p></li>
<li><p>It supports the BGEN, PLINK bed/bim/fam and PLINK2 pgen/pvar/psam genetic data formats</p></li>
<li><p>It can be installed with Conda</p></li>
</ul>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-aschard2015adjusting" class="csl-entry">
Aschard, H., Vilhj√°lmsson, B.J., Joshi, A.D., Price, A.L., &amp; Kraft, P. (2015). <a href="https://doi.org/10.1016%2Fj.ajhg.2014.12.021">Adjusting for heritable covariates can bias effect estimates in genome-wide association studies</a>. <em>The American Journal of Human Genetics</em>, <em>96</em>, 329‚Äì339.
</div>
<div id="ref-day2016robust" class="csl-entry">
Day, F.R., Loh, P.-R., Scott, R.A., Ong, K.K., &amp; Perry, J.R. (2016). <a href="https://doi.org/10.1016/j.ajhg.2015.12.019">A robust example of collider bias in a genetic association study</a>. <em>The American Journal of Human Genetics</em>, <em>98</em>, 392‚Äì393.
</div>
<div id="ref-mbatchou2021computationally" class="csl-entry">
Mbatchou, J., Barnard, L., Backman, J., Marcketta, A., Kosmicki, J.A., Ziyatdinov, A., et al. (2021). <a href="https://doi.org/10.1038/s41588-021-00870-7">Computationally efficient whole-genome regression for quantitative and binary traits</a>. <em>Nature Genetics</em>, <em>53</em>, 1097‚Äì1103.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="population-structure.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="fine-mapping-with-susie.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
