<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Genome-Wide Association Study (GWAS) | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian PrivÃ© &amp; Isabelle McGrath" />


<meta name="date" content="2025-07-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="population-structure.html"/>
<link rel="next" href="gwas-summary-statistics.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#main-author"><i class="fa fa-check"></i><b>1.3</b> Main author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#imputed-data"><i class="fa fa-check"></i><b>2.1</b> Imputed data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.2</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>2.3</b> Main motivation for developing R packages bigstatsr and bigsnpr</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.4</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.5</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.6</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.7" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.7</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.7.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.7.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.7.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.7.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.7.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.7.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.7.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.8</b> Exercise</a></li>
<li class="chapter" data-level="2.9" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#matching-genetic-variants-between-datasets"><i class="fa fa-check"></i><b>2.9</b> Matching genetic variants between datasets</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
<li class="chapter" data-level="5.4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses"><i class="fa fa-check"></i><b>5.4</b> Some post GWAS analyses</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#one-use-of-prs"><i class="fa fa-check"></i><b>8.1</b> One use of PRS</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.2</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.3</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.3.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.3.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.3.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.5</b> Future of PGS</a></li>
<li class="chapter" data-level="8.6" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.6</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.6.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.6.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.6.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.6.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.6.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Letâs take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create an LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="some-other-analyses.html"><a href="some-other-analyses.html"><i class="fa fa-check"></i><b>10</b> Some other analyses</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="genome-wide-association-study-gwas" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">Chapter 5</span> Genome-Wide Association Study (GWAS)<a href="genome-wide-association-study-gwas.html#genome-wide-association-study-gwas" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In bigstatsr, you can perform both standard linear and logistic regressions GWAS, using either <code>big_univLinReg()</code> or <code>big_univLogReg()</code>.
Function <code>big_univLinReg()</code> should be very fast, while <code>big_univLogReg()</code> is slower.</p>
<div class="infobox info">
<p>This type of association, where each variable is considered independently, can be performed for any type of FBM (i.e.Â it does not have to be a genotype matrix). This is why these two functions are in package bigstatsr, and not bigsnpr.</p>
</div>
<div id="exo-gwas" class="section level2 hasAnchor" number="5.1">
<h2><span class="header-section-number">5.1</span> Example<a href="genome-wide-association-study-gwas.html#exo-gwas" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Letâs reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">3.3</a> and the PCs in <a href="population-structure.html#exo-pca">4.4</a>.</p>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb141-1"><a href="genome-wide-association-study-gwas.html#cb141-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb143-1"><a href="genome-wide-association-study-gwas.html#cb143-1" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb143-2"><a href="genome-wide-association-study-gwas.html#cb143-2" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span></code></pre></div>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="genome-wide-association-study-gwas.html#cb144-1" tabindex="-1"></a>obj.svd <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;tmp-data/PCA_GWAS_data.rds&quot;</span>)</span>
<span id="cb144-2"><a href="genome-wide-association-study-gwas.html#cb144-2" tabindex="-1"></a>PC <span class="ot">&lt;-</span> <span class="fu">predict</span>(obj.svd)</span></code></pre></div>
<p>The clinical data includes age, sex, high-density lipoprotein (HDL)-cholesterol (<code>hdl</code>), low-density lipoprotein (LDL)-cholesterol (<code>ldl</code>), triglycerides (<code>tg</code>) and coronary artery disease status (<code>CAD</code>).</p>
<p>For the set of covariates, we will use sex, age, and the first 6 PCs:</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb145-1"><a href="genome-wide-association-study-gwas.html#cb145-1" tabindex="-1"></a>covar <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.matrix</span>(obj.bigsnp<span class="sc">$</span>fam[<span class="fu">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>)]), PC[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span></code></pre></div>
<div class="infobox info">
<p>PCs are used as covariates in GWAS to correct for population structure <span class="citation">(<a href="#ref-price2006principal">Price et al., 2006</a>)</span>.</p>
</div>
<div class="infobox caution">
<p>You probably should not account for other information such as cholesterol as they are heritable covariates, which can lead to collider bias <span class="citation">(<a href="#ref-aschard2015adjusting">Aschard, VilhjÃ¡lmsson, Joshi, Price, &amp; Kraft, 2015</a>; <a href="#ref-day2016robust">Day, Loh, Scott, Ong, &amp; Perry, 2016</a>)</span>.</p>
</div>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="genome-wide-association-study-gwas.html#cb146-1" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb146-2"><a href="genome-wide-association-study-gwas.html#cb146-2" tabindex="-1"></a>y <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>CAD</span>
<span id="cb146-3"><a href="genome-wide-association-study-gwas.html#cb146-3" tabindex="-1"></a>ind.gwas <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(y) <span class="sc">&amp;</span> <span class="fu">complete.cases</span>(covar))</span></code></pre></div>
<div class="infobox caution">
<p>To only use a subset of the data stored as an FBM (<code>G</code> here), you should almost never make a copy of the data; instead, use parameters <code>ind.row</code> (or <code>ind.train</code>) and <code>ind.col</code> to apply functions to a subset of the data.</p>
</div>
<p>Letâs perform a case-control GWAS for CAD:</p>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb147-1"><a href="genome-wide-association-study-gwas.html#cb147-1" tabindex="-1"></a>gwas <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb147-2"><a href="genome-wide-association-study-gwas.html#cb147-2" tabindex="-1"></a>  <span class="fu">big_univLogReg</span>(G, y[ind.gwas], <span class="at">ind.train =</span> ind.gwas,</span>
<span id="cb147-3"><a href="genome-wide-association-study-gwas.html#cb147-3" tabindex="-1"></a>                 <span class="at">covar.train =</span> covar[ind.gwas, ], <span class="at">ncores =</span> NCORES),</span>
<span id="cb147-4"><a href="genome-wide-association-study-gwas.html#cb147-4" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/GWAS_CAD.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;    0.17    0.14  157.05 
#&gt; Code finished running at 2025-06-10 13:44:27 CEST</code></pre>
<div class="infobox info">
<p>This takes about two minutes with 4 cores on my laptop. Note that <code>big_univLinReg()</code> takes two seconds only, and should give very similar p-values, if you just need something quick.</p>
</div>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="genome-wide-association-study-gwas.html#cb149-1" tabindex="-1"></a><span class="fu">plot</span>(gwas)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-6-1.png" width="95%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="genome-wide-association-study-gwas.html#cb150-1" tabindex="-1"></a>CHR <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>map<span class="sc">$</span>chromosome</span>
<span id="cb150-2"><a href="genome-wide-association-study-gwas.html#cb150-2" tabindex="-1"></a>POS <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>map<span class="sc">$</span>physical.pos</span>
<span id="cb150-3"><a href="genome-wide-association-study-gwas.html#cb150-3" tabindex="-1"></a><span class="fu">snp_manhattan</span>(gwas, CHR, POS, <span class="at">npoints =</span> <span class="fl">50e3</span>) <span class="sc">+</span></span>
<span id="cb150-4"><a href="genome-wide-association-study-gwas.html#cb150-4" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-6-2.png" width="95%" style="display: block; margin: auto;" />
Here, nothing is genome-wide significant because of the small sample size.</p>
<div class="infobox info">
<p>Normally, a p-value threshold of <span class="math inline">\(5 \times 10^{-8}\)</span> is used when reporting significant findings, which corresponds to correcting for one million independent tests. Now, with larger GWAS datasets that include tens of millions of variants, researchers have started using an even more stringent threshold of <span class="math inline">\(5 \times 10^{-9}\)</span>.</p>
</div>
<div class="infobox exo">
<p>Perform a GWAS for the other phenotypes, and look at the histogram, Q-Q plot and Manhattan plot.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb151"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb151-1"><a href="genome-wide-association-study-gwas.html#cb151-1" tabindex="-1"></a>y2 <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>hdl</span>
<span id="cb151-2"><a href="genome-wide-association-study-gwas.html#cb151-2" tabindex="-1"></a>ind.gwas2 <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(y2) <span class="sc">&amp;</span> <span class="fu">complete.cases</span>(covar))</span>
<span id="cb151-3"><a href="genome-wide-association-study-gwas.html#cb151-3" tabindex="-1"></a>gwas2 <span class="ot">&lt;-</span> <span class="fu">big_univLinReg</span>(G, y2[ind.gwas2], <span class="at">ind.train =</span> ind.gwas2,</span>
<span id="cb151-4"><a href="genome-wide-association-study-gwas.html#cb151-4" tabindex="-1"></a>                        <span class="at">covar.train =</span> covar[ind.gwas2, ], <span class="at">ncores =</span> NCORES)</span>
<span id="cb151-5"><a href="genome-wide-association-study-gwas.html#cb151-5" tabindex="-1"></a><span class="fu">snp_manhattan</span>(gwas2, CHR, POS, <span class="at">npoints =</span> <span class="fl">50e3</span>) <span class="sc">+</span></span>
<span id="cb151-6"><a href="genome-wide-association-study-gwas.html#cb151-6" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<img src="GWAS_files/figure-html/unnamed-chunk-7-1.png" width="95%" style="display: block; margin: auto;" />
</details>
<hr />
<p>To report independent findings (instead of all variants in one peak), researchers have used LD clumping, which iteratively keeps the most associated variant and remove those that are too correlated with this one. Another strategy is to use GCTA-COJO <span class="citation">(<a href="#ref-yang2012conditional">Yang et al., 2012</a>)</span>.</p>
<div class="infobox exo">
<p>Perform some LD clumping for all variants with a p-value lower than <span class="math inline">\(10^{-5}\)</span>, using a stringent <span class="math inline">\(r^2\)</span> threshold of 0.05 over a 5 Mb (= 5000 kb) window. How many variants are left?</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="genome-wide-association-study-gwas.html#cb152-1" tabindex="-1"></a>lpval <span class="ot">&lt;-</span> <span class="sc">-</span><span class="fu">predict</span>(gwas)  <span class="co"># -log10(pval)</span></span>
<span id="cb152-2"><a href="genome-wide-association-study-gwas.html#cb152-2" tabindex="-1"></a>ind_keep <span class="ot">&lt;-</span> <span class="fu">snp_clumping</span>(G, CHR, <span class="at">S =</span> lpval, <span class="at">thr.r2 =</span> <span class="fl">0.05</span>,</span>
<span id="cb152-3"><a href="genome-wide-association-study-gwas.html#cb152-3" tabindex="-1"></a>                         <span class="at">size =</span> <span class="dv">5000</span>, <span class="at">infos.pos =</span> POS, </span>
<span id="cb152-4"><a href="genome-wide-association-study-gwas.html#cb152-4" tabindex="-1"></a>                         <span class="at">exclude =</span> <span class="fu">which</span>(lpval <span class="sc">&lt;</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">1e-5</span>)),</span>
<span id="cb152-5"><a href="genome-wide-association-study-gwas.html#cb152-5" tabindex="-1"></a>                         <span class="at">ncores =</span> NCORES)</span>
<span id="cb152-6"><a href="genome-wide-association-study-gwas.html#cb152-6" tabindex="-1"></a>obj.bigsnp<span class="sc">$</span>map[ind_keep, ]</span></code></pre></div>
<pre><code>#&gt;        chromosome marker.ID genetic.dist physical.pos allele1 allele2
#&gt; 121561          4 rs4863155            0    190300221       G       A
#&gt; 136435          5 rs4957723            0    107157011       C       T
#&gt; 224210          9 rs9632884            0     22072301       G       C</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="genome-wide-association-study-gwas.html#cb154-1" tabindex="-1"></a><span class="fu">snp_manhattan</span>(gwas, CHR, POS, <span class="at">npoints =</span> <span class="fl">50e3</span>, <span class="at">ind.highlight =</span> ind_keep) <span class="sc">+</span></span>
<span id="cb154-2"><a href="genome-wide-association-study-gwas.html#cb154-2" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">5e-8</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb154-3"><a href="genome-wide-association-study-gwas.html#cb154-3" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fu">log10</span>(<span class="fl">1e-5</span>), <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;blue&quot;</span>)</span></code></pre></div>
<p><img src="GWAS_files/figure-html/unnamed-chunk-8-1.png" width="95%" style="display: block; margin: auto;" /></p>
</details>
<hr />
<p>For some example code to run a GWAS on multiple nodes on an HPC cluster:</p>
<ul>
<li><a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/3-GWAS.R">GWAS in iPSYCH</a>; you can perform the GWAS on multiple nodes in parallel that would each process a chunk of the variants only</li>
<li><a href="https://github.com/privefl/UKBB-PGS/blob/main/code/run-GWAS-large.R">GWAS for very large data and multiple phenotypes</a>; you should perform the GWAS for all phenotypes for a âsmallâ chunk of columns to avoid repeated access from disk, and can process these chunks on multiple nodes in parallel</li>
<li><a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/batchtools.slurm.tmpl">some template for {future.batchtools} when using Slurm</a></li>
</ul>
</div>
<div id="regenie" class="section level2 hasAnchor" number="5.2">
<h2><span class="header-section-number">5.2</span> REGENIE<a href="genome-wide-association-study-gwas.html#regenie" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://rgcgithub.github.io/regenie/">regenie</a> <span class="citation">(<a href="#ref-mbatchou2021computationally">Mbatchou et al., 2021</a>)</span> is a C++ program for whole genome regression modeling of large genome-wide association studies.
It is developed and supported by a team of scientists at the Regeneron Genetics Center.
It is fancier than simple linear/logistic regressions and is currently (one of if not) the state-of-the-art GWAS method.</p>
<p>REGENIE has the following properties:</p>
<ul>
<li><p>It supports the BGEN, PLINK bed/bim/fam and PLINK2 pgen/pvar/psam genetic data formats</p></li>
<li><p>It works on both quantitative and binary traits, including binary traits with unbalanced case-control ratios</p></li>
<li><p>It can handle population structure and relatedness</p></li>
<li><p>It can provide some power boost</p></li>
<li><p>It is fast and memory efficient ð¥</p></li>
<li><p>It can process multiple phenotypes at once efficiently</p></li>
<li><p>For binary traits, it supports Firth logistic regression and an SPA test (useful in case of very unbalanced phenotypes)</p></li>
<li><p>It can perform gene/region-based tests (burden, SBAT, SKAT/SKATO, ACATV/ACATO)</p></li>
<li><p>It can perform interaction tests (GxE, GxG) as well as conditional analyses</p></li>
<li><p>Meta-analysis of REGENIE summary statistics can be performed using REMETA</p></li>
<li><p>It can be installed with Conda</p></li>
</ul>
</div>
<div id="genomic-control-and-ld-score-regression" class="section level2 hasAnchor" number="5.3">
<h2><span class="header-section-number">5.3</span> Genomic control and LD score regression<a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><span class="math inline">\(\lambda_\text{GC}\)</span>, the genomic inflation factor (GIF), was used a lot to assess whether a GWAS is confounded by population structure or relatedness (when e.g.Â larger than 1.05).
It corresponds to the ratio between the median of test statistics over the median of expected test statistics (under the null hypothesis).
However, it is no longer used because, with sufficient GWAS power and trait polygenicity, a GIF larger than 1 is actually expected <span class="citation">(<a href="#ref-yang2011genomic">Yang et al., 2011</a>)</span>.</p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="genome-wide-association-study-gwas.html#cb155-1" tabindex="-1"></a><span class="fu">snp_qq</span>(gwas) <span class="sc">+</span> </span>
<span id="cb155-2"><a href="genome-wide-association-study-gwas.html#cb155-2" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">ylim</span>(<span class="dv">1</span>, <span class="cn">NA</span>)  <span class="co"># to plot less points</span></span></code></pre></div>
<pre><code>#&gt; Warning: Removed 363713 rows containing missing values or values outside the scale range
#&gt; (`geom_point()`).</code></pre>
<p><img src="GWAS_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<hr />
<p>Then, the intercept of LD score regression was introduced to test for confounding <span class="citation">(<a href="#ref-bulik2015ld">B. K. Bulik-Sullivan et al., 2015</a>)</span>.
However, note that it was observed that LD score regression intercepts tend to increase with SNP-heritability and GWAS sample size <span class="citation">(<a href="#ref-loh2018mixed">Loh, Kichaev, Gazal, Schoech, &amp; Price, 2018</a>)</span>.</p>
</div>
<div id="some-post-gwas-analyses" class="section level2 hasAnchor" number="5.4">
<h2><span class="header-section-number">5.4</span> Some post GWAS analyses<a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>It is challenging to interpret a GWAS on its own as there are often many associations, many of which are in non-coding regions. There is a large collection of downstream analyses to aid in gaining biological understanding from your GWAS, some of which are based on GWAS summary statistics which will cover later in this course.</p>
<ul>
<li><p>There are several online tools/resources that allow you to quickly get a sense of the meaning of a variant or GWAS results. If you want to understand what is known about a particular variant, <a href="https://www.snpedia.com/index.php/SNPedia">SNPedia</a>, <a href="https://www.ncbi.nlm.nih.gov/snp/">dbSNP</a>, and <a href="https://platform.opentargets.org/">OpenTargets</a> are good places to start. Here you can find information on the peer-reviewed papers that mention this particular variant, its allele frequencies across populations, its predicted consequences, and associations with other traits.</p></li>
<li><p><a href="https://www.gtexportal.org/home/">GTEx</a> is a database of the genetic regulation of gene expression in humans, and is also useful to get a sense of whether a variant of interest may influence the expression of a particular gene. Variants in non-coding regions are generally hypothesized to affect the trait through regulatory mechanisms. This database is not an exhaustive list of genetic regulation of gene expression â a variant may influence gene expression in certain cell or tissue types or contexts not captured in the data.</p></li>
<li><p><a href="https://fuma.ctglab.nl/">FUMA</a> is a website where you can upload GWAS summary statistics and it will run a range of analyses to link genetic variants to genes. You can then further investigate these genes with pathways analyses on the same platform. To my knowledge, there is no good tool that can run a range of analyses locally in one go â this would certainly save me a lot of time and be more efficient!</p></li>
<li><p>Resources for helping linking GWAS hits to eQTLs:</p>
<ul>
<li><a href="https://www.ebi.ac.uk/eqtl/">eQTL Catalogue</a></li>
<li><a href="https://www.gtexportal.org/home/">GTEx</a></li>
<li><a href="https://cellxgene.cziscience.com/">CellxGene</a></li>
<li><a href="https://onek1k.org/">OneK1K</a></li>
</ul></li>
<li><p>Read more in <a href="other-files/Resources-for-Post-GWAS-Analysis.pptx">this PowerPoint presentation</a>.</p></li>
</ul>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-aschard2015adjusting" class="csl-entry">
Aschard, H., VilhjÃ¡lmsson, B.J., Joshi, A.D., Price, A.L., &amp; Kraft, P. (2015). <a href="https://doi.org/10.1016%2Fj.ajhg.2014.12.021">Adjusting for heritable covariates can bias effect estimates in genome-wide association studies</a>. <em>The American Journal of Human Genetics</em>, <em>96</em>, 329â339.
</div>
<div id="ref-bulik2015ld" class="csl-entry">
Bulik-Sullivan, B.K., Loh, P.-R., Finucane, H.K., Ripke, S., Yang, J., Psychiatric Genomics Consortium, S.W.G. of the, et al. (2015). <a href="https://doi.org/10.1038/ng.3211"><span>LD</span> score regression distinguishes confounding from polygenicity in genome-wide association studies</a>. <em>Nature Genetics</em>, <em>47</em>, 291â295.
</div>
<div id="ref-day2016robust" class="csl-entry">
Day, F.R., Loh, P.-R., Scott, R.A., Ong, K.K., &amp; Perry, J.R. (2016). <a href="https://doi.org/10.1016/j.ajhg.2015.12.019">A robust example of collider bias in a genetic association study</a>. <em>The American Journal of Human Genetics</em>, <em>98</em>, 392â393.
</div>
<div id="ref-loh2018mixed" class="csl-entry">
Loh, P.-R., Kichaev, G., Gazal, S., Schoech, A.P., &amp; Price, A.L. (2018). <a href="https://doi.org/10.1038/s41588-018-0144-6">Mixed-model association for biobank-scale datasets</a>. <em>Nature Genetics</em>, <em>50</em>, 906â908.
</div>
<div id="ref-mbatchou2021computationally" class="csl-entry">
Mbatchou, J., Barnard, L., Backman, J., Marcketta, A., Kosmicki, J.A., Ziyatdinov, A., et al. (2021). <a href="https://doi.org/10.1038/s41588-021-00870-7">Computationally efficient whole-genome regression for quantitative and binary traits</a>. <em>Nature Genetics</em>, <em>53</em>, 1097â1103.
</div>
<div id="ref-price2006principal" class="csl-entry">
Price, A.L., Patterson, N.J., Plenge, R.M., Weinblatt, M.E., Shadick, N.A., &amp; Reich, D. (2006). <a href="https://doi.org/10.1038/ng1847">Principal components analysis corrects for stratification in genome-wide association studies</a>. <em>Nature Genetics</em>, <em>38</em>, 904â909.
</div>
<div id="ref-yang2012conditional" class="csl-entry">
Yang, J., Ferreira, T., Morris, A.P., Medland, S.E., ANthropometric Traits (GIANT) Consortium, G.I. of, Consortium, D.G.R.A.M. (DIAGRAM), et al. (2012). <a href="https://doi.org/10.1038/ng.2213">Conditional and joint multiple-<span>SNP</span> analysis of <span>GWAS</span> summary statistics identifies additional variants influencing complex traits</a>. <em>Nature Genetics</em>, <em>44</em>, 369â375.
</div>
<div id="ref-yang2011genomic" class="csl-entry">
Yang, J., Weedon, M.N., Purcell, S., Lettre, G., Estrada, K., Willer, C.J., et al. (2011). <a href="https://doi.org/10.1038/ejhg.2011.39">Genomic inflation factors under polygenic inheritance</a>. <em>European Journal of Human Genetics</em>, <em>19</em>, 807â812.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="population-structure.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="gwas-summary-statistics.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
