<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 GWAS summary statistics | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 6 GWAS summary statistics | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 GWAS summary statistics | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 GWAS summary statistics | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="genome-wide-association-study-gwas.html"/>
<link rel="next" href="linkage-disequilibrium-ld.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#main-author"><i class="fa fa-check"></i><b>1.3</b> Main author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> Main motivation for developing R packages bigstatsr and bigsnpr</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.3</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.4</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.5</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.6</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.6.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.6.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.6.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.6.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.6.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.6.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.6.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.6.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.7" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.7</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
<li class="chapter" data-level="5.4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses"><i class="fa fa-check"></i><b>5.4</b> Some post GWAS analyses</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.1</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.2</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.2.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.2.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.2.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.3</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Future of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.5</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.5.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.5.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.5.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.5.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.5.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create an LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="some-other-analyses.html"><a href="some-other-analyses.html"><i class="fa fa-check"></i><b>10</b> Some other analyses</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gwas-summary-statistics" class="section level1 hasAnchor" number="6">
<h1><span class="header-section-number">Chapter 6</span> GWAS summary statistics<a href="gwas-summary-statistics.html#gwas-summary-statistics" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>When genome-wide association studies (GWAS) were first introduced to explore genetic factors associated with diseases, researchers may not have realized how valuable the resulting GWAS summary statistics would be.
Over the past decade, GWAS summary statistics have become essential for many genetic analyses.</p>
<p>The summary statistics generated by GWAS do not include personal information, making GWAS summary statistics easy to share and use for further analysis.
Collaboration among researchers in large consortia has facilitated meta-analyses of GWAS summary statistics from numerous study sites, resulting in unprecedentedly large GWAS sample sizes <span class="citation">(<a href="#ref-levey2021bi">Levey et al., 2021</a>; <a href="#ref-okbay2022polygenic">Okbay et al., 2022</a>; <a href="#ref-suzuki2023multi">Suzuki et al., 2023</a>; <a href="#ref-yengo2022saturated">Yengo et al., 2022</a>; <a href="#ref-zhou2021global">Zhou et al., 2022</a>)</span>.</p>
<div id="format" class="section level2 hasAnchor" number="6.1">
<h2><span class="header-section-number">6.1</span> Format<a href="gwas-summary-statistics.html#format" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This an example of GWAS summary statistics (a subset of it at least).</p>
<div class="infobox caution">
<p>Fields included in GWAS summary statistics and their specific names in the header of the file can change very much from one file to the other; this is why tools like MungeSumstats have been developed <span class="citation">(<a href="#ref-murphy2021mungesumstats">Murphy, Schilder, &amp; Skene, 2021</a>)</span>.</p>
</div>
<div class="sourceCode" id="cb144"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb144-1"><a href="gwas-summary-statistics.html#cb144-1" tabindex="-1"></a>tgz <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb144-2"><a href="gwas-summary-statistics.html#cb144-2" tabindex="-1"></a>  <span class="st">&quot;https://figshare.com/ndownloader/files/55262768&quot;</span>, </span>
<span id="cb144-3"><a href="gwas-summary-statistics.html#cb144-3" tabindex="-1"></a>  <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;FT_sumstats_small.tsv.gz&quot;</span>)</span>
<span id="cb144-4"><a href="gwas-summary-statistics.html#cb144-4" tabindex="-1"></a></span>
<span id="cb144-5"><a href="gwas-summary-statistics.html#cb144-5" tabindex="-1"></a><span class="fu">writeLines</span>(<span class="fu">readLines</span>(tgz, <span class="at">n =</span> <span class="dv">6</span>))</span></code></pre></div>
<pre><code>#&gt; variant_id   chromosome  base_pair_location  effect_allele   other_allele    effect_allele_frequency imputation_quality  beta    standard_error  p_value
#&gt; rs72858371   1   6409383 T   C   0.986224    0.977552    -0.00615364 0.0108886   0.64
#&gt; rs72633437   1   6409494 C   T   0.906138    0.988959    0.00289667  0.00434703  0.39
#&gt; rs151043271  1   6409495 G   A   0.998537    0.732147    -0.0244749  0.0386808   0.4
#&gt; rs58484986   1   6409653 C   G   0.986228    0.976894    -0.00623617 0.0108958   0.63
#&gt; rs60772726   1   6409662 A   T   0.986228    0.976844    -0.00625052 0.0108957   0.63</code></pre>
<p>They often include</p>
<ul>
<li><p>Information to match variants (chromosome, physical position, alleles, rsid).</p></li>
<li><p><span class="math inline">\(\hat{\beta}_j\)</span> or <span class="math inline">\(\hat{\gamma}_j\)</span> — the GWAS effect size of variant <span class="math inline">\(j\)</span> (marginal effect),</p></li>
<li><p><span class="math inline">\(\text{se}(\hat{\gamma}_j)\)</span> — its standard error,</p></li>
<li><p><span class="math inline">\(z_j = \frac{\hat{\gamma}_j}{\text{se}(\hat{\gamma}_j)}\)</span> — the Z-score of variant <span class="math inline">\(j\)</span>,</p></li>
<li><p>p-values derived from Z-scores</p></li>
</ul>
<p>but they more rarely include</p>
<ul>
<li><p><span class="math inline">\(n_j\)</span> — the GWAS sample size associated with variant <span class="math inline">\(j\)</span>,</p></li>
<li><p><span class="math inline">\(f_j\)</span> — the allele frequency of variant <span class="math inline">\(j\)</span>,</p></li>
<li><p><span class="math inline">\(\text{INFO}_j\)</span> — the imputation INFO score (imputation quality) of variant <span class="math inline">\(j\)</span></p></li>
</ul>
<div class="infobox question">
<p>Why would different variants have different GWAS sample sizes?</p>
</div>
<div class="infobox exo">
<p>Read-in these GWAS summary statistics and derive the p-values again, using the other columns. Recall that the Z-score squared is <span class="math inline">\(\chi^2\)</span>-distributed with one degree of freedom.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb146"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb146-1"><a href="gwas-summary-statistics.html#cb146-1" tabindex="-1"></a>(sumstats <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(tgz))</span></code></pre></div>
<pre><code>#&gt;    variant_id chromosome base_pair_location effect_allele other_allele
#&gt; 1  rs72858371          1            6409383             T            C
#&gt; 2  rs72633437          1            6409494             C            T
#&gt; 3 rs151043271          1            6409495             G            A
#&gt; 4  rs58484986          1            6409653             C            G
#&gt; 5  rs60772726          1            6409662             A            T
#&gt;   effect_allele_frequency imputation_quality        beta standard_error
#&gt; 1                0.986224           0.977552 -0.00615364     0.01088860
#&gt; 2                0.906138           0.988959  0.00289667     0.00434703
#&gt; 3                0.998537           0.732147 -0.02447490     0.03868080
#&gt; 4                0.986228           0.976894 -0.00623617     0.01089580
#&gt; 5                0.986228           0.976844 -0.00625052     0.01089570
#&gt;   p_value
#&gt; 1    0.64
#&gt; 2    0.39
#&gt; 3    0.40
#&gt; 4    0.63
#&gt; 5    0.63
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 41805 rows ]</code></pre>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="gwas-summary-statistics.html#cb148-1" tabindex="-1"></a>chi2 <span class="ot">&lt;-</span> <span class="fu">with</span>(sumstats, (beta <span class="sc">/</span> standard_error)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb148-2"><a href="gwas-summary-statistics.html#cb148-2" tabindex="-1"></a>pval <span class="ot">&lt;-</span> <span class="fu">pchisq</span>(chi2, <span class="at">df =</span> <span class="dv">1</span>, <span class="at">lower.tail =</span> <span class="cn">FALSE</span>)</span>
<span id="cb148-3"><a href="gwas-summary-statistics.html#cb148-3" tabindex="-1"></a><span class="fu">plot</span>(pval, sumstats<span class="sc">$</span>p_value, <span class="at">log =</span> <span class="st">&quot;xy&quot;</span>); <span class="fu">abline</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<img src="sumstats_files/figure-html/unnamed-chunk-2-1.png" width="70%" style="display: block; margin: auto;" />
</details>
</div>
<div id="quality-control" class="section level2 hasAnchor" number="6.2">
<h2><span class="header-section-number">6.2</span> Quality control<a href="gwas-summary-statistics.html#quality-control" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>However, <strong>pooling many GWAS summary statistics in large meta-analyses has been a double-edged sword</strong>. On the one hand, this has given much-needed power to many genetic analyses. On the other hand, the quality and standardization of GWAS summary statistics varies widely, and the more studies that are pooled together, the more issues arise when using these pooled summary data.</p>
<p>An overview of known issues in using GWAS summary statistics is presented in Figure <a href="gwas-summary-statistics.html#fig:overview-misspec">6.1</a>. This includes weakening the construction of polygenic scores <span class="citation">(<a href="#ref-prive2021identifying">Privé, Arbel, et al., 2022</a>)</span>, biasing heritability estimates <span class="citation">(<a href="#ref-gazal2018functional">Gazal et al., 2018</a>; <a href="#ref-grotzinger2023pervasive">Grotzinger, Fuente, Privé, Nivard, &amp; Tucker-Drob, 2023</a>)</span>, identifying spurious causal genes <span class="citation">(<a href="#ref-kanai2022meta">Kanai et al., 2022</a>; <a href="#ref-zou2022fine">Zou, Carbonetto, Wang, &amp; Stephens, 2022</a>)</span>, as well as undermining other genetic analyses <span class="citation">(<a href="#ref-chen2021improved">Chen et al., 2021</a>; <a href="#ref-julienne2021multitrait">Julienne et al., 2021</a>)</span>.</p>
<div class="figure" style="text-align: center"><span style="display:block;" id="fig:overview-misspec"></span>
<img src="https://github.com/privefl/thesis-docs/blob/master/figures/overview-misspec.jpg?raw=true" alt="Overview of possible errors and misspecifications in GWAS summary statistics, with possible harmful consequences, as well as possible remedies [@prive2021identifying]. $\hat{\gamma}$: GWAS effect sizes; SE: standard errors; $n^{eff}$: effective GWAS sample sizes; SD: standard deviations of genotypes; INFO: measure of imputation quality." width="100%" />
<p class="caption">
Figure 6.1: Overview of possible errors and misspecifications in GWAS summary statistics, with possible harmful consequences, as well as possible remedies <span class="citation">(<a href="#ref-prive2021identifying">Privé, Arbel, et al., 2022</a>)</span>. <span class="math inline">\(\hat{\gamma}\)</span>: GWAS effect sizes; SE: standard errors; <span class="math inline">\(n^{eff}\)</span>: effective GWAS sample sizes; SD: standard deviations of genotypes; INFO: measure of imputation quality.
</p>
</div>
<hr />
<p>The QC I recommend to perform consist in <strong>comparing standard deviations</strong> of genotypes estimated in 2 ways <span class="citation">(<a href="#ref-prive2021identifying">Privé, Arbel, et al., 2022</a>; <a href="#ref-prive2020ldpred2">Privé, Arbel, &amp; Vilhjálmsson, 2020</a>)</span>:</p>
<p><br></p>
<ol style="list-style-type: decimal">
<li><ul>
<li><p>When linear regression was used
<span class="math display" id="eq:sdlin">\[\begin{equation}
\text{sd}(G_j) \approx \dfrac{\text{sd}(y)}{\sqrt{n_j \cdot \text{se}(\hat{\gamma}_j)^2 + \hat{\gamma}_j^2}} \tag{6.1}
\end{equation}\]</span></p></li>
<li><p>When logistic regression was used (case-control phenotype)
<span class="math display" id="eq:sdlog">\[\begin{equation}\label{eq:approx-sd-log}
\text{sd}(G_j) \approx \dfrac{2}{\sqrt{n_j^\text{eff} \cdot \text{se}(\hat{\gamma}_j)^2 + \hat{\gamma}_j^2}}~, \tag{6.2}
\end{equation}\]</span>
where <span class="math inline">\(n_\text{eff} = \frac{4}{1 / n_\text{ca} + 1 / n_\text{co}}\)</span>
<br></p></li>
</ul></li>
<li><span class="math display" id="eq:sdaf">\[\begin{equation}
\text{sd}(G_j) \approx \sqrt{2 \cdot f_j \cdot (1 - f_j) \cdot \text{INFO}_j} \tag{6.3}
\end{equation}\]</span></li>
</ol>
<hr />
<p>With this QC, you can detect differences in per-variant GWAS sample sizes</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/simu-qc-plot.jpg?raw=true" width="72%" style="display: block; margin: auto;" /></p>
<p>When <span class="math inline">\(n_j\)</span> are missing, you can use this to perform some QC (e.g. at 70% of max n) or to impute <span class="math inline">\(n_j\)</span>.</p>
<hr />
<p>You can detect bias in total effective GWAS sample size</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/cad_quick_qc.png?raw=true" width="72%" style="display: block; margin: auto;" /></p>
<p><span class="math inline">\(N_\text{eff} = \frac{4}{1 / N_\text{ca} + 1 / N_\text{co}}\)</span></p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/cad_neff_perstudy.png?raw=true" width="30%" style="display: block; margin: auto;" /></p>
<p>For example, overestimating the sample size leads to underestimating the SNP heritability in many methods such as LD score regression <span class="citation">(<a href="#ref-grotzinger2023pervasive">Grotzinger et al., 2023</a>)</span>.</p>
<hr />
<p>You can detect and QC low imputation INFO scores, &amp; other issues</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/brca_onco_qc.png?raw=true" width="100%" style="display: block; margin: auto;" /></p>
<hr />
<p>Beware that multi-ancestry INFO scores are overestimated (e.g. in the UK Biobank)</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/overestim-info.png?raw=true" width="65%" style="display: block; margin: auto;" /></p>
<hr />
<div class="infobox exo">
<p>Use the following GWAS results for HDL and allele frequencies to estimate standard deviations and compare both estimates. How can you estimate <code>sd(y)</code>?</p>
</div>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb149-1"><a href="gwas-summary-statistics.html#cb149-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb149-2"><a href="gwas-summary-statistics.html#cb149-2" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb149-3"><a href="gwas-summary-statistics.html#cb149-3" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span>
<span id="cb149-4"><a href="gwas-summary-statistics.html#cb149-4" tabindex="-1"></a>obj.svd <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;tmp-data/PCA_GWAS_data.rds&quot;</span>)</span>
<span id="cb149-5"><a href="gwas-summary-statistics.html#cb149-5" tabindex="-1"></a>PC <span class="ot">&lt;-</span> <span class="fu">predict</span>(obj.svd)</span>
<span id="cb149-6"><a href="gwas-summary-statistics.html#cb149-6" tabindex="-1"></a>covar <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.matrix</span>(obj.bigsnp<span class="sc">$</span>fam[<span class="fu">c</span>(<span class="st">&quot;sex&quot;</span>, <span class="st">&quot;age&quot;</span>)]), PC[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb149-7"><a href="gwas-summary-statistics.html#cb149-7" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb149-8"><a href="gwas-summary-statistics.html#cb149-8" tabindex="-1"></a>y <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>hdl</span>
<span id="cb149-9"><a href="gwas-summary-statistics.html#cb149-9" tabindex="-1"></a>ind.gwas <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="sc">!</span><span class="fu">is.na</span>(y) <span class="sc">&amp;</span> <span class="fu">complete.cases</span>(covar))</span>
<span id="cb149-10"><a href="gwas-summary-statistics.html#cb149-10" tabindex="-1"></a>gwas <span class="ot">&lt;-</span> <span class="fu">big_univLinReg</span>(G, y[ind.gwas], <span class="at">ind.train =</span> ind.gwas,</span>
<span id="cb149-11"><a href="gwas-summary-statistics.html#cb149-11" tabindex="-1"></a>                       <span class="at">covar.train =</span> covar[ind.gwas, ], <span class="at">ncores =</span> NCORES)</span>
<span id="cb149-12"><a href="gwas-summary-statistics.html#cb149-12" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="fu">length</span>(ind.gwas)</span>
<span id="cb149-13"><a href="gwas-summary-statistics.html#cb149-13" tabindex="-1"></a>af <span class="ot">&lt;-</span> <span class="fu">big_colstats</span>(G, <span class="at">ind.row =</span> ind.gwas, <span class="at">ncores =</span> NCORES)<span class="sc">$</span>sum <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">*</span> N)</span></code></pre></div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="gwas-summary-statistics.html#cb150-1" tabindex="-1"></a>sd_y <span class="ot">&lt;-</span> <span class="fu">with</span>(gwas, <span class="fu">quantile</span>(<span class="fu">sqrt</span>(<span class="fl">0.5</span> <span class="sc">*</span> (N <span class="sc">*</span> std.err<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> estim<span class="sc">^</span><span class="dv">2</span>)), <span class="fl">0.01</span>))</span>
<span id="cb150-2"><a href="gwas-summary-statistics.html#cb150-2" tabindex="-1"></a><span class="fu">c</span>(<span class="fu">sd</span>(y[ind.gwas]), sd_y)</span></code></pre></div>
<pre><code>#&gt;                1% 
#&gt; 12.96917 11.84793</code></pre>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="gwas-summary-statistics.html#cb152-1" tabindex="-1"></a>sd_ss <span class="ot">&lt;-</span> sd_y <span class="sc">/</span> <span class="fu">with</span>(gwas, <span class="fu">sqrt</span>(N <span class="sc">*</span> std.err<span class="sc">^</span><span class="dv">2</span> <span class="sc">+</span> estim<span class="sc">^</span><span class="dv">2</span>))</span>
<span id="cb152-2"><a href="gwas-summary-statistics.html#cb152-2" tabindex="-1"></a>sd_af <span class="ot">&lt;-</span> <span class="fu">sqrt</span>(<span class="dv">2</span> <span class="sc">*</span> af <span class="sc">*</span> (<span class="dv">1</span> <span class="sc">-</span> af))</span>
<span id="cb152-3"><a href="gwas-summary-statistics.html#cb152-3" tabindex="-1"></a></span>
<span id="cb152-4"><a href="gwas-summary-statistics.html#cb152-4" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code></pre></div>
<pre><code>#&gt; Warning: package &#39;ggplot2&#39; was built under R version 4.2.3</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="gwas-summary-statistics.html#cb154-1" tabindex="-1"></a><span class="fu">ggplot</span>(dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="fu">data.frame</span>(sd_af, sd_ss), <span class="at">n =</span> <span class="fl">100e3</span>)) <span class="sc">+</span></span>
<span id="cb154-2"><a href="gwas-summary-statistics.html#cb154-2" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(sd_af, sd_ss), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb154-3"><a href="gwas-summary-statistics.html#cb154-3" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>(<span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb154-4"><a href="gwas-summary-statistics.html#cb154-4" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb154-5"><a href="gwas-summary-statistics.html#cb154-5" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Standard deviations derived from the allele frequencies&quot;</span>,</span>
<span id="cb154-6"><a href="gwas-summary-statistics.html#cb154-6" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Standard deviations derived from the summary statistics&quot;</span>)</span></code></pre></div>
<p><img src="sumstats_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb155-1"><a href="gwas-summary-statistics.html#cb155-1" tabindex="-1"></a>nb_na <span class="ot">&lt;-</span> <span class="fu">big_counts</span>(G<span class="sc">$</span><span class="fu">copy</span>(CODE_012), <span class="at">ind.row =</span> ind.gwas)[<span class="dv">4</span>, ]</span>
<span id="cb155-2"><a href="gwas-summary-statistics.html#cb155-2" tabindex="-1"></a><span class="fu">ggplot</span>(dplyr<span class="sc">::</span><span class="fu">slice_sample</span>(<span class="fu">data.frame</span>(sd_af, sd_ss, nb_na), <span class="at">n =</span> <span class="fl">100e3</span>)) <span class="sc">+</span></span>
<span id="cb155-3"><a href="gwas-summary-statistics.html#cb155-3" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(sd_af, sd_ss, <span class="at">color =</span> nb_na), <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb155-4"><a href="gwas-summary-statistics.html#cb155-4" tabindex="-1"></a>  <span class="fu">theme_bigstatsr</span>(<span class="fl">0.9</span>) <span class="sc">+</span></span>
<span id="cb155-5"><a href="gwas-summary-statistics.html#cb155-5" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">trans =</span> <span class="st">&quot;sqrt&quot;</span>) <span class="sc">+</span></span>
<span id="cb155-6"><a href="gwas-summary-statistics.html#cb155-6" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>) <span class="sc">+</span></span>
<span id="cb155-7"><a href="gwas-summary-statistics.html#cb155-7" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">&quot;Standard deviations derived from the allele frequencies&quot;</span>,</span>
<span id="cb155-8"><a href="gwas-summary-statistics.html#cb155-8" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">&quot;Standard deviations derived from the summary statistics&quot;</span>)</span></code></pre></div>
<p><img src="sumstats_files/figure-html/unnamed-chunk-10-1.png" width="70%" style="display: block; margin: auto;" /></p>
</details>
<hr />
<p>We will practice more QC on GWAS summary statistics in <a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2">8.5</a>.</p>
<p>I provide some example R script in <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">the LDpred2 tutorial</a> that implements this QC. You can also find some other scripts with examples how to prepare several GWAS summary statistics <a href="https://github.com/privefl/paper-misspec/tree/main/code">here</a>.</p>
<p>I am currently working on implementing a method that performs a complementary QC, as well as the imputation of GWAS summary statistics (i.e. GWAS results for more variants).</p>
</div>
<div id="ssancestry" class="section level2 hasAnchor" number="6.3">
<h2><span class="header-section-number">6.3</span> Ancestry inference<a href="gwas-summary-statistics.html#ssancestry" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>In <a href="population-structure.html#ancestry">4.5</a>, we’ve seen how to infer ancestry for individual-level data using reference data provided in <span class="citation">Privé (<a href="#ref-prive2021using">2022</a>)</span>; we can also use this to infer the ancestry composition of a GWAS dataset using only allele frequencies from the GWAS summary statistics. The tutorial accompanying <span class="citation">Privé (<a href="#ref-prive2021using">2022</a>)</span> is <a href="https://privefl.github.io/bigsnpr/articles/ancestry.html">here</a>.</p>
<div class="infobox exo">
<p>Reuse some of the code from <a href="population-structure.html#ancestry">4.5</a> and allele frequencies (that you should compute) to get the ancestry composition of this dataset using function <code>snp_ancestry_summary()</code>.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="gwas-summary-statistics.html#cb156-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb156-2"><a href="gwas-summary-statistics.html#cb156-2" tabindex="-1"></a>obj.bed <span class="ot">&lt;-</span> <span class="fu">bed</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.bed&quot;</span>)</span>
<span id="cb156-3"><a href="gwas-summary-statistics.html#cb156-3" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span></code></pre></div>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb157-1"><a href="gwas-summary-statistics.html#cb157-1" tabindex="-1"></a>all_freq <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb157-2"><a href="gwas-summary-statistics.html#cb157-2" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb157-3"><a href="gwas-summary-statistics.html#cb157-3" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019027&quot;</span>,  <span class="co"># subset for the tutorial (46 MB)</span></span>
<span id="cb157-4"><a href="gwas-summary-statistics.html#cb157-4" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620968&quot;,  # for real analyses (849 MB)</span></span>
<span id="cb157-5"><a href="gwas-summary-statistics.html#cb157-5" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;ref_freqs.csv.gz&quot;</span>))</span>
<span id="cb157-6"><a href="gwas-summary-statistics.html#cb157-6" tabindex="-1"></a></span>
<span id="cb157-7"><a href="gwas-summary-statistics.html#cb157-7" tabindex="-1"></a>projection <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb157-8"><a href="gwas-summary-statistics.html#cb157-8" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb157-9"><a href="gwas-summary-statistics.html#cb157-9" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019024&quot;</span>,  <span class="co"># subset for the tutorial (44 MB)</span></span>
<span id="cb157-10"><a href="gwas-summary-statistics.html#cb157-10" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620953&quot;,  # for real analyses (847 MB)</span></span>
<span id="cb157-11"><a href="gwas-summary-statistics.html#cb157-11" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;projection.csv.gz&quot;</span>))</span>
<span id="cb157-12"><a href="gwas-summary-statistics.html#cb157-12" tabindex="-1"></a></span>
<span id="cb157-13"><a href="gwas-summary-statistics.html#cb157-13" tabindex="-1"></a><span class="co"># coefficients to correct for overfitting of PCA</span></span>
<span id="cb157-14"><a href="gwas-summary-statistics.html#cb157-14" tabindex="-1"></a>correction <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">1.008</span>, <span class="fl">1.021</span>, <span class="fl">1.034</span>, <span class="fl">1.052</span>, <span class="fl">1.074</span>, <span class="fl">1.099</span>,</span>
<span id="cb157-15"><a href="gwas-summary-statistics.html#cb157-15" tabindex="-1"></a>                <span class="fl">1.123</span>, <span class="fl">1.15</span>, <span class="fl">1.195</span>, <span class="fl">1.256</span>, <span class="fl">1.321</span>, <span class="fl">1.382</span>, <span class="fl">1.443</span>)</span></code></pre></div>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="gwas-summary-statistics.html#cb158-1" tabindex="-1"></a><span class="co"># match variants between the two datasets</span></span>
<span id="cb158-2"><a href="gwas-summary-statistics.html#cb158-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb158-3"><a href="gwas-summary-statistics.html#cb158-3" tabindex="-1"></a>matched <span class="ot">&lt;-</span> obj.bed<span class="sc">$</span>map <span class="sc">%&gt;%</span></span>
<span id="cb158-4"><a href="gwas-summary-statistics.html#cb158-4" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">chr =</span> chromosome, <span class="at">pos =</span> physical.pos, <span class="at">a1 =</span> allele1, <span class="at">a0 =</span> allele2) <span class="sc">%&gt;%</span></span>
<span id="cb158-5"><a href="gwas-summary-statistics.html#cb158-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">beta =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb158-6"><a href="gwas-summary-statistics.html#cb158-6" tabindex="-1"></a>  <span class="fu">snp_match</span>(all_freq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb158-7"><a href="gwas-summary-statistics.html#cb158-7" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>#&gt;   chr     pos a0 a1 beta _NUM_ID_.ss       rsid _NUM_ID_
#&gt; 1   1  752566  G  A   -1           2  rs3094315        1
#&gt; 2   1  785989  T  C   -1           4  rs2980300        2
#&gt; 3   1  798959  G  A    1           5 rs11240777        3
#&gt; 4   1  947034  G  A   -1           6  rs2465126        4
#&gt; 5   1  949608  G  A    1           7     rs1921        5
#&gt; 6   1 1018704  A  G   -1           8  rs9442372        6
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 301150 rows ]</code></pre>
<div class="infobox caution">
<p>When matching between variants from two datasets, sometimes physical positions are in two different genome builds, usually it can be hg19 and hg38 (for newer datasets). You will need to either convert one of the two sets of positions with <code>snp_modifyBuild()</code> (that uses liftOver, not available for Windows) or by matching using rsIDs instead of positions (by using <code>join_by_pos = FALSE</code>).</p>
</div>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="gwas-summary-statistics.html#cb160-1" tabindex="-1"></a><span class="co"># further subsetting on missing values</span></span>
<span id="cb160-2"><a href="gwas-summary-statistics.html#cb160-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">bed_counts</span>(obj.bed, <span class="at">ind.col =</span> matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>, <span class="at">ncores =</span> NCORES)</span>
<span id="cb160-3"><a href="gwas-summary-statistics.html#cb160-3" tabindex="-1"></a>ind <span class="ot">&lt;-</span> <span class="fu">which</span>((counts[<span class="dv">4</span>, ] <span class="sc">/</span> <span class="fu">nrow</span>(obj.bed)) <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb160-4"><a href="gwas-summary-statistics.html#cb160-4" tabindex="-1"></a>matched2 <span class="ot">&lt;-</span> matched[ind, ]</span></code></pre></div>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="gwas-summary-statistics.html#cb161-1" tabindex="-1"></a><span class="co"># compute allele frequencies</span></span>
<span id="cb161-2"><a href="gwas-summary-statistics.html#cb161-2" tabindex="-1"></a>af <span class="ot">&lt;-</span> <span class="fu">bed_MAF</span>(obj.bed, <span class="at">ind.col =</span> matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>)<span class="sc">$</span>af</span>
<span id="cb161-3"><a href="gwas-summary-statistics.html#cb161-3" tabindex="-1"></a>af2 <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(matched2<span class="sc">$</span>beta <span class="sc">&gt;</span> <span class="dv">0</span>, af, <span class="dv">1</span> <span class="sc">-</span> af)</span>
<span id="cb161-4"><a href="gwas-summary-statistics.html#cb161-4" tabindex="-1"></a></span>
<span id="cb161-5"><a href="gwas-summary-statistics.html#cb161-5" tabindex="-1"></a><span class="co"># get ancestry composition</span></span>
<span id="cb161-6"><a href="gwas-summary-statistics.html#cb161-6" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">snp_ancestry_summary</span>(</span>
<span id="cb161-7"><a href="gwas-summary-statistics.html#cb161-7" tabindex="-1"></a>  <span class="at">freq =</span> af2,</span>
<span id="cb161-8"><a href="gwas-summary-statistics.html#cb161-8" tabindex="-1"></a>  <span class="at">info_freq_ref =</span> all_freq[matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>, <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)],</span>
<span id="cb161-9"><a href="gwas-summary-statistics.html#cb161-9" tabindex="-1"></a>  <span class="at">projection =</span> projection[matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>, <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)],</span>
<span id="cb161-10"><a href="gwas-summary-statistics.html#cb161-10" tabindex="-1"></a>  <span class="at">correction =</span> correction</span>
<span id="cb161-11"><a href="gwas-summary-statistics.html#cb161-11" tabindex="-1"></a>)</span>
<span id="cb161-12"><a href="gwas-summary-statistics.html#cb161-12" tabindex="-1"></a>res</span></code></pre></div>
<pre><code>#&gt;       Africa (West)      Africa (South)       Africa (East) 
#&gt;           0.0015918           0.0000000           0.0000000 
#&gt;      Africa (North)         Middle East           Ashkenazi 
#&gt;           0.0000000           0.0000000           0.0824672 
#&gt;               Italy Europe (South East) Europe (North East) 
#&gt;           0.2020482           0.1529422           0.0000000 
#&gt;             Finland         Scandinavia      United Kingdom 
#&gt;           0.0051918           0.1891938           0.1510849 
#&gt;             Ireland Europe (South West)       South America 
#&gt;           0.2154802           0.0000000           0.0000000 
#&gt;           Sri Lanka            Pakistan          Bangladesh 
#&gt;           0.0000000           0.0000000           0.0000000 
#&gt;         Asia (East)               Japan         Philippines 
#&gt;           0.0000000           0.0000000           0.0000000 
#&gt; attr(,&quot;cor_each&quot;)
#&gt;       Africa (West)      Africa (South)       Africa (East) 
#&gt;           0.6109985           0.6137032           0.8576549 
#&gt;      Africa (North)         Middle East           Ashkenazi 
#&gt;           0.9560451           0.9775514           0.9809802 
#&gt;               Italy Europe (South East) Europe (North East) 
#&gt;           0.9924371           0.9953612           0.9902400 
#&gt;             Finland         Scandinavia      United Kingdom 
#&gt;           0.9782142           0.9946803           0.9975349 
#&gt;             Ireland Europe (South West)       South America 
#&gt;           0.9956773           0.9949990           0.9002633 
#&gt;           Sri Lanka            Pakistan          Bangladesh 
#&gt;           0.8909370           0.9428874           0.8967786 
#&gt;         Asia (East)               Japan         Philippines 
#&gt;           0.7286470           0.7290332           0.7343768 
#&gt; attr(,&quot;cor_pred&quot;)
#&gt; [1] 0.9990652</code></pre>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb163-1"><a href="gwas-summary-statistics.html#cb163-1" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">colnames</span>(all_freq)[<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]</span>
<span id="cb163-2"><a href="gwas-summary-statistics.html#cb163-2" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Scandinavia&quot;</span>, <span class="st">&quot;United Kingdom&quot;</span>, <span class="st">&quot;Ireland&quot;</span>)]   <span class="ot">&lt;-</span> <span class="st">&quot;Europe (North West)&quot;</span></span>
<span id="cb163-3"><a href="gwas-summary-statistics.html#cb163-3" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Europe (South East)&quot;</span>, <span class="st">&quot;Europe (North East)&quot;</span>)] <span class="ot">&lt;-</span> <span class="st">&quot;Europe (East)&quot;</span></span>
<span id="cb163-4"><a href="gwas-summary-statistics.html#cb163-4" tabindex="-1"></a>grp_fct <span class="ot">&lt;-</span> <span class="fu">factor</span>(group, <span class="fu">unique</span>(group))</span>
<span id="cb163-5"><a href="gwas-summary-statistics.html#cb163-5" tabindex="-1"></a></span>
<span id="cb163-6"><a href="gwas-summary-statistics.html#cb163-6" tabindex="-1"></a>final_res <span class="ot">&lt;-</span> <span class="fu">tapply</span>(res, grp_fct, sum)</span>
<span id="cb163-7"><a href="gwas-summary-statistics.html#cb163-7" tabindex="-1"></a><span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> final_res, <span class="dv">1</span>)</span></code></pre></div>
<pre><code>#&gt;       Africa (West)      Africa (South)       Africa (East) 
#&gt;                 0.2                 0.0                 0.0 
#&gt;      Africa (North)         Middle East           Ashkenazi 
#&gt;                 0.0                 0.0                 8.2 
#&gt;               Italy       Europe (East)             Finland 
#&gt;                20.2                15.3                 0.5 
#&gt; Europe (North West) Europe (South West)       South America 
#&gt;                55.6                 0.0                 0.0 
#&gt;           Sri Lanka            Pakistan          Bangladesh 
#&gt;                 0.0                 0.0                 0.0 
#&gt;         Asia (East)               Japan         Philippines 
#&gt;                 0.0                 0.0                 0.0</code></pre>
</details>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-chen2021improved" class="csl-entry">
Chen, W., Wu, Y., Zheng, Z., Qi, T., Visscher, P.M., Zhu, Z., &amp; Yang, J. (2021). <a href="https://doi.org/10.1038/s41467-021-27438-7">Improved analyses of <span>GWAS</span> summary statistics by reducing data heterogeneity and errors</a>. <em>Nature Communications</em>, <em>12</em>, 1–10.
</div>
<div id="ref-gazal2018functional" class="csl-entry">
Gazal, S., Loh, P.-R., Finucane, H.K., Ganna, A., Schoech, A., Sunyaev, S., &amp; Price, A.L. (2018). <a href="https://doi.org/10.1038/s41588-018-0231-8">Functional architecture of low-frequency variants highlights strength of negative selection across coding and non-coding annotations</a>. <em>Nature Genetics</em>, <em>50</em>, 1600–1607.
</div>
<div id="ref-grotzinger2023pervasive" class="csl-entry">
Grotzinger, A.D., Fuente, J. de la, Privé, F., Nivard, M.G., &amp; Tucker-Drob, E.M. (2023). <a href="https://doi.org/10.1016/j.biopsych.2022.05.029">Pervasive downward bias in estimates of liability-scale heritability in genome-wide association study meta-analysis: A simple solution</a>. <em>Biological Psychiatry</em>, <em>93</em>, 29–36.
</div>
<div id="ref-julienne2021multitrait" class="csl-entry">
Julienne, H., Laville, V., McCaw, Z.R., He, Z., Guillemot, V., Lasry, C., et al. (2021). <a href="https://doi.org/10.1371/journal.pgen.1009713">Multitrait <span>GWAS</span> to connect disease variants and biological mechanisms</a>. <em>PLoS Genetics</em>, <em>17</em>, e1009713.
</div>
<div id="ref-kanai2022meta" class="csl-entry">
Kanai, M., Elzur, R., Zhou, W., Wu, K.-H.H., Rasheed, H., Tsuo, K., et al. (2022). <a href="https://doi.org/10.1016/j.xgen.2022.100210">Meta-analysis fine-mapping is often miscalibrated at single-variant resolution</a>. <em>Cell Genomics</em>, <em>2</em>, 100210.
</div>
<div id="ref-levey2021bi" class="csl-entry">
Levey, D.F., Stein, M.B., Wendt, F.R., Pathak, G.A., Zhou, H., Aslan, M., et al. (2021). <a href="https://doi.org/10.1038/s41593-021-00860-2">Bi-ancestral depression <span>GWAS</span> in the <span>Million Veteran Program</span> and meta-analysis in&gt; 1.2 million individuals highlight new therapeutic directions</a>. <em>Nature Neuroscience</em>, <em>24</em>, 954–963.
</div>
<div id="ref-murphy2021mungesumstats" class="csl-entry">
Murphy, A.E., Schilder, B.M., &amp; Skene, N.G. (2021). <a href="https://doi.org/10.1093/bioinformatics/btab665"><span>MungeSumstats</span>: A <span>Bioconductor</span> package for the standardization and quality control of many <span>GWAS</span> summary statistics</a>. <em>Bioinformatics</em>, <em>37</em>, 4593–4596.
</div>
<div id="ref-okbay2022polygenic" class="csl-entry">
Okbay, A., Wu, Y., Wang, N., Jayashankar, H., Bennett, M., Nehzati, S.M., et al. (2022). <a href="https://doi.org/10.1038/s41588-022-01016-z">Polygenic prediction of educational attainment within and between families from genome-wide association analyses in 3 million individuals</a>. <em>Nature Genetics</em>, <em>54</em>, 437–449.
</div>
<div id="ref-prive2021using" class="csl-entry">
Privé, F. (2022). <a href="https://doi.org/10.1093/bioinformatics/btac348"><span class="nocase">Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics</span></a>. <em>Bioinformatics</em>, <em>38</em>, 3477–3480.
</div>
<div id="ref-prive2021identifying" class="csl-entry">
Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136">Identifying and correcting for misspecifications in <span>GWAS</span> summary statistics and polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>, <em>3</em>, 100136.
</div>
<div id="ref-prive2020ldpred2" class="csl-entry">
Privé, F., Arbel, J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa1029"><span class="nocase"><span>LDpred2</span>: better, faster, stronger</span></a>. <em>Bioinformatics</em>, <em>36</em>, 5424–5431.
</div>
<div id="ref-suzuki2023multi" class="csl-entry">
Suzuki, K., Hatzikotoulas, K., Southam, L., Taylor, H.J., Yin, X., Lorenz, K.M., et al. (2023). Multi-ancestry genome-wide study in &gt; 2.5 million individuals reveals heterogeneity in mechanistic pathways of type 2 diabetes and complications. <em>medRxiv</em>. Retrieved from <a href="https://doi.org/10.1101/2023.03.31.23287839">https://doi.org/10.1101/2023.03.31.23287839</a>
</div>
<div id="ref-yengo2022saturated" class="csl-entry">
Yengo, L., Vedantam, S., Marouli, E., Sidorenko, J., Bartell, E., Sakaue, S., et al. (2022). <a href="https://doi.org/10.1038/s41586-022-05275-y">A saturated map of common genetic variants associated with human height</a>. <em>Nature</em>, <em>610</em>, 704–712.
</div>
<div id="ref-zhou2021global" class="csl-entry">
Zhou, W., Kanai, M., Wu, K.-H.H., Rasheed, H., Tsuo, K., Hirbo, J.B., et al. (2022). <a href="https://doi.org/10.1016/j.xgen.2022.100192"><span class="nocase">Global Biobank Meta-analysis Initiative</span>: Powering genetic discovery across human disease</a>. <em>Cell Genomics</em>, <em>2</em>, 100192.
</div>
<div id="ref-zou2022fine" class="csl-entry">
Zou, Y., Carbonetto, P., Wang, G., &amp; Stephens, M. (2022). <a href="https://doi.org/10.1371/journal.pgen.1010299">Fine-mapping from summary data with the <span class="nocase">"Sum of Single Effects"</span> model</a>. <em>PLoS Genetics</em>, <em>18</em>, e1010299.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="genome-wide-association-study-gwas.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="linkage-disequilibrium-ld.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
