<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Inputs and formats | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 2 Inputs and formats | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Inputs and formats | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Inputs and formats | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-07-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="preprocessing.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#main-author"><i class="fa fa-check"></i><b>1.3</b> Main author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#imputed-data"><i class="fa fa-check"></i><b>2.1</b> Imputed data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.2</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>2.3</b> Main motivation for developing R packages bigstatsr and bigsnpr</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.4</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.5</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.6</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.7" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.7</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.7.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.7.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.7.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.7.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.7.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.7.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.7.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.8</b> Exercise</a></li>
<li class="chapter" data-level="2.9" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#matching-genetic-variants-between-datasets"><i class="fa fa-check"></i><b>2.9</b> Matching genetic variants between datasets</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
<li class="chapter" data-level="5.4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses"><i class="fa fa-check"></i><b>5.4</b> Some post GWAS analyses</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#one-use-of-prs"><i class="fa fa-check"></i><b>8.1</b> One use of PRS</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.2</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.3</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.3.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.3.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.3.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.5</b> Future of PGS</a></li>
<li class="chapter" data-level="8.6" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.6</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.6.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.6.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.6.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.6.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.6.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create an LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="some-other-analyses.html"><a href="some-other-analyses.html"><i class="fa fa-check"></i><b>10</b> Some other analyses</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="inputs-and-formats" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Inputs and formats<a href="inputs-and-formats.html#inputs-and-formats" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Genetic data is represented as sequences of pairs of alleles (one from both parents). In the following figure, the reference alleles are shown in black, and the alternative alleles—arising from mutations—are shown in blue.
Here, three DNA positions show alternative alleles; these are called genetic variants. And we refer to them as common variants because the alternative allele is frequent enough in the population.
A particularly common type of genetic variant is the single nucleotide polymorphism (SNP), which involves a substitution of a single base at a specific genomic position.
We can encode this data into a numeric genotype matrix G by counting the number of alternative alleles per individual at each variant.</p>
<p><img src="images/overview_geneticvariants.png" width="65%" style="display: block; margin: auto;" /></p>
<div id="imputed-data" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Imputed data<a href="inputs-and-formats.html#imputed-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Genotype imputation is a widely used statistical technique that infers <em>untyped</em> genetic variants (possibly, dozens of millions) by leveraging known haplotype structures from external reference panels <span class="citation">(<a href="#ref-marchini2010genotype">Marchini &amp; Howie, 2010</a>)</span>. It increases variant density, improves power for association testing, and facilitates meta-analysis across studies genotyped on different arrays. Imputation is typically performed after genotype quality control and phasing, using reference panels such as the 1000 Genomes Project, the Haplotype Reference Consortium (HRC), or TOPMed, depending on the ancestral composition and genotyping density of the study cohort.</p>
<p>Several online imputation services are available, notably the Michigan Imputation Server <span class="citation">(<a href="#ref-das2016next">Das et al., 2016</a>)</span> and the Sanger Imputation Service, which provide streamlined pipelines for phasing (e.g., with SHAPEIT or Eagle) and imputation (e.g., using Minimac or IMPUTE). These platforms accept standard VCF inputs, automatically match strand orientation, and return imputed genotypes with associated quality metrics such as the INFO score. It is crucial to select a reference panel that closely matches the ancestry of the study population, as mismatches can reduce imputation accuracy and introduce bias in downstream analyses.</p>
</div>
<div id="formats-of-genetic-data" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Formats of genetic data<a href="inputs-and-formats.html#formats-of-genetic-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There exist many different data formats for genetic data:</p>
<ul>
<li><em>bed</em>/<em>bim</em>/<em>fam</em> files (also called PLINK1 files) that respectively store genotype calls only (0, 1, 2, or NA) in a very condensed way (binary using one byte only for 4 genotypes), information on the genetic variants (text), and information on the individuals (text). PLINK 1.9 and 2.0 <span class="citation">(<a href="#ref-chang2015second">Chang et al., 2015</a>)</span> provide many functions to work with this format.</li>
</ul>
<div class="infobox caution">
<p>The <em>bed</em> format presented here is not the same as <a href="https://en.wikipedia.org/wiki/BED_(file_format)">the <em>BED</em> format</a> (a text file format used to store genomic regions as coordinates and associated annotations).</p>
</div>
<ul>
<li><p><em>bgen</em> files (usually one per chromosome) that can store imputed probabilities (P(0), P(1), P(2)) that are often transformed to dosage information (expected values: P(1) + 2 P(2)). Each variant is stored compressed, which is very efficient, especially for low-frequency variants. They are accompanied by <em>bgen.bgi</em> files that store information on the genetic variants and the position of their corresponding data in the <em>bgen</em> files, and by a <em>sample</em> file that stores information on the individual IDs. This is the format used for the original UK Biobank imputed data <span class="citation">(<a href="#ref-bycroft2018uk">Bycroft et al., 2018</a>)</span>.</p></li>
<li><p><em>pgen</em>/<em>pvar</em>/<em>psam</em> files (also called PLINK2 files) that can store imputed data as well, and <em>pgen</em> files seem more compressed than <em>bgen</em> files. PLINK2 provides many functions to work with this format.</p></li>
<li><p>Many other formats that you can usually convert from using PLINK.</p></li>
</ul>
</div>
<div id="main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Main motivation for developing R packages bigstatsr and bigsnpr<a href="inputs-and-formats.html#main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>At the time, there was a notable lack of user-friendly and efficient R packages for genetic analyses, which posed challenges for researchers.
The existing workflows often required the use of disparate software tools with inconsistent input formats, reliance on text files for parameter settings, and limited compatibility with exploratory data analysis and familiar R packages.
Additionally, the development of new methods was hindered by the absence of tools supporting a simple matrix-like data structure.
To address these challenges, I initiated the development of the R package bigsnpr in 2016, reimplementing the statistical methods commonly used in genetic analyses within a cohesive and accessible R framework.</p>
<p>At some point, I realized that many functions (e.g. to perform genome-wide association studies (GWAS), principal component analysis (PCA), summary statistics, etc.) were not specific to genotype data. Indeed, both association studies and PCA are applicable to other omics data, such as transcriptomic or epigenetic datasets. Therefore I decided to move all these functions that could be used on any data stored as a matrix into a new R package, bigstatsr. This is why there are two packages, where bigstatsr can basically be used by any field using data stored as large numeric matrices, while bigsnpr provides some tools more specific to genotype data, largely building on top of bigstatsr. The initial description of the two packages is available in <span class="citation">Privé, Aschard, Ziyatdinov, &amp; Blum (<a href="#ref-prive2017efficient">2018</a>)</span>.</p>
</div>
<div id="the-bigsnp-format-from-bigsnpr" class="section level2 hasAnchor" number="2.4">
<h2><span class="header-section-number">2.4</span> The bigSNP format from bigsnpr<a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>My R package bigsnpr <span class="citation">(<a href="#ref-prive2017efficient">Privé et al., 2018</a>)</span> uses a class called <code>bigSNP</code> for representing SNP data. A <code>bigSNP</code> object is merely a list with the three following elements:</p>
<ul>
<li><code>$genotypes</code>: A <a href="https://privefl.github.io/bigstatsr/reference/FBM.code256-class.html"><code>FBM.code256</code></a>. Rows are samples and columns are genetic variants. This stores genotype calls or <em>dosages</em> (rounded to 2 decimal places). More about this format below.</li>
<li><code>$fam</code>: A <code>data.frame</code> with some information on the samples.</li>
<li><code>$map</code>: A <code>data.frame</code> with some information on the genetic variants.</li>
</ul>
<p>Package bigsnpr also provides functions for directly working on bed files with a small percentage of missing values <span class="citation">(<a href="#ref-prive2020efficient">Privé, Luu, Blum, McGrath, &amp; Vilhjálmsson, 2020</a>)</span>.</p>
</div>
<div id="getting-bigsnp" class="section level2 hasAnchor" number="2.5">
<h2><span class="header-section-number">2.5</span> Getting a bigSNP object<a href="inputs-and-formats.html#getting-bigsnp" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<ul>
<li><p>To read a bigSNP object from <em>bed</em>/<em>bim</em>/<em>fam</em> files, you can use functions <code>snp_readBed()</code> and <code>snp_readBed2()</code> (the second can read a subset of individuals/variants and use parallelism).</p></li>
<li><p>To read dosages from <em>BGEN</em> files, you can use function <code>snp_readBGEN()</code>. This function takes around 40 minutes to read 1M variants for 400K individuals using 15 cores. Note that this function currently works only for BGEN v1.2 with probabilities stored as 8 bits (cf. <a href="https://github.com/privefl/bigsnpr/issues/141#issuecomment-691285671">this issue</a>), which is the case for e.g. the UK Biobank files.</p></li>
<li><p>The previous functions create two new files: <code>&lt;backingfile&gt;.bk</code> (binary file that stores the content of the genetic matrix) and <code>&lt;backingfile&gt;.rds</code> (stores the R bigSNP object, including some information on how to link with the <em>bk</em> file).</p></li>
<li><p>To read any format used in genetics, you can always convert blocks of the data to text files using PLINK, read these using <code>bigreadr::fread2()</code>, and fill part of the resulting FBM. For example, see <a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/1-ricopili-to-bigsnp.R">the code I used to convert the iPSYCH imputed data from the RICOPILI pipeline to my bigSNP format</a>.</p></li>
</ul>
<p><strong>Example converting a bed file to bigSNP:</strong></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="inputs-and-formats.html#cb3-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="inputs-and-formats.html#cb5-1" tabindex="-1"></a>bedfile <span class="ot">&lt;-</span> <span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>, <span class="st">&quot;example.bed&quot;</span>, <span class="at">package =</span> <span class="st">&quot;bigsnpr&quot;</span>)</span></code></pre></div>
<div class="infobox exo">
<p>Before using some functions you don’t know, have a look at their documentation to see how to use them and some examples (<code>?snp_readBed</code>).</p>
<p>Use <code>snp_readBed()</code> to transform this data into a bigSNP object.
What do you get? Use <code>snp_attach()</code> to get the data into R and explore it a bit.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="inputs-and-formats.html#cb6-1" tabindex="-1"></a>(rds <span class="ot">&lt;-</span> <span class="fu">snp_readBed2</span>(bedfile, <span class="at">backingfile =</span> <span class="fu">tempfile</span>()))  <span class="co"># get path to new file</span></span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmpc5dgam\\file626c1cf44536.rds&quot;</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="inputs-and-formats.html#cb8-1" tabindex="-1"></a>bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(rds)  <span class="co"># can then read in the bigSNP object in any R session</span></span>
<span id="cb8-2"><a href="inputs-and-formats.html#cb8-2" tabindex="-1"></a>(G <span class="ot">&lt;-</span> bigsnp<span class="sc">$</span>genotypes)</span></code></pre></div>
<pre><code>#&gt; A Filebacked Big Matrix of type &#39;code 256&#39; with 517 rows and 4542 columns.</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="inputs-and-formats.html#cb10-1" tabindex="-1"></a>G[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]  <span class="co"># first 8 genotypes for first 5 individuals</span></span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; [1,]    0    0    2    0    1    1    0    2
#&gt; [2,]    1    0    1    0    0    1    0    2
#&gt; [3,]    0    1    1    0    2    1    0    2
#&gt; [4,]    0    0    2    0    2    1    0    2
#&gt; [5,]    1    0    0    0    2    2    1    0</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="inputs-and-formats.html#cb12-1" tabindex="-1"></a><span class="fu">str</span>(bigsnp<span class="sc">$</span>fam)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    517 obs. of  6 variables:
#&gt;  $ family.ID  : chr  &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; ...
#&gt;  $ sample.ID  : chr  &quot;IND0&quot; &quot;IND1&quot; &quot;IND2&quot; &quot;IND3&quot; ...
#&gt;  $ paternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ maternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ sex        : int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ affection  : int  1 1 2 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="inputs-and-formats.html#cb14-1" tabindex="-1"></a><span class="fu">str</span>(bigsnp<span class="sc">$</span>map)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    4542 obs. of  6 variables:
#&gt;  $ chromosome  : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ marker.ID   : chr  &quot;SNP0&quot; &quot;SNP1&quot; &quot;SNP2&quot; &quot;SNP3&quot; ...
#&gt;  $ genetic.dist: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ physical.pos: int  112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;  $ allele1     : chr  &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; ...
#&gt;  $ allele2     : chr  &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...</code></pre>
</details>
<hr />
<p><strong>Example directly mapping the bed file:</strong></p>
<div class="infobox exo">
<p>Map the bed file directly using <code>bed()</code>. Can you access the same data as with the bigSNP object?</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="inputs-and-formats.html#cb16-1" tabindex="-1"></a>obj.bed <span class="ot">&lt;-</span> <span class="fu">bed</span>(bedfile)</span>
<span id="cb16-2"><a href="inputs-and-formats.html#cb16-2" tabindex="-1"></a>obj.bed[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]  <span class="co"># first 8 genotypes for first 5 individuals</span></span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; [1,]    0    0    2    0    1    1    0    2
#&gt; [2,]    1    0    1    0    0    1    0    2
#&gt; [3,]    0    1    1    0    2    1    0    2
#&gt; [4,]    0    0    2    0    2    1    0    2
#&gt; [5,]    1    0    0    0    2    2    1    0</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="inputs-and-formats.html#cb18-1" tabindex="-1"></a><span class="fu">str</span>(obj.bed<span class="sc">$</span>fam)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    517 obs. of  6 variables:
#&gt;  $ family.ID  : chr  &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; &quot;POP1&quot; ...
#&gt;  $ sample.ID  : chr  &quot;IND0&quot; &quot;IND1&quot; &quot;IND2&quot; &quot;IND3&quot; ...
#&gt;  $ paternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ maternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ sex        : int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ affection  : int  1 1 2 1 1 1 1 1 1 1 ...</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="inputs-and-formats.html#cb20-1" tabindex="-1"></a><span class="fu">str</span>(obj.bed<span class="sc">$</span>map)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    4542 obs. of  6 variables:
#&gt;  $ chromosome  : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ marker.ID   : chr  &quot;SNP0&quot; &quot;SNP1&quot; &quot;SNP2&quot; &quot;SNP3&quot; ...
#&gt;  $ genetic.dist: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ physical.pos: int  112 1098 2089 3107 4091 5091 6107 7103 8090 9074 ...
#&gt;  $ allele1     : chr  &quot;A&quot; &quot;T&quot; &quot;T&quot; &quot;T&quot; ...
#&gt;  $ allele2     : chr  &quot;T&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...</code></pre>
</details>
<hr />
<p><strong>Example converting a bgen file to bigSNP:</strong></p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="inputs-and-formats.html#cb22-1" tabindex="-1"></a>bgen <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb22-2"><a href="inputs-and-formats.html#cb22-2" tabindex="-1"></a>  <span class="st">&quot;https://enkre.net/cgi-bin/code/bgen/raw/3ec770a829a753282b5cb45afc3f4eda036b246705b76f9037b6cc98c41a4194?at=example.8bits.bgen&quot;</span>,</span>
<span id="cb22-3"><a href="inputs-and-formats.html#cb22-3" tabindex="-1"></a>  <span class="at">fname =</span> <span class="st">&quot;example.bgen&quot;</span>)</span>
<span id="cb22-4"><a href="inputs-and-formats.html#cb22-4" tabindex="-1"></a>bgi <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb22-5"><a href="inputs-and-formats.html#cb22-5" tabindex="-1"></a>  <span class="st">&quot;https://enkre.net/cgi-bin/code/bgen/raw/dc7276e0f0e2e096f58d2dac645aa5711de2cd64c3b29a07a80575e175344f78?at=example.8bits.bgen.bgi&quot;</span>,</span>
<span id="cb22-6"><a href="inputs-and-formats.html#cb22-6" tabindex="-1"></a>  <span class="at">fname =</span> <span class="st">&quot;example.bgen.bgi&quot;</span>)</span></code></pre></div>
<div class="infobox exo">
<p>First, use <code>snp_readBGI()</code> to get information on the variants.</p>
<p>Then read the first 10 variants using <code>snp_readBGEN()</code> (pay attention to the specific format of <code>list_snp_id</code>), and <code>snp_attach()</code>.</p>
<p>What extra information about the variants do you get in the bigSNP object?</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="inputs-and-formats.html#cb23-1" tabindex="-1"></a>(var_info <span class="ot">&lt;-</span> <span class="fu">snp_readBGI</span>(bgi))</span></code></pre></div>
<pre><code>#&gt; # A tibble: 199 × 8
#&gt;    chromosome position rsid     number_of_alleles allele1 allele2
#&gt;    &lt;chr&gt;         &lt;int&gt; &lt;chr&gt;                &lt;int&gt; &lt;chr&gt;   &lt;chr&gt;  
#&gt;  1 01             1001 RSID_101                 2 A       G      
#&gt;  2 01             2000 RSID_2                   2 A       G      
#&gt;  3 01             2001 RSID_102                 2 A       G      
#&gt;  4 01             3000 RSID_3                   2 A       G      
#&gt;  5 01             3001 RSID_103                 2 A       G      
#&gt;  6 01             4000 RSID_4                   2 A       G      
#&gt;  7 01             4001 RSID_104                 2 A       G      
#&gt;  8 01             5000 RSID_5                   2 A       G      
#&gt;  9 01             5001 RSID_105                 2 A       G      
#&gt; 10 01             6000 RSID_6                   2 A       G      
#&gt; # ℹ 189 more rows
#&gt; # ℹ 2 more variables: file_start_position &lt;dbl&gt;, size_in_bytes &lt;int&gt;</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="inputs-and-formats.html#cb25-1" tabindex="-1"></a>(snp_id <span class="ot">&lt;-</span> <span class="fu">with</span>(var_info[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, ], <span class="fu">paste</span>(chromosome, position, allele1, allele2, <span class="at">sep =</span> <span class="st">&quot;_&quot;</span>)))</span></code></pre></div>
<pre><code>#&gt;  [1] &quot;01_1001_A_G&quot; &quot;01_2000_A_G&quot; &quot;01_2001_A_G&quot; &quot;01_3000_A_G&quot; &quot;01_3001_A_G&quot;
#&gt;  [6] &quot;01_4000_A_G&quot; &quot;01_4001_A_G&quot; &quot;01_5000_A_G&quot; &quot;01_5001_A_G&quot; &quot;01_6000_A_G&quot;</code></pre>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="inputs-and-formats.html#cb27-1" tabindex="-1"></a>(rds <span class="ot">&lt;-</span> <span class="fu">snp_readBGEN</span>(bgen, <span class="at">backingfile =</span> <span class="fu">tempfile</span>(), <span class="at">list_snp_id =</span> <span class="fu">list</span>(snp_id)))</span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\AppData\\Local\\Temp\\Rtmpc5dgam\\file626c390b316d.rds&quot;</code></pre>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="inputs-and-formats.html#cb29-1" tabindex="-1"></a>bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(rds)</span>
<span id="cb29-2"><a href="inputs-and-formats.html#cb29-2" tabindex="-1"></a>(G <span class="ot">&lt;-</span> bigsnp<span class="sc">$</span>genotypes)</span></code></pre></div>
<pre><code>#&gt; A Filebacked Big Matrix of type &#39;code 256&#39; with 500 rows and 10 columns.</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="inputs-and-formats.html#cb31-1" tabindex="-1"></a><span class="fu">str</span>(bigsnp<span class="sc">$</span>map)  <span class="co"># `$freq` and `$info` are computed when reading the data</span></span></code></pre></div>
<pre><code>#&gt; tibble [10 × 8] (S3: tbl_df/tbl/data.frame)
#&gt;  $ chromosome  : chr [1:10] &quot;01&quot; &quot;01&quot; &quot;01&quot; &quot;01&quot; ...
#&gt;  $ marker.ID   : chr [1:10] &quot;SNPID_101&quot; &quot;SNPID_2&quot; &quot;SNPID_102&quot; &quot;SNPID_3&quot; ...
#&gt;  $ rsid        : chr [1:10] &quot;RSID_101&quot; &quot;RSID_2&quot; &quot;RSID_102&quot; &quot;RSID_3&quot; ...
#&gt;  $ physical.pos: int [1:10] 1001 2000 2001 3000 3001 4000 4001 5000 5001 6000
#&gt;  $ allele1     : chr [1:10] &quot;A&quot; &quot;A&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;  $ allele2     : chr [1:10] &quot;G&quot; &quot;G&quot; &quot;G&quot; &quot;G&quot; ...
#&gt;  $ freq        : num [1:10] 0.583 0.802 0.198 0.483 0.517 ...
#&gt;  $ info        : num [1:10] 0.975 0.668 0.666 0.952 0.952 ...</code></pre>
</details>
<p><br></p>
<div class="infobox caution">
<p>There is no <code>$fam</code> information when reading BGEN files, which should be read from other data and matched/joined using IDs from the <em>sample</em> file. From the BGEN, you can also use these IDs to read a subset of individuals only, as done <a href="https://github.com/privefl/paper-misspec/blob/main/code/prepare-genotypes.R#L33-L34">in this code</a>.</p>
</div>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="inputs-and-formats.html#cb33-1" tabindex="-1"></a>sample <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb33-2"><a href="inputs-and-formats.html#cb33-2" tabindex="-1"></a>  <span class="st">&quot;https://enkre.net/cgi-bin/code/bgen/raw/a3c4d8e4c132048a502dc00a3e51362f98eda5a2889df695ba260dc48c327fd9?at=example.sample&quot;</span>,</span>
<span id="cb33-3"><a href="inputs-and-formats.html#cb33-3" tabindex="-1"></a>  <span class="at">fname =</span> <span class="st">&quot;example.sample&quot;</span>)</span>
<span id="cb33-4"><a href="inputs-and-formats.html#cb33-4" tabindex="-1"></a><span class="fu">readLines</span>(sample, <span class="at">n =</span> <span class="dv">6</span>)</span></code></pre></div>
<pre><code>#&gt; [1] &quot;ID_1&quot;       &quot;0&quot;          &quot;sample_001&quot; &quot;sample_002&quot; &quot;sample_003&quot;
#&gt; [6] &quot;sample_004&quot;</code></pre>
<div class="infobox caution">
<p>What should you be careful about when reading the <em>sample</em> file?</p>
</div>
</div>
<div id="the-fbm-format-from-bigstatsr" class="section level2 hasAnchor" number="2.6">
<h2><span class="header-section-number">2.6</span> The FBM format from bigstatsr<a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>The format provided in R package bigstatsr is called a Filebacked Big Matrix (FBM). It is an on-disk matrix format that is accessed through memory-mapping.</p>
<p><img src="https://raw.githubusercontent.com/privefl/R-presentation/master/memory-solution.svg" width="70%" style="display: block; margin: auto;" /></p>
<p>How memory-mapping works:</p>
<ul>
<li>when you access the 1st element (1st row, 1st col), it accesses a block (say the first column) from disk into memory (RAM)</li>
<li>when you access the 2nd element (2nd row, 1st col), it is already in memory so it is accessed very fast</li>
<li>when you access the second column, you access from disk again (once)</li>
<li>you can access many columns like that, until you do not have enough memory anymore</li>
<li>some space is freed automatically so that new columns can be accessed into memory</li>
<li>everything is seamlessly managed by the operating system (OS)</li>
<li>it is also very convenient for parallelism as data is shared between processes</li>
</ul>
<p>All the elements of an FBM have the same type; supported types are:</p>
<ul>
<li><code>"double"</code> (the default, double precision – 8 bytes per element)</li>
<li><code>"float"</code> (single precision – 4 bytes)</li>
<li><code>"integer"</code> (signed, so between <span class="math inline">\(\text{-}2^{31}\)</span> and (<span class="math inline">\(2^{31} \text{ - } 1\)</span>) – 4 bytes)</li>
<li><code>"unsigned short"</code>: can store integer values from <span class="math inline">\(0\)</span> to <span class="math inline">\(65535\)</span> (2 bytes)</li>
<li><code>"raw"</code> or <code>"unsigned char"</code>: can store integer values from <span class="math inline">\(0\)</span> to <span class="math inline">\(255\)</span> (1 byte). <strong>It is the basis for class <em>FBM.code256</em> that can access 256 arbitrary different numeric values (decoded using a <code>CODE_*</code>), which is used in bigsnpr.</strong></li>
</ul>
<p>The code used in class <em>FBM.code256</em> for imputed data is e.g. </p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="inputs-and-formats.html#cb35-1" tabindex="-1"></a>bigsnpr<span class="sc">::</span>CODE_DOSAGE</span></code></pre></div>
<pre><code>#&gt;   [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04 0.05 0.06 0.07
#&gt;  [16] 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18 0.19 0.20 0.21 0.22
#&gt;  [31] 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31 0.32 0.33 0.34 0.35 0.36 0.37
#&gt;  [46] 0.38 0.39 0.40 0.41 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.50 0.51 0.52
#&gt;  [61] 0.53 0.54 0.55 0.56 0.57 0.58 0.59 0.60 0.61 0.62 0.63 0.64 0.65 0.66 0.67
#&gt;  [76] 0.68 0.69 0.70 0.71 0.72 0.73 0.74 0.75 0.76 0.77 0.78 0.79 0.80 0.81 0.82
#&gt;  [91] 0.83 0.84 0.85 0.86 0.87 0.88 0.89 0.90 0.91 0.92 0.93 0.94 0.95 0.96 0.97
#&gt; [106] 0.98 0.99 1.00 1.01 1.02 1.03 1.04 1.05 1.06 1.07 1.08 1.09 1.10 1.11 1.12
#&gt; [121] 1.13 1.14 1.15 1.16 1.17 1.18 1.19 1.20 1.21 1.22 1.23 1.24 1.25 1.26 1.27
#&gt; [136] 1.28 1.29 1.30 1.31 1.32 1.33 1.34 1.35 1.36 1.37 1.38 1.39 1.40 1.41 1.42
#&gt; [151] 1.43 1.44 1.45 1.46 1.47 1.48 1.49 1.50 1.51 1.52 1.53 1.54 1.55 1.56 1.57
#&gt; [166] 1.58 1.59 1.60 1.61 1.62 1.63 1.64 1.65 1.66 1.67 1.68 1.69 1.70 1.71 1.72
#&gt; [181] 1.73 1.74 1.75 1.76 1.77 1.78 1.79 1.80 1.81 1.82 1.83 1.84 1.85 1.86 1.87
#&gt; [196] 1.88 1.89 1.90 1.91 1.92 1.93 1.94 1.95 1.96 1.97 1.98 1.99 2.00   NA   NA
#&gt; [211]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [226]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [241]   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA   NA
#&gt; [256]   NA</code></pre>
<p>where the first four elements are used to store genotype calls, the next three to store imputed allele counts, and the next 201 values to store dosages rounded to 2 decimal places. This allows for handling many data types (genotype calls and dosages) while storing each element using one byte only (x4 compared to bed files, but /8 compared to double-precision floating-point numbers).</p>
</div>
<div id="working-with-an-fbm" class="section level2 hasAnchor" number="2.7">
<h2><span class="header-section-number">2.7</span> Working with an FBM<a href="inputs-and-formats.html#working-with-an-fbm" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="similar-accessor-as-r-matrices" class="section level3 hasAnchor" number="2.7.1">
<h3><span class="header-section-number">2.7.1</span> Similar accessor as R matrices<a href="inputs-and-formats.html#similar-accessor-as-r-matrices" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="inputs-and-formats.html#cb37-1" tabindex="-1"></a><span class="fu">library</span>(bigstatsr)</span>
<span id="cb37-2"><a href="inputs-and-formats.html#cb37-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">FBM</span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="at">init =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">backingfile =</span> <span class="st">&quot;tmp-data/test&quot;</span>)<span class="sc">$</span><span class="fu">save</span>()</span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="inputs-and-formats.html#cb38-1" tabindex="-1"></a>X<span class="sc">$</span>backingfile  <span class="co"># the file where numeric data is stored</span></span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\OneDrive - Aarhus universitet\\Desktop\\statgen-course\\tmp-data\\test.bk&quot;</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="inputs-and-formats.html#cb40-1" tabindex="-1"></a>X<span class="sc">$</span>rds  <span class="co"># the file where information about the data is stored</span></span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\OneDrive - Aarhus universitet\\Desktop\\statgen-course\\tmp-data\\test.rds&quot;</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="inputs-and-formats.html#cb42-1" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">big_attach</span>(<span class="st">&quot;tmp-data/test.rds&quot;</span>)  <span class="co"># can get the FBM from any R session</span></span></code></pre></div>
<p>You can access the whole FBM as an R matrix in memory using <code>X[]</code>.
However, if the matrix is too large to fit in memory, you should always access only a subset of columns.
Note that the elements of an FBM are stored column-wise (as for a standard R matrix). Therefore, be careful not to access a subset of rows, since it would read non-contiguous elements from the whole matrix from disk.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="inputs-and-formats.html#cb43-1" tabindex="-1"></a>X[, <span class="dv">1</span>]  <span class="co"># ok (must read first column only)</span></span></code></pre></div>
<pre><code>#&gt; [1] 1 2</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="inputs-and-formats.html#cb45-1" tabindex="-1"></a>X[<span class="dv">1</span>, ]  <span class="co"># bad (must read all data from disk)</span></span></code></pre></div>
<pre><code>#&gt; [1] 1 3 5 7 9</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="inputs-and-formats.html#cb47-1" tabindex="-1"></a>X[]     <span class="co"># super bad (standard R matrix in memory)</span></span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5]
#&gt; [1,]    1    3    5    7    9
#&gt; [2,]    2    4    6    8   10</code></pre>
</div>
<div id="split-parapply-combine-strategy" class="section level3 hasAnchor" number="2.7.2">
<h3><span class="header-section-number">2.7.2</span> Split-(par)Apply-Combine Strategy<a href="inputs-and-formats.html#split-parapply-combine-strategy" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="inputs-and-formats.html#cb49-1" tabindex="-1"></a><span class="fu">colSums</span>(X[])  <span class="co"># super bad</span></span></code></pre></div>
<pre><code>#&gt; [1]  3  7 11 15 19</code></pre>
<p>Instead, the split-apply-combine strategy works well for applying standard R functions to FBMs (possibly in parallel), as implemented in <code>big_apply()</code>.</p>
<p><img src="https://raw.githubusercontent.com/privefl/R-presentation/master/split-apply-combine.svg" width="90%" style="display: block; margin: auto;" /></p>
<p>Learn more with <a href="https://privefl.github.io/bigstatsr/articles/big-apply.html">this tutorial on <code>big_apply()</code></a>.</p>
<div class="infobox exo">
<p>Compute the sum of each column of <code>X &lt;- big_attachExtdata()</code> using <code>big_apply()</code>.</p>
</div>
</div>
<div id="similar-accessor-as-rcpp-matrices" class="section level3 hasAnchor" number="2.7.3">
<h3><span class="header-section-number">2.7.3</span> Similar accessor as Rcpp matrices<a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>In case you want to develop new R functions for the FBM format while coding in C++ (which is the case for many bigstatsr/bigsnpr functions). Note that it is easy to use only a subset of the data without having to change anything in the code.</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode cpp"><code class="sourceCode cpp"><span id="cb51-1"><a href="inputs-and-formats.html#cb51-1" tabindex="-1"></a><span class="co">// [[Rcpp::plugins(cpp11)]]</span></span>
<span id="cb51-2"><a href="inputs-and-formats.html#cb51-2" tabindex="-1"></a><span class="co">// [[Rcpp::depends(bigstatsr, rmio)]]</span></span>
<span id="cb51-3"><a href="inputs-and-formats.html#cb51-3" tabindex="-1"></a><span class="pp">#include </span><span class="im">&lt;bigstatsr/BMCodeAcc.h&gt;</span></span>
<span id="cb51-4"><a href="inputs-and-formats.html#cb51-4" tabindex="-1"></a></span>
<span id="cb51-5"><a href="inputs-and-formats.html#cb51-5" tabindex="-1"></a><span class="co">// [[Rcpp::export]]</span></span>
<span id="cb51-6"><a href="inputs-and-formats.html#cb51-6" tabindex="-1"></a>NumericVector bigcolsums<span class="op">(</span>Environment BM<span class="op">,</span></span>
<span id="cb51-7"><a href="inputs-and-formats.html#cb51-7" tabindex="-1"></a>                         <span class="at">const</span> IntegerVector<span class="op">&amp;</span> rowInd<span class="op">,</span></span>
<span id="cb51-8"><a href="inputs-and-formats.html#cb51-8" tabindex="-1"></a>                         <span class="at">const</span> IntegerVector<span class="op">&amp;</span> colInd<span class="op">)</span> <span class="op">{</span></span>
<span id="cb51-9"><a href="inputs-and-formats.html#cb51-9" tabindex="-1"></a>  </span>
<span id="cb51-10"><a href="inputs-and-formats.html#cb51-10" tabindex="-1"></a>  <span class="co">// the external pointer</span></span>
<span id="cb51-11"><a href="inputs-and-formats.html#cb51-11" tabindex="-1"></a>  XPtr<span class="op">&lt;</span>FBM<span class="op">&gt;</span> xpBM <span class="op">=</span> BM<span class="op">[</span><span class="st">&quot;address&quot;</span><span class="op">];</span></span>
<span id="cb51-12"><a href="inputs-and-formats.html#cb51-12" tabindex="-1"></a>  <span class="co">// accessor to a sub-view of the data + automatically decoded</span></span>
<span id="cb51-13"><a href="inputs-and-formats.html#cb51-13" tabindex="-1"></a>  SubBMCode256Acc<span class="op">(</span>xpBM<span class="op">,</span> rowInd<span class="op">,</span> colInd<span class="op">,</span> BM<span class="op">[</span><span class="st">&quot;code256&quot;</span><span class="op">],</span> <span class="dv">1</span><span class="op">)</span></span>
<span id="cb51-14"><a href="inputs-and-formats.html#cb51-14" tabindex="-1"></a>  </span>
<span id="cb51-15"><a href="inputs-and-formats.html#cb51-15" tabindex="-1"></a>  <span class="dt">size_t</span> n <span class="op">=</span> macc<span class="op">.</span>nrow<span class="op">();</span>          <span class="co">// similar code as for an Rcpp::NumericMatrix</span></span>
<span id="cb51-16"><a href="inputs-and-formats.html#cb51-16" tabindex="-1"></a>  <span class="dt">size_t</span> m <span class="op">=</span> macc<span class="op">.</span>ncol<span class="op">();</span>          <span class="co">// similar code as for an Rcpp::NumericMatrix</span></span>
<span id="cb51-17"><a href="inputs-and-formats.html#cb51-17" tabindex="-1"></a>  </span>
<span id="cb51-18"><a href="inputs-and-formats.html#cb51-18" tabindex="-1"></a>  NumericVector res<span class="op">(</span>m<span class="op">);</span></span>
<span id="cb51-19"><a href="inputs-and-formats.html#cb51-19" tabindex="-1"></a>  </span>
<span id="cb51-20"><a href="inputs-and-formats.html#cb51-20" tabindex="-1"></a>  <span class="cf">for</span> <span class="op">(</span><span class="dt">size_t</span> j <span class="op">=</span> <span class="dv">0</span><span class="op">;</span> j <span class="op">&lt;</span> m<span class="op">;</span> j<span class="op">++)</span> </span>
<span id="cb51-21"><a href="inputs-and-formats.html#cb51-21" tabindex="-1"></a>    <span class="cf">for</span> <span class="op">(</span><span class="dt">size_t</span> i <span class="op">=</span> <span class="dv">0</span><span class="op">;</span> i <span class="op">&lt;</span> n<span class="op">;</span> i<span class="op">++)</span></span>
<span id="cb51-22"><a href="inputs-and-formats.html#cb51-22" tabindex="-1"></a>      res<span class="op">[</span>j<span class="op">]</span> <span class="op">+=</span> macc<span class="op">(</span>i<span class="op">,</span> j<span class="op">);</span>        <span class="co">// similar code as for an Rcpp::NumericMatrix</span></span>
<span id="cb51-23"><a href="inputs-and-formats.html#cb51-23" tabindex="-1"></a>  </span>
<span id="cb51-24"><a href="inputs-and-formats.html#cb51-24" tabindex="-1"></a>  <span class="cf">return</span> res<span class="op">;</span></span>
<span id="cb51-25"><a href="inputs-and-formats.html#cb51-25" tabindex="-1"></a><span class="op">}</span></span></code></pre></div>
</div>
<div id="some-summary-functions-are-already-implemented" class="section level3 hasAnchor" number="2.7.4">
<h3><span class="header-section-number">2.7.4</span> Some summary functions are already implemented<a href="inputs-and-formats.html#some-summary-functions-are-already-implemented" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="inputs-and-formats.html#cb52-1" tabindex="-1"></a>(X2 <span class="ot">&lt;-</span> <span class="fu">snp_attachExtdata</span>()<span class="sc">$</span>genotypes)</span></code></pre></div>
<pre><code>#&gt; A Filebacked Big Matrix of type &#39;code 256&#39; with 517 rows and 4542 columns.</code></pre>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="inputs-and-formats.html#cb54-1" tabindex="-1"></a><span class="fu">big_colstats</span>(X2)      <span class="co"># sum and var (for each column)</span></span></code></pre></div>
<pre><code>#&gt;    sum       var
#&gt; 1  354 0.4604831
#&gt; 2  213 0.3241195
#&gt; 3  245 0.3932122
#&gt; 4  191 0.3109247
#&gt; 5  472 0.5176030
#&gt; 6  368 0.4613528
#&gt; 7  132 0.2292594
#&gt; 8  497 0.4791207
#&gt; 9  498 0.5160886
#&gt; 10 481 0.5416535
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 4532 rows ]</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="inputs-and-formats.html#cb56-1" tabindex="-1"></a><span class="fu">big_scale</span>()(X2)       <span class="co"># mean and sd (for each column)</span></span></code></pre></div>
<pre><code>#&gt;       center     scale
#&gt; 1  0.6847195 0.6785891
#&gt; 2  0.4119923 0.5693149
#&gt; 3  0.4738878 0.6270663
#&gt; 4  0.3694391 0.5576062
#&gt; 5  0.9129594 0.7194463
#&gt; 6  0.7117988 0.6792295
#&gt; 7  0.2553191 0.4788104
#&gt; 8  0.9613153 0.6921855
#&gt; 9  0.9632495 0.7183931
#&gt; 10 0.9303675 0.7359712
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 4532 rows ]</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="inputs-and-formats.html#cb58-1" tabindex="-1"></a><span class="fu">snp_scaleBinom</span>()(X2)  <span class="co"># 2 * af (mean) and sqrt(2 * af * (1 - af)) (~sd)</span></span></code></pre></div>
<pre><code>#&gt;       center     scale
#&gt; 1  0.6847195 0.6710433
#&gt; 2  0.4119923 0.5719471
#&gt; 3  0.4738878 0.6013343
#&gt; 4  0.3694391 0.5488137
#&gt; 5  0.9129594 0.7044231
#&gt; 6  0.7117988 0.6771042
#&gt; 7  0.2553191 0.4719377
#&gt; 8  0.9613153 0.7065775
#&gt; 9  0.9632495 0.7066291
#&gt; 10 0.9303675 0.7053904
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 4532 rows ]</code></pre>
<div class="infobox info">
<p>Functions starting with <code>big_</code> are part of bigstatsr and work for any type of FBM, while functions starting with <code>snp_</code> or <code>bed_</code> are part of bigsnpr and are more specific to genetic data.</p>
<p>There are now many functions implemented in the packages. You can find a comprehensive list of available functions on the package website <a href="https://privefl.github.io/bigstatsr/reference/index.html">of bigstatsr</a> and <a href="https://privefl.github.io/bigsnpr/reference/index.html">of bigsnpr</a>.</p>
</div>
<div class="infobox exo">
<p>Take some time to have a look at the different functions available in both bigstatsr and bigsnpr.</p>
</div>
<div class="infobox question">
<p>What is the difference between functions starting with <code>snp_</code> vs <code>bed_</code>?</p>
</div>
<div class="infobox caution">
<p>To only use a subset of the data stored as an FBM, you should almost never make a copy of the data. Instead, to apply functions to a subset of the data, use their parameters <code>ind.row</code> (or <code>ind.train</code>) and <code>ind.col</code>.</p>
</div>
</div>
</div>
<div id="exercise" class="section level2 hasAnchor" number="2.8">
<h2><span class="header-section-number">2.8</span> Exercise<a href="inputs-and-formats.html#exercise" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="inputs-and-formats.html#cb60-1" tabindex="-1"></a>zip <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb60-2"><a href="inputs-and-formats.html#cb60-2" tabindex="-1"></a>  <span class="st">&quot;https://figshare.com/ndownloader/files/38019072&quot;</span>,</span>
<span id="cb60-3"><a href="inputs-and-formats.html#cb60-3" tabindex="-1"></a>  <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;GWAS_data.zip&quot;</span>)</span>
<span id="cb60-4"><a href="inputs-and-formats.html#cb60-4" tabindex="-1"></a><span class="fu">unzip</span>(zip, <span class="at">exdir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">overwrite =</span> <span class="cn">FALSE</span>)</span>
<span id="cb60-5"><a href="inputs-and-formats.html#cb60-5" tabindex="-1"></a>bedfile <span class="ot">&lt;-</span> <span class="st">&quot;tmp-data/GWAS_data.bed&quot;</span></span></code></pre></div>
<div class="infobox exo">
<p>Map this data with <code>bed()</code> and start exploring it a bit, e.g. what summaries you could compute, what functions you could use.</p>
</div>
</div>
<div id="matching-genetic-variants-between-datasets" class="section level2 hasAnchor" number="2.9">
<h2><span class="header-section-number">2.9</span> Matching genetic variants between datasets<a href="inputs-and-formats.html#matching-genetic-variants-between-datasets" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>This is some operation we will have to perform several times in this course.</p>
<p>When matching genetic variants across datasets—such as between GWAS summary statistics and a reference panel—it is essential to ensure that alleles are correctly aligned. Two common issues can lead to misalignment:</p>
<ul>
<li><p>Strand flips occur when alleles are reported on opposite DNA strands. For example, a SNP recorded as A/G in one dataset may appear as T/C in another. These are complements (A↔︎T, G↔︎C), and although they refer to the same variant, their labels differ due to strand orientation. If strand flips are not corrected, allele matching will fail.</p></li>
<li><p>Allele reversals happen when the effect and non-effect (or reference and alternate) alleles are swapped. For instance, if one dataset reports A/G and another G/A, the alleles are technically the same, but the interpretation of the effect direction (e.g., positive or negative beta) will be reversed. This can introduce sign errors in association results or polygenic scores.</p></li>
</ul>
<p>These issues are especially problematic for palindromic SNPs such as A/T or C/G, where the alleles are self-complementary. In such cases, even checking for strand flips or reversals is insufficient without additional information, such as allele frequencies, since both datasets may report A/T with no way to determine whether they refer to the same strand or orientation.</p>
<div class="infobox exo">
<p>Look carefully at the following results from <code>snp_match()</code> to understand what is going on.</p>
</div>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="inputs-and-formats.html#cb61-1" tabindex="-1"></a>sumstats <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb61-2"><a href="inputs-and-formats.html#cb61-2" tabindex="-1"></a>  <span class="at">chr =</span> <span class="dv">1</span>,</span>
<span id="cb61-3"><a href="inputs-and-formats.html#cb61-3" tabindex="-1"></a>  <span class="at">pos =</span> <span class="fu">c</span>(<span class="dv">86303</span>, <span class="dv">86331</span>, <span class="dv">162463</span>, <span class="dv">752566</span>, <span class="dv">755890</span>, <span class="dv">758144</span>),</span>
<span id="cb61-4"><a href="inputs-and-formats.html#cb61-4" tabindex="-1"></a>  <span class="at">a0 =</span> <span class="fu">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;C&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>),</span>
<span id="cb61-5"><a href="inputs-and-formats.html#cb61-5" tabindex="-1"></a>  <span class="at">a1 =</span> <span class="fu">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;A&quot;</span>),</span>
<span id="cb61-6"><a href="inputs-and-formats.html#cb61-6" tabindex="-1"></a>  <span class="at">beta =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">1.868</span>, <span class="fl">0.250</span>, <span class="sc">-</span><span class="fl">0.671</span>, <span class="fl">2.112</span>, <span class="fl">0.239</span>, <span class="fl">1.272</span>),</span>
<span id="cb61-7"><a href="inputs-and-formats.html#cb61-7" tabindex="-1"></a>  <span class="at">p =</span> <span class="fu">c</span>(<span class="fl">0.860</span>, <span class="fl">0.346</span>, <span class="fl">0.900</span>, <span class="fl">0.456</span>, <span class="fl">0.776</span>, <span class="fl">0.383</span>)</span>
<span id="cb61-8"><a href="inputs-and-formats.html#cb61-8" tabindex="-1"></a>)</span>
<span id="cb61-9"><a href="inputs-and-formats.html#cb61-9" tabindex="-1"></a></span>
<span id="cb61-10"><a href="inputs-and-formats.html#cb61-10" tabindex="-1"></a>info_snp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb61-11"><a href="inputs-and-formats.html#cb61-11" tabindex="-1"></a>  <span class="at">id =</span> <span class="fu">c</span>(<span class="st">&quot;rs2949417&quot;</span>, <span class="st">&quot;rs115209712&quot;</span>, <span class="st">&quot;rs143399298&quot;</span>, <span class="st">&quot;rs3094315&quot;</span>, <span class="st">&quot;rs3115858&quot;</span>),</span>
<span id="cb61-12"><a href="inputs-and-formats.html#cb61-12" tabindex="-1"></a>  <span class="at">chr =</span> <span class="dv">1</span>,</span>
<span id="cb61-13"><a href="inputs-and-formats.html#cb61-13" tabindex="-1"></a>  <span class="at">pos =</span> <span class="fu">c</span>(<span class="dv">86303</span>, <span class="dv">86331</span>, <span class="dv">162463</span>, <span class="dv">752566</span>, <span class="dv">755890</span>),</span>
<span id="cb61-14"><a href="inputs-and-formats.html#cb61-14" tabindex="-1"></a>  <span class="at">a0 =</span> <span class="fu">c</span>(<span class="st">&quot;T&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;T&quot;</span>),</span>
<span id="cb61-15"><a href="inputs-and-formats.html#cb61-15" tabindex="-1"></a>  <span class="at">a1 =</span> <span class="fu">c</span>(<span class="st">&quot;G&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>, <span class="st">&quot;G&quot;</span>, <span class="st">&quot;A&quot;</span>)</span>
<span id="cb61-16"><a href="inputs-and-formats.html#cb61-16" tabindex="-1"></a>)</span>
<span id="cb61-17"><a href="inputs-and-formats.html#cb61-17" tabindex="-1"></a></span>
<span id="cb61-18"><a href="inputs-and-formats.html#cb61-18" tabindex="-1"></a><span class="fu">snp_match</span>(sumstats, info_snp, <span class="at">return_flip_and_rev =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<pre><code>#&gt; 6 variants to be matched.</code></pre>
<pre><code>#&gt; 1 ambiguous SNPs have been removed.</code></pre>
<pre><code>#&gt; 4 variants have been matched; 1 were flipped and 1 were reversed.</code></pre>
<pre><code>#&gt;   chr    pos a0 a1   beta     p _NUM_ID_.ss _FLIP_ _REV_          id _NUM_ID_
#&gt; 1   1  86303  T  G -1.868 0.860           1  FALSE FALSE   rs2949417        1
#&gt; 2   1  86331  A  G -0.250 0.346           2  FALSE  TRUE rs115209712        2
#&gt; 3   1 162463  G  A -0.671 0.900           3   TRUE FALSE rs143399298        3
#&gt; 4   1 752566  A  G  2.112 0.456           4  FALSE FALSE   rs3094315        4</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="inputs-and-formats.html#cb66-1" tabindex="-1"></a><span class="fu">snp_match</span>(sumstats, info_snp, <span class="at">strand_flip =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>#&gt; 6 variants to be matched.</code></pre>
<pre><code>#&gt; 4 variants have been matched; 0 were flipped and 1 were reversed.</code></pre>
<pre><code>#&gt;   chr    pos a0 a1   beta     p _NUM_ID_.ss          id _NUM_ID_
#&gt; 1   1  86303  T  G -1.868 0.860           1   rs2949417        1
#&gt; 2   1  86331  A  G -0.250 0.346           2 rs115209712        2
#&gt; 3   1 752566  A  G  2.112 0.456           4   rs3094315        4
#&gt; 4   1 755890  T  A  0.239 0.776           5   rs3115858        5</code></pre>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bycroft2018uk" class="csl-entry">
Bycroft, C., Freeman, C., Petkova, D., Band, G., Elliott, L.T., Sharp, K., et al. (2018). <a href="https://doi.org/10.1038/s41586-018-0579-z">The <span>UK Biobank</span> resource with deep phenotyping and genomic data</a>. <em>Nature</em>, <em>562</em>, 203–209.
</div>
<div id="ref-chang2015second" class="csl-entry">
Chang, C.C., Chow, C.C., Tellier, L.C., Vattikuti, S., Purcell, S.M., &amp; Lee, J.J. (2015). <a href="https://doi.org/10.1186/s13742-015-0047-8">Second-generation <span>PLINK</span>: Rising to the challenge of larger and richer datasets</a>. <em>Gigascience</em>, <em>4</em>, s13742–015.
</div>
<div id="ref-das2016next" class="csl-entry">
Das, S., Forer, L., Schönherr, S., Sidore, C., Locke, A.E., Kwong, A., et al. (2016). <a href="https://doi.org/10.1038/ng.3656">Next-generation genotype imputation service and methods</a>. <em>Nature Genetics</em>, <em>48</em>, 1284–1287.
</div>
<div id="ref-marchini2010genotype" class="csl-entry">
Marchini, J., &amp; Howie, B. (2010). <a href="https://doi.org/10.1038/nrg2796">Genotype imputation for genome-wide association studies</a>. <em>Nature Reviews Genetics</em>, <em>11</em>, 499–511.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
Privé, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781–2787.
</div>
<div id="ref-prive2020efficient" class="csl-entry">
Privé, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449–4457.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="preprocessing.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
