<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 4 Population structure | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 4 Population structure | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 4 Population structure | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 4 Population structure | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-17" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="preprocessing.html"/>
<link rel="next" href="genome-wide-association-study-gwas.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i><b>1.3</b> Author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.3</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.4</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.5</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.5.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.5.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.5.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.5.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.5.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.5.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.5.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.1</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.2</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.2.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.2.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.2.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.3</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Future of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.5</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.5.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.5.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.5.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.5.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.5.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="population-structure" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">Chapter 4</span> Population structure<a href="population-structure.html#population-structure" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="principal-component-analysis-pca" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Principal Component Analysis (PCA)<a href="population-structure.html#principal-component-analysis-pca" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>PCA on the genotype matrix can be used to capture population structure.
However, PCA can actually capture different kinds of structure <span class="citation">(<a href="#ref-prive2020efficient">Privé, Luu, Blum, et al., 2020</a>)</span>:</p>
<ul>
<li><p>population structure (what we want),</p></li>
<li><p>Linkage Disequilibrium (LD) structure, when there are too many correlated variants (e.g. within long-range LD regions) and not enough population structure (see <a href="https://privefl.github.io/bigsnpr/articles/how-to-PCA.html">this vignette</a>),</p></li>
<li><p>relatedness structure, when there are related individuals who usually cluster together in later PCs,</p></li>
<li><p>noise, basically just circles when looking at PC scores.</p></li>
</ul>
<p>Capturing population structure with PCA is the second main topic of my research work (after polygenic scores).</p>
<ul>
<li><p>In <span class="citation">Privé et al. (<a href="#ref-prive2017efficient">2018</a>)</span>, I introduced an algorithm to compute PCA for a <code>bigSNP</code> object while accounting for the LD problem by using clumping (not pruning, cf. <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html">this vignette</a>) and an <em>automatic</em> detection and removal of long-range LD regions.</p></li>
<li><p>In <span class="citation">Privé, Luu, Vilhjálmsson, &amp; Blum (<a href="#ref-prive2020performing">2020</a>)</span>, I improved the implementation of the pcadapt algorithm, which detects variants associated with population structure (i.e. some kind of GWAS for population structure).</p></li>
<li><p>In <span class="citation">Privé, Luu, Blum, et al. (<a href="#ref-prive2020efficient">2020</a>)</span>, I extended bigsnpr to also be able to run PCA on PLINK bed files directly with a small percentage of missing values, and investigated best practices for PCA in more detail.</p></li>
<li><p>In <span class="citation">Privé, Aschard, et al. (<a href="#ref-prive2021high">2022</a>)</span> and <span class="citation">Privé (<a href="#ref-prive2021using">2022</a>)</span>, I showed how to use PCA for ancestry inference, including grouping individuals in homogeneous ancestry groups, and inferring ancestry proportions from genotype data but also from allele frequencies only (see <a href="https://privefl.github.io/bigsnpr/articles/ancestry.html">this vignette</a>).</p></li>
</ul>
</div>
<div id="the-problem-with-ld" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> The problem with LD<a href="population-structure.html#the-problem-with-ld" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let’s reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">3.3</a>.</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="population-structure.html#cb86-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb86-2"><a href="population-structure.html#cb86-2" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="population-structure.html#cb88-1" tabindex="-1"></a>bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb88-2"><a href="population-structure.html#cb88-2" tabindex="-1"></a>G <span class="ot">&lt;-</span> bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb88-3"><a href="population-structure.html#cb88-3" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span></code></pre></div>
<div class="infobox exo">
<p>Use <code>big_randomSVD()</code> to perform a SVD/PCA on <code>G</code>. Do not forget to use some scaling (at least some centering).</p>
<p>Look at PC scores and PC loadings with <code>plot()</code>. What do you see for PC4? Color PC scores using the most associated variant with PC4.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="population-structure.html#cb89-1" tabindex="-1"></a>svd <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb89-2"><a href="population-structure.html#cb89-2" tabindex="-1"></a>  <span class="fu">big_randomSVD</span>(G, <span class="at">fun.scaling =</span> <span class="fu">big_scale</span>(), <span class="at">ncores =</span> NCORES),</span>
<span id="cb89-3"><a href="population-structure.html#cb89-3" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/svd_with_ld.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;    0.73    0.27   59.08 
#&gt; Code finished running at 2025-06-10 13:33:51 CEST</code></pre>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="population-structure.html#cb91-1" tabindex="-1"></a><span class="fu">plot</span>(svd)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="population-structure.html#cb92-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-2.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="population-structure.html#cb93-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-3.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="population-structure.html#cb94-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;loadings&quot;</span>, <span class="at">loadings =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">coef =</span> <span class="fl">0.4</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-4.png" width="70%" style="display: block; margin: auto;" /></p>
<p>What is going on with PC4?</p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="population-structure.html#cb95-1" tabindex="-1"></a>the_max <span class="ot">&lt;-</span> <span class="fu">which.max</span>(<span class="fu">abs</span>(svd<span class="sc">$</span>v[, <span class="dv">4</span>]))</span>
<span id="cb95-2"><a href="population-structure.html#cb95-2" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb95-3"><a href="population-structure.html#cb95-3" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">as.factor</span>(G[, the_max])) <span class="sc">+</span> </span>
<span id="cb95-4"><a href="population-structure.html#cb95-4" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">&quot;Genotype at most influential SNP&quot;</span>) <span class="sc">+</span></span>
<span id="cb95-5"><a href="population-structure.html#cb95-5" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">3</span>))) <span class="sc">+</span></span>
<span id="cb95-6"><a href="population-structure.html#cb95-6" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;bottom&quot;</span>,</span>
<span id="cb95-7"><a href="population-structure.html#cb95-7" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">r =</span> <span class="dv">6</span>)))</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-3-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>PC4 is basically capturing variation in a block of LD (the peak you see for loadings of PC4) and is then very correlated to genetic variants in this region.</p>
</details>
<hr />
<div class="infobox caution">
<p>You really want to avoid using PCs that capture LD e.g. as covariates in GWAS because it will cause some collider bias <span class="citation">(<a href="#ref-grinde2024adjusting">Grinde, Browning, Reiner, Thornton, &amp; Browning, 2024</a>; <a href="#ref-prive2021identifying">Privé, Arbel, Aschard, &amp; Vilhjálmsson, 2022</a>)</span>.</p>
</div>
<p>To avoid capturing LD in PCA, it has been recommended to perform some pruning (removing variants too correlated with one another) AND remove some known list of long-range LD regions (that can still be captured by PCA, even after a stringent pruning step) <span class="citation">(<a href="#ref-abdellaoui2013population">Abdellaoui et al., 2013</a>; <a href="#ref-price2008long">Price et al., 2008</a>)</span>. But this is not enough; this is exactly what the UK Biobank did <span class="citation">(<a href="#ref-bycroft2018uk">Bycroft et al., 2018</a>)</span>, and this how the PC loadings of the 40 PCs they provide look like:</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/loadings-ukbb.jpeg?raw=true" width="50%" style="display: block; margin: auto;" /></p>
<p><strong>I recommend using only the first 16 PCs provided by the UK Biobank <span class="citation">(<a href="#ref-prive2020efficient">Privé, Luu, Blum, et al., 2020</a>)</span>.</strong> If you want to compute PCs yourself, I recommend using the autoSVD functions I developed that perform an <em>automatic</em> detection and removal of long-range LD regions.</p>
</div>
<div id="best-practices-for-pca-of-genetic-data" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Best practices for PCA of genetic data<a href="population-structure.html#best-practices-for-pca-of-genetic-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There can be many steps to properly perform a PCA; you can find more about this in <span class="citation">Privé, Luu, Blum, et al. (<a href="#ref-prive2020efficient">2020</a>)</span>.
Let’s have a look at <a href="https://privefl.github.io/bigsnpr/articles/bedpca.html">the corresponding tutorial from the bigsnpr website</a>.</p>
</div>
<div id="exo-pca" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Exercise<a href="population-structure.html#exo-pca" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="infobox exo">
<p>Let’s reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">3.3</a>.</p>
<p>Follow the previous tutorial to perform a PCA for this data, using either <code>snp_*</code> functions on the bigSNP object or <code>bed_*</code> functions on the bed file mapped.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<p>First, let’s get an idea of the relatedness in the data using</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="population-structure.html#cb96-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb96-2"><a href="population-structure.html#cb96-2" tabindex="-1"></a>(NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>())</span>
<span id="cb96-3"><a href="population-structure.html#cb96-3" tabindex="-1"></a>plink2 <span class="ot">&lt;-</span> <span class="fu">download_plink2</span>(<span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb96-4"><a href="population-structure.html#cb96-4" tabindex="-1"></a>rel <span class="ot">&lt;-</span> <span class="fu">snp_plinkKINGQC</span>(plink2, <span class="st">&quot;tmp-data/GWAS_data_sorted_QC.bed&quot;</span>,</span>
<span id="cb96-5"><a href="population-structure.html#cb96-5" tabindex="-1"></a>                       <span class="at">thr.king =</span> <span class="dv">2</span><span class="sc">^-</span><span class="fl">4.5</span>, <span class="at">make.bed =</span> <span class="cn">FALSE</span>, <span class="at">ncores =</span> NCORES)</span></code></pre></div>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="population-structure.html#cb97-1" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(rel<span class="sc">$</span>KINSHIP), <span class="st">&quot;FD&quot;</span>); <span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">4.5</span>, <span class="sc">-</span><span class="fl">3.5</span>), <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-6-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="infobox info">
<p>When computing relatedness with KING, LD pruning is <em>NOT</em> recommended. However, it may be useful to filter out some variants that are highly associated with population structure, e.g. as performed in the UK Biobank <span class="citation">(<a href="#ref-bycroft2018uk">Bycroft et al., 2018</a>)</span>. For example, see <a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/2-rel-and-pca.R">this code</a>.</p>
</div>
<p>Relatedness should not be a major issue here. Let’s now compute PCs.</p>
<div class="infobox info">
<p>All the code that follows could be run on the bigSNP object we made before. Nevertheless, to showcase the <code>bed_*</code> functions here, we will run the following analyses on the bed file directly.</p>
</div>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="population-structure.html#cb98-1" tabindex="-1"></a><span class="co"># Memory-map a bed file directly</span></span>
<span id="cb98-2"><a href="population-structure.html#cb98-2" tabindex="-1"></a>obj.bed <span class="ot">&lt;-</span> <span class="fu">bed</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.bed&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="population-structure.html#cb99-1" tabindex="-1"></a>obj.svd <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb99-2"><a href="population-structure.html#cb99-2" tabindex="-1"></a>  <span class="fu">bed_autoSVD</span>(obj.bed, <span class="at">k =</span> <span class="dv">12</span>, <span class="at">ncores =</span> NCORES),</span>
<span id="cb99-3"><a href="population-structure.html#cb99-3" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/PCA_GWAS_data.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;   80.42    1.28  392.85 
#&gt; Code finished running at 2025-06-10 13:41:05 CEST</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="population-structure.html#cb101-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="population-structure.html#cb102-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">coeff =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-9-2.png" width="70%" style="display: block; margin: auto;" /></p>
<p>There is some population structure (maybe up to 6 PCs). You should also check loadings to make sure there is no LD structure (peaks on loadings):</p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="population-structure.html#cb103-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd, <span class="at">type =</span> <span class="st">&quot;loadings&quot;</span>, <span class="at">loadings =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">coeff =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-10-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>No peaks, but loadings of PC4 are a bit odd.</p>
<p>If you expect the individuals to mostly come from one population (e.g. in a national biobank), you can simply use a robust distance to identify a homogeneous subset of individuals, then look at the histogram of log-distances to choose a threshold based on visual inspection (here I would probably choose 4.5).</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="population-structure.html#cb104-1" tabindex="-1"></a>PC <span class="ot">&lt;-</span> <span class="fu">predict</span>(obj.svd)</span>
<span id="cb104-2"><a href="population-structure.html#cb104-2" tabindex="-1"></a>ldist <span class="ot">&lt;-</span> <span class="fu">log</span>(bigutilsr<span class="sc">::</span><span class="fu">dist_ogk</span>(PC[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]))</span>
<span id="cb104-3"><a href="population-structure.html#cb104-3" tabindex="-1"></a><span class="fu">hist</span>(ldist, <span class="st">&quot;FD&quot;</span>); <span class="fu">abline</span>(<span class="at">v =</span> <span class="fl">4.5</span>, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-11-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="population-structure.html#cb105-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb105-2"><a href="population-structure.html#cb105-2" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/privefl/paper4-bedpca/master/code/plot_grid2.R&quot;</span>)</span>
<span id="cb105-3"><a href="population-structure.html#cb105-3" tabindex="-1"></a></span>
<span id="cb105-4"><a href="population-structure.html#cb105-4" tabindex="-1"></a><span class="fu">plot_grid2</span>(<span class="at">plotlist =</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(k) {</span>
<span id="cb105-5"><a href="population-structure.html#cb105-5" tabindex="-1"></a>  k1 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb105-6"><a href="population-structure.html#cb105-6" tabindex="-1"></a>  k2 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k</span>
<span id="cb105-7"><a href="population-structure.html#cb105-7" tabindex="-1"></a>  <span class="fu">qplot</span>(PC[, k1], PC[, k2], <span class="at">color =</span> ldist, <span class="at">size =</span> <span class="fu">I</span>(<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb105-8"><a href="population-structure.html#cb105-8" tabindex="-1"></a>    <span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb105-9"><a href="population-structure.html#cb105-9" tabindex="-1"></a>    <span class="fu">theme_bigstatsr</span>(<span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb105-10"><a href="population-structure.html#cb105-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k1), <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k2), <span class="at">color =</span> <span class="st">&quot;log-distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb105-11"><a href="population-structure.html#cb105-11" tabindex="-1"></a>    <span class="fu">coord_equal</span>()</span>
<span id="cb105-12"><a href="population-structure.html#cb105-12" tabindex="-1"></a>}), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">legend_ratio =</span> <span class="fl">0.2</span>, <span class="at">title_ratio =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>#&gt; Warning: `qplot()` was deprecated in ggplot2 3.4.0.
#&gt; This warning is displayed once every 8 hours.
#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#&gt; generated.</code></pre>
<p><img src="popstruct_files/figure-html/unnamed-chunk-12-1.png" width="70%" style="display: block; margin: auto;" /></p>
</details>
</div>
<div id="ancestry" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> Ancestry inference<a href="population-structure.html#ancestry" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>It would be nice if we could get a better sense of the ancestry of these individuals.
To achieve this, we can project this data onto the PCA space of many known population groups defined in <span class="citation">Privé (<a href="#ref-prive2021using">2022</a>)</span> (based on the UK Biobank and some individuals from the 1000 Genomes).</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="population-structure.html#cb107-1" tabindex="-1"></a>all_freq <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb107-2"><a href="population-structure.html#cb107-2" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb107-3"><a href="population-structure.html#cb107-3" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019027&quot;</span>,  <span class="co"># subset for the tutorial (46 MB)</span></span>
<span id="cb107-4"><a href="population-structure.html#cb107-4" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620968&quot;,  # for real analyses (849 MB)</span></span>
<span id="cb107-5"><a href="population-structure.html#cb107-5" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;ref_freqs.csv.gz&quot;</span>))</span>
<span id="cb107-6"><a href="population-structure.html#cb107-6" tabindex="-1"></a></span>
<span id="cb107-7"><a href="population-structure.html#cb107-7" tabindex="-1"></a>projection <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb107-8"><a href="population-structure.html#cb107-8" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb107-9"><a href="population-structure.html#cb107-9" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019024&quot;</span>,  <span class="co"># subset for the tutorial (44 MB)</span></span>
<span id="cb107-10"><a href="population-structure.html#cb107-10" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620953&quot;,  # for real analyses (847 MB)</span></span>
<span id="cb107-11"><a href="population-structure.html#cb107-11" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;projection.csv.gz&quot;</span>))</span>
<span id="cb107-12"><a href="population-structure.html#cb107-12" tabindex="-1"></a></span>
<span id="cb107-13"><a href="population-structure.html#cb107-13" tabindex="-1"></a><span class="co"># coefficients to correct for overfitting of PCA</span></span>
<span id="cb107-14"><a href="population-structure.html#cb107-14" tabindex="-1"></a>correction <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">1.008</span>, <span class="fl">1.021</span>, <span class="fl">1.034</span>, <span class="fl">1.052</span>, <span class="fl">1.074</span>, <span class="fl">1.099</span>,</span>
<span id="cb107-15"><a href="population-structure.html#cb107-15" tabindex="-1"></a>                <span class="fl">1.123</span>, <span class="fl">1.15</span>, <span class="fl">1.195</span>, <span class="fl">1.256</span>, <span class="fl">1.321</span>, <span class="fl">1.382</span>, <span class="fl">1.443</span>)</span></code></pre></div>
<div class="sourceCode" id="cb108"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="population-structure.html#cb108-1" tabindex="-1"></a><span class="fu">str</span>(all_freq)    <span class="co"># reference allele frequencies</span></span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    301156 obs. of  26 variables:
#&gt;  $ chr                : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ pos                : int  752566 785989 798959 947034 949608 1018704 1041700 1129672 1130727 1165310 ...
#&gt;  $ a0                 : chr  &quot;G&quot; &quot;T&quot; &quot;G&quot; &quot;G&quot; ...
#&gt;  $ a1                 : chr  &quot;A&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;  $ rsid               : chr  &quot;rs3094315&quot; &quot;rs2980300&quot; &quot;rs11240777&quot; &quot;rs2465126&quot; ...
#&gt;  $ Africa (West)      : num  0.307 0.179 0.765 0.493 0.437 ...
#&gt;  $ Africa (South)     : num  0.417 0.233 0.737 0.414 0.352 ...
#&gt;  $ Africa (East)      : num  0.547 0.493 0.608 0.597 0.261 ...
#&gt;  $ Africa (North)     : num  0.662 0.645 0.427 0.762 0.319 ...
#&gt;  $ Middle East        : num  0.794 0.821 0.317 0.886 0.295 ...
#&gt;  $ Ashkenazi          : num  0.775 0.84 0.203 0.899 0.298 ...
#&gt;  $ Italy              : num  0.806 0.864 0.243 0.903 0.357 ...
#&gt;  $ Europe (South East): num  0.842 0.872 0.201 0.935 0.408 ...
#&gt;  $ Europe (North East): num  0.783 0.825 0.203 0.967 0.38 ...
#&gt;  $ Finland            : num  0.812 0.831 0.224 0.969 0.483 ...
#&gt;  $ Scandinavia        : num  0.819 0.843 0.228 0.964 0.412 ...
#&gt;  $ United Kingdom     : num  0.839 0.872 0.2 0.963 0.401 ...
#&gt;  $ Ireland            : num  0.86 0.887 0.202 0.973 0.387 ...
#&gt;  $ Europe (South West): num  0.828 0.867 0.216 0.932 0.347 ...
#&gt;  $ South America      : num  0.775 0.722 0.324 0.928 0.244 ...
#&gt;  $ Sri Lanka          : num  0.749 0.753 0.384 0.967 0.293 ...
#&gt;  $ Pakistan           : num  0.781 0.79 0.347 0.943 0.338 ...
#&gt;  $ Bangladesh         : num  0.799 0.808 0.33 0.975 0.325 ...
#&gt;  $ Asia (East)        : num  0.887 0.758 0.319 0.953 0.191 ...
#&gt;  $ Japan              : num  0.855 0.67 0.397 0.974 0.167 ...
#&gt;  $ Philippines        : num  0.874 0.845 0.224 0.989 0.212 ...</code></pre>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="population-structure.html#cb110-1" tabindex="-1"></a><span class="fu">str</span>(projection)  <span class="co"># PC loadings of the reference data</span></span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    301156 obs. of  21 variables:
#&gt;  $ chr : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ pos : int  752566 785989 798959 947034 949608 1018704 1041700 1129672 1130727 1165310 ...
#&gt;  $ a0  : chr  &quot;G&quot; &quot;T&quot; &quot;G&quot; &quot;G&quot; ...
#&gt;  $ a1  : chr  &quot;A&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;  $ rsid: chr  &quot;rs3094315&quot; &quot;rs2980300&quot; &quot;rs11240777&quot; &quot;rs2465126&quot; ...
#&gt;  $ PC1 : num  8.73e-05 5.91e-05 3.81e-05 8.08e-05 9.40e-05 ...
#&gt;  $ PC2 : num  -1.25e-04 -1.22e-04 1.54e-04 -1.01e-04 2.02e-05 ...
#&gt;  $ PC3 : num  -5.23e-05 -1.25e-05 1.02e-05 -3.68e-05 1.37e-04 ...
#&gt;  $ PC4 : num  -3.58e-05 1.01e-06 3.73e-05 -1.24e-05 -1.17e-04 ...
#&gt;  $ PC5 : num  2.69e-06 -7.52e-06 4.28e-05 4.53e-05 1.52e-04 ...
#&gt;  $ PC6 : num  3.22e-05 2.57e-05 -1.12e-05 1.91e-05 1.76e-04 ...
#&gt;  $ PC7 : num  -1.64e-05 1.86e-05 -1.21e-04 5.86e-05 1.39e-04 ...
#&gt;  $ PC8 : num  2.03e-05 7.16e-05 -1.05e-04 1.00e-05 3.29e-05 ...
#&gt;  $ PC9 : num  -7.71e-05 -3.57e-05 2.06e-05 -1.10e-05 1.57e-04 ...
#&gt;  $ PC10: num  7.27e-06 1.28e-05 1.65e-05 1.68e-05 1.78e-05 ...
#&gt;  $ PC11: num  1.35e-04 4.22e-05 -1.22e-05 -5.77e-05 -1.60e-04 ...
#&gt;  $ PC12: num  -6.57e-05 -4.03e-05 4.44e-05 -2.68e-06 7.79e-05 ...
#&gt;  $ PC13: num  -6.45e-05 -1.94e-05 5.16e-05 3.27e-05 5.64e-05 ...
#&gt;  $ PC14: num  -3.13e-05 9.18e-09 -4.06e-05 2.15e-05 -1.59e-04 ...
#&gt;  $ PC15: num  1.16e-05 -5.76e-06 1.68e-05 -1.26e-06 -7.81e-05 ...
#&gt;  $ PC16: num  6.42e-05 6.47e-05 -4.77e-05 2.51e-05 -9.73e-07 ...</code></pre>
<div class="infobox exo">
<p>Match variants between <code>obj.bed</code> and <code>all_freq</code> using <code>snp_match()</code>.</p>
<p>For the variants matched, further remove the variants with more than 5% of missing values.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="population-structure.html#cb112-1" tabindex="-1"></a><span class="co"># match variants between the two datasets</span></span>
<span id="cb112-2"><a href="population-structure.html#cb112-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb112-3"><a href="population-structure.html#cb112-3" tabindex="-1"></a>matched <span class="ot">&lt;-</span> obj.bed<span class="sc">$</span>map <span class="sc">%&gt;%</span></span>
<span id="cb112-4"><a href="population-structure.html#cb112-4" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">chr =</span> chromosome, <span class="at">pos =</span> physical.pos, <span class="at">a1 =</span> allele1, <span class="at">a0 =</span> allele2) <span class="sc">%&gt;%</span></span>
<span id="cb112-5"><a href="population-structure.html#cb112-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">beta =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb112-6"><a href="population-structure.html#cb112-6" tabindex="-1"></a>  <span class="fu">snp_match</span>(all_freq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb112-7"><a href="population-structure.html#cb112-7" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>#&gt;   chr     pos a0 a1 beta _NUM_ID_.ss       rsid _NUM_ID_
#&gt; 1   1  752566  G  A   -1           2  rs3094315        1
#&gt; 2   1  785989  T  C   -1           4  rs2980300        2
#&gt; 3   1  798959  G  A    1           5 rs11240777        3
#&gt; 4   1  947034  G  A   -1           6  rs2465126        4
#&gt; 5   1  949608  G  A    1           7     rs1921        5
#&gt; 6   1 1018704  A  G   -1           8  rs9442372        6
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 301150 rows ]</code></pre>
<div class="infobox caution">
<p>Some datasets may have the convention that <code>a0 = allele1, a1 = allele2</code>, therefore all effects or allele frequencies are reversed (after matching).</p>
</div>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="population-structure.html#cb114-1" tabindex="-1"></a><span class="co"># further subsetting on missing values</span></span>
<span id="cb114-2"><a href="population-structure.html#cb114-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">bed_counts</span>(obj.bed, <span class="at">ind.col =</span> matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>, <span class="at">ncores =</span> NCORES)</span>
<span id="cb114-3"><a href="population-structure.html#cb114-3" tabindex="-1"></a><span class="fu">hist</span>(counts[<span class="dv">4</span>, ])</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-16-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="population-structure.html#cb115-1" tabindex="-1"></a>ind <span class="ot">&lt;-</span> <span class="fu">which</span>((counts[<span class="dv">4</span>, ] <span class="sc">/</span> <span class="fu">nrow</span>(obj.bed)) <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb115-2"><a href="population-structure.html#cb115-2" tabindex="-1"></a>matched2 <span class="ot">&lt;-</span> matched[ind, ]</span></code></pre></div>
</details>
<hr />
<div class="sourceCode" id="cb116"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="population-structure.html#cb116-1" tabindex="-1"></a><span class="co"># projection matrix</span></span>
<span id="cb116-2"><a href="population-structure.html#cb116-2" tabindex="-1"></a>proj_mat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(projection[matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>, <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)])</span>
<span id="cb116-3"><a href="population-structure.html#cb116-3" tabindex="-1"></a><span class="co"># reference allele frequencies</span></span>
<span id="cb116-4"><a href="population-structure.html#cb116-4" tabindex="-1"></a>ref_mat  <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(all_freq  [matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>, <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)])</span>
<span id="cb116-5"><a href="population-structure.html#cb116-5" tabindex="-1"></a><span class="co"># centers of reference populations in the PCA</span></span>
<span id="cb116-6"><a href="population-structure.html#cb116-6" tabindex="-1"></a>all_centers <span class="ot">&lt;-</span> <span class="fu">crossprod</span>(ref_mat, proj_mat)</span></code></pre></div>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="population-structure.html#cb117-1" tabindex="-1"></a><span class="co"># project individuals (divided by 2 -&gt; as AF) onto the PC space</span></span>
<span id="cb117-2"><a href="population-structure.html#cb117-2" tabindex="-1"></a>all_proj <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">sweep</span>(proj_mat, <span class="dv">2</span>, correction <span class="sc">/</span> <span class="dv">2</span>, <span class="st">&#39;*&#39;</span>), <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb117-3"><a href="population-structure.html#cb117-3" tabindex="-1"></a>  <span class="fu">bed_prodVec</span>(obj.bed, x, <span class="at">ind.col =</span> matched2<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>, <span class="at">ncores =</span> NCORES,</span>
<span id="cb117-4"><a href="population-structure.html#cb117-4" tabindex="-1"></a>              <span class="co"># scaling to get G if beta = 1 and (2 - G) if beta = -1</span></span>
<span id="cb117-5"><a href="population-structure.html#cb117-5" tabindex="-1"></a>              <span class="at">center =</span> <span class="dv">1</span> <span class="sc">-</span> matched2<span class="sc">$</span>beta, <span class="at">scale =</span> matched2<span class="sc">$</span>beta)</span>
<span id="cb117-6"><a href="population-structure.html#cb117-6" tabindex="-1"></a>})</span></code></pre></div>
<div class="infobox question">
<p>When is it needed to use the <code>correction</code>? Why can’t we use these provided PC loadings to project the UK Biobank individuals?</p>
</div>
<hr />
<p>We can then assign each individual to their closest center:</p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="population-structure.html#cb118-1" tabindex="-1"></a><span class="co"># distance of individuals from each reference population (the center)</span></span>
<span id="cb118-2"><a href="population-structure.html#cb118-2" tabindex="-1"></a>all_sq_dist <span class="ot">&lt;-</span> <span class="fu">apply</span>(all_centers, <span class="dv">1</span>, <span class="cf">function</span>(one_center) {</span>
<span id="cb118-3"><a href="population-structure.html#cb118-3" tabindex="-1"></a>  <span class="fu">rowSums</span>(<span class="fu">sweep</span>(all_proj, <span class="dv">2</span>, one_center, <span class="st">&#39;-&#39;</span>)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb118-4"><a href="population-structure.html#cb118-4" tabindex="-1"></a>})</span>
<span id="cb118-5"><a href="population-structure.html#cb118-5" tabindex="-1"></a></span>
<span id="cb118-6"><a href="population-structure.html#cb118-6" tabindex="-1"></a>THR <span class="ot">&lt;-</span> <span class="fl">0.002</span>  <span class="co"># you can adjust this threshold</span></span>
<span id="cb118-7"><a href="population-structure.html#cb118-7" tabindex="-1"></a>thr_sq_dist <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">dist</span>(all_centers)<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> THR <span class="sc">/</span> <span class="fl">0.16</span></span>
<span id="cb118-8"><a href="population-structure.html#cb118-8" tabindex="-1"></a></span>
<span id="cb118-9"><a href="population-structure.html#cb118-9" tabindex="-1"></a><span class="co"># gather some groups</span></span>
<span id="cb118-10"><a href="population-structure.html#cb118-10" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">colnames</span>(all_freq)[<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]</span>
<span id="cb118-11"><a href="population-structure.html#cb118-11" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Scandinavia&quot;</span>, <span class="st">&quot;United Kingdom&quot;</span>, <span class="st">&quot;Ireland&quot;</span>)]   <span class="ot">&lt;-</span> <span class="st">&quot;Europe (North West)&quot;</span></span>
<span id="cb118-12"><a href="population-structure.html#cb118-12" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Europe (South East)&quot;</span>, <span class="st">&quot;Europe (North East)&quot;</span>)] <span class="ot">&lt;-</span> <span class="st">&quot;Europe (East)&quot;</span></span>
<span id="cb118-13"><a href="population-structure.html#cb118-13" tabindex="-1"></a></span>
<span id="cb118-14"><a href="population-structure.html#cb118-14" tabindex="-1"></a><span class="co"># assign to closest cluster, when close enough</span></span>
<span id="cb118-15"><a href="population-structure.html#cb118-15" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">apply</span>(all_sq_dist, <span class="dv">1</span>, <span class="cf">function</span>(sq_dist) {</span>
<span id="cb118-16"><a href="population-structure.html#cb118-16" tabindex="-1"></a>  ind <span class="ot">&lt;-</span> <span class="fu">which.min</span>(sq_dist)</span>
<span id="cb118-17"><a href="population-structure.html#cb118-17" tabindex="-1"></a>  <span class="cf">if</span> (sq_dist[ind] <span class="sc">&lt;</span> thr_sq_dist) group[ind] <span class="cf">else</span> <span class="cn">NA</span></span>
<span id="cb118-18"><a href="population-structure.html#cb118-18" tabindex="-1"></a>})</span>
<span id="cb118-19"><a href="population-structure.html#cb118-19" tabindex="-1"></a></span>
<span id="cb118-20"><a href="population-structure.html#cb118-20" tabindex="-1"></a><span class="fu">table</span>(cluster, <span class="at">exclude =</span> <span class="cn">NULL</span>)  <span class="co"># 3 NAs -&gt; almost all assigned</span></span></code></pre></div>
<pre><code>#&gt; cluster
#&gt;           Ashkenazi       Europe (East) Europe (North West) 
#&gt;                 110                 148                 872 
#&gt; Europe (South West)               Italy         Middle East 
#&gt;                  45                 219                   4 
#&gt;                &lt;NA&gt; 
#&gt;                   3</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="population-structure.html#cb120-1" tabindex="-1"></a><span class="fu">plot_grid2</span>(<span class="at">plotlist =</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(k) {</span>
<span id="cb120-2"><a href="population-structure.html#cb120-2" tabindex="-1"></a>  k1 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb120-3"><a href="population-structure.html#cb120-3" tabindex="-1"></a>  k2 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k</span>
<span id="cb120-4"><a href="population-structure.html#cb120-4" tabindex="-1"></a>  <span class="fu">qplot</span>(PC[, k1], PC[, k2], <span class="at">color =</span> cluster, <span class="at">size =</span> <span class="fu">I</span>(<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb120-5"><a href="population-structure.html#cb120-5" tabindex="-1"></a>    <span class="fu">theme_bigstatsr</span>(<span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb120-6"><a href="population-structure.html#cb120-6" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k1), <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k2), <span class="at">color =</span> <span class="st">&quot;Assigned group&quot;</span>) <span class="sc">+</span></span>
<span id="cb120-7"><a href="population-structure.html#cb120-7" tabindex="-1"></a>    <span class="fu">coord_equal</span>()</span>
<span id="cb120-8"><a href="population-structure.html#cb120-8" tabindex="-1"></a>}), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">legend_ratio =</span> <span class="fl">0.25</span>, <span class="at">title_ratio =</span> <span class="dv">0</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-20-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>These are mostly European individuals.
PC4 is definitively a bit odd.</p>
<div class="infobox exo">
<p>Try to find out what’s going on with PC4.</p>
</div>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-abdellaoui2013population" class="csl-entry">
Abdellaoui, A., Hottenga, J.-J., Knijff, P. de, Nivard, M.G., Xiao, X., Scheet, P., et al. (2013). <a href="https://doi.org/10.1038/ejhg.2013.48">Population structure, migration, and diversifying selection in the <span>Netherlands</span></a>. <em>European Journal of Human Genetics</em>, <em>21</em>, 1277–1285.
</div>
<div id="ref-bycroft2018uk" class="csl-entry">
Bycroft, C., Freeman, C., Petkova, D., Band, G., Elliott, L.T., Sharp, K., et al. (2018). <a href="https://doi.org/10.1038/s41586-018-0579-z">The <span>UK Biobank</span> resource with deep phenotyping and genomic data</a>. <em>Nature</em>, <em>562</em>, 203–209.
</div>
<div id="ref-grinde2024adjusting" class="csl-entry">
Grinde, K.E., Browning, B.L., Reiner, A.P., Thornton, T.A., &amp; Browning, S.R. (2024). <a href="https://doi.org/10.1371/journal.pgen.1011242">Adjusting for principal components can induce collider bias in genome-wide association studies</a>. <em>PLoS Genetics</em>, <em>20</em>, e1011242.
</div>
<div id="ref-price2008long" class="csl-entry">
Price, A.L., Weale, M.E., Patterson, N., Myers, S.R., Need, A.C., Shianna, K.V., et al. (2008). <a href="https://doi.org/10.1016/j.ajhg.2008.06.005">Long-range <span>LD</span> can confound genome scans in admixed populations</a>. <em>The American Journal of Human Genetics</em>, <em>83</em>, 132–135.
</div>
<div id="ref-prive2021using" class="csl-entry">
Privé, F. (2022). <a href="https://doi.org/10.1093/bioinformatics/btac348"><span class="nocase">Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics</span></a>. <em>Bioinformatics</em>, <em>38</em>, 3477–3480.
</div>
<div id="ref-prive2021identifying" class="csl-entry">
Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136">Identifying and correcting for misspecifications in <span>GWAS</span> summary statistics and polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>, <em>3</em>, 100136.
</div>
<div id="ref-prive2021high" class="csl-entry">
Privé, F., Aschard, H., Carmi, S., Folkersen, L., Hoggart, C., O’Reilly, P.F., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.ajhg.2021.11.008">Portability of 245 polygenic scores when derived from the <span>UK Biobank</span> and applied to 9 ancestry groups from the same cohort</a>. <em>The American Journal of Human Genetics</em>, <em>109</em>, 12–23.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
Privé, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781–2787.
</div>
<div id="ref-prive2020efficient" class="csl-entry">
Privé, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449–4457.
</div>
<div id="ref-prive2020performing" class="csl-entry">
Privé, F., Luu, K., Vilhjálmsson, B.J., &amp; Blum, M.G. (2020). <a href="https://doi.org/10.1093/molbev/msaa053">Performing highly efficient genome scans for local adaptation with <span>R</span> package pcadapt version 4</a>. <em>Molecular Biology and Evolution</em>, <em>37</em>, 2153–2154.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="preprocessing.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genome-wide-association-study-gwas.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
