<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Population structure | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 3 Population structure | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Population structure | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Population structure | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian PrivÃ© &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="preprocessing.html"/>
<link rel="next" href="genome-wide-association-study-gwas.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i>Prerequisites</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i>License</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i>Author</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i>Contact</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>1</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="1.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>1.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="1.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>1.2</b> The bigSNP format from {bigsnpr}</a></li>
<li class="chapter" data-level="1.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>1.3</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="1.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>1.4</b> The FBM format from {bigstatsr}</a></li>
<li class="chapter" data-level="1.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>1.5</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="1.5.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>1.5.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="1.5.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>1.5.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="1.5.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>1.5.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="1.5.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>1.5.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="1.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>1.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>2</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="2.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>2.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="2.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>2.2</b> Imputation</a></li>
<li class="chapter" data-level="2.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>2.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>3</b> Population structure</a>
<ul>
<li class="chapter" data-level="3.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>3.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="3.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>3.2</b> The problem with LD</a></li>
<li class="chapter" data-level="3.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>3.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="3.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>3.4</b> Exercise</a></li>
<li class="chapter" data-level="3.5" data-path="population-structure.html"><a href="population-structure.html#ancestry-inference"><i class="fa fa-check"></i><b>3.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>4</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="4.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>4.1</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>5</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="5.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>5.1</b> Objective</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>5.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="5.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>5.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>5.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="5.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>5.2.1</b> Load data</a></li>
<li class="chapter" data-level="5.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>5.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="5.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>5.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="5.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>5.2.4</b> Letâs take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="5.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>5.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>5.3</b> Real data</a>
<ul>
<li class="chapter" data-level="5.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>5.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="5.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>5.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="5.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#create-ld-matrix"><i class="fa fa-check"></i><b>5.3.3</b> Create LD matrix</a></li>
<li class="chapter" data-level="5.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>5.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="5.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>5.3.5</b> Examine results</a></li>
<li class="chapter" data-level="5.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>5.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="5.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>5.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="population-structure" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Population structure<a href="population-structure.html#population-structure" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="principal-component-analysis-pca" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Principal Component Analysis (PCA)<a href="population-structure.html#principal-component-analysis-pca" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>PCA on the genotype matrix can be used to capture population structure.
However, PCA can actually capture different kinds of structure <span class="citation">(<a href="#ref-prive2020efficient">PrivÃ©, Luu, Blum, et al., 2020</a>)</span>:</p>
<ul>
<li><p>population structure (what we want),</p></li>
<li><p>Linkage Disequilibrium (LD) structure, when there are two many correlated variants (e.g.Â within long-range LD regions) and not enough population structure (see <a href="https://privefl.github.io/bigsnpr/articles/how-to-PCA.html">this vignette</a>),</p></li>
<li><p>relatedness structure, when there are related individuals who usually cluster together in later PCs,</p></li>
<li><p>noise, basically just circles when looking at PC scores.</p></li>
</ul>
<p>Population structure is the second main topic of my research work (after polygenic scores).</p>
<ul>
<li><p>In <span class="citation">PrivÃ© et al. (<a href="#ref-prive2017efficient">2018</a>)</span>, I introduced an algorithm to compute PCA for a <code>bigSNP</code> object while accounting for the LD problem by using clumping (not pruning, cf.Â <a href="https://privefl.github.io/bigsnpr/articles/pruning-vs-clumping.html">this vignette</a>) and an <em>automatic</em> detection and removal of long-range LD regions.</p></li>
<li><p>In <span class="citation">PrivÃ©, Luu, VilhjÃ¡lmsson, &amp; Blum (<a href="#ref-prive2020performing">2020</a>)</span>, I improved the implementation of the pcadapt algorithm, which detects variants associated with population structure (i.e.Â some kind of GWAS for population structure).</p></li>
<li><p>In <span class="citation">PrivÃ©, Luu, Blum, et al. (<a href="#ref-prive2020efficient">2020</a>)</span>, I extended {bigsnpr} to also be able to run PCA on PLINK bed files directly with a small percentage of missing values, and investigated best practices for PCA in more detail.</p></li>
<li><p>In <span class="citation">PrivÃ© et al. (<a href="#ref-prive2021high">2022</a>)</span> and <span class="citation">PrivÃ© (<a href="#ref-prive2021using">2022</a>)</span>, I showed how to use PCA for ancestry inference, including grouping individuals in homogeneous ancestry groups, and inferring ancestry proportions from genotype data but also from allele frequencies only (see <a href="https://privefl.github.io/bigsnpr/articles/ancestry.html">this vignette</a>).</p></li>
</ul>
</div>
<div id="the-problem-with-ld" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> The problem with LD<a href="population-structure.html#the-problem-with-ld" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Let us reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">2.3</a>.</p>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="population-structure.html#cb87-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb87-2"><a href="population-structure.html#cb87-2" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span></code></pre></div>
<pre><code>#&gt; Loading required package: bigstatsr</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="population-structure.html#cb89-1" tabindex="-1"></a>bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.rds&quot;</span>)</span>
<span id="cb89-2"><a href="population-structure.html#cb89-2" tabindex="-1"></a>G <span class="ot">&lt;-</span> bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb89-3"><a href="population-structure.html#cb89-3" tabindex="-1"></a>NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>()</span></code></pre></div>
<div class="infobox exo">
<p>Use <code>big_randomSVD()</code> to perform a SVD/PCA on <code>G</code>. Do not forget to use some scaling (at least some centering). Look at PC scores and PC loadings with <code>plot()</code>. What do you see for PC4?</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="population-structure.html#cb90-1" tabindex="-1"></a>svd <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb90-2"><a href="population-structure.html#cb90-2" tabindex="-1"></a>  <span class="fu">big_randomSVD</span>(G, <span class="at">fun.scaling =</span> <span class="fu">big_scale</span>(), <span class="at">ncores =</span> NCORES),</span>
<span id="cb90-3"><a href="population-structure.html#cb90-3" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/svd_with_ld.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;    0.73    0.27   59.08 
#&gt; Code finished running at 2025-06-10 13:33:51 CEST</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="population-structure.html#cb92-1" tabindex="-1"></a><span class="fu">plot</span>(svd)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="population-structure.html#cb93-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-2.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="population-structure.html#cb94-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-3.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb95-1"><a href="population-structure.html#cb95-1" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;loadings&quot;</span>, <span class="at">loadings =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">coef =</span> <span class="fl">0.4</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-2-4.png" width="70%" style="display: block; margin: auto;" /></p>
<p>What is going on with PC4?</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="population-structure.html#cb96-1" tabindex="-1"></a>the_max <span class="ot">&lt;-</span> <span class="fu">which.max</span>(<span class="fu">abs</span>(svd<span class="sc">$</span>v[, <span class="dv">4</span>]))</span>
<span id="cb96-2"><a href="population-structure.html#cb96-2" tabindex="-1"></a><span class="fu">plot</span>(svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">3</span><span class="sc">:</span><span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb96-3"><a href="population-structure.html#cb96-3" tabindex="-1"></a>  <span class="fu">aes</span>(<span class="at">color =</span> <span class="fu">as.factor</span>(G[, the_max])) <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">&quot;One geno&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-3-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>PC4 is basically capturing variation in a block of LD and is then very correlated to genetic variants in this region. You really want to avoid using this type of PCs e.g.Â as covariates in GWAS because it will cause some collider bias.</p>
</details>
<hr />
<p>To avoid capturing LD in PCA, it has been recommended to perform some pruning (removing variants too correlated with one another) AND remove some known list of long-range LD regions (that can be captured by PCA, even after a stringent pruning step). But this is not enoughâ¦ This is exactly what the UK Biobank did, and this how the PC loadings of the 40 PCs they provide look like:</p>
<p><img src="https://github.com/privefl/thesis-docs/blob/master/figures/loadings-ukbb.jpeg?raw=true" width="50%" style="display: block; margin: auto;" /></p>
<p><strong>I recommend using only the first 16 PCs provided by the UK Biobank <span class="citation">(<a href="#ref-prive2020efficient">PrivÃ©, Luu, Blum, et al., 2020</a>)</span>.</strong></p>
</div>
<div id="best-practices-for-pca-of-genetic-data" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Best practices for PCA of genetic data<a href="population-structure.html#best-practices-for-pca-of-genetic-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>There can be many steps to properly perform a PCA; you can find more about this in <span class="citation">PrivÃ©, Luu, Blum, et al. (<a href="#ref-prive2020efficient">2020</a>)</span>.
Letâs have a look at <a href="https://privefl.github.io/bigsnpr/articles/bedpca.html">the corresponding tutorial from the bigsnpr website</a>.</p>
</div>
<div id="exo-pca" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Exercise<a href="population-structure.html#exo-pca" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="infobox exo">
<p>Let us reuse the data prepared in <a href="preprocessing.html#exo-preprocessing">2.3</a>.</p>
<p>Follow the previous tutorial to perform a PCA for this data, either using <code>snp_*</code> functions on the bigSNP object or <code>bed_*</code> functions on the bed file mapped.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<p>First, let us get an idea of the relatedness in the data using</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="population-structure.html#cb97-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb97-2"><a href="population-structure.html#cb97-2" tabindex="-1"></a>(NCORES <span class="ot">&lt;-</span> <span class="fu">nb_cores</span>())</span>
<span id="cb97-3"><a href="population-structure.html#cb97-3" tabindex="-1"></a>plink2 <span class="ot">&lt;-</span> <span class="fu">download_plink2</span>(<span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb97-4"><a href="population-structure.html#cb97-4" tabindex="-1"></a>rel <span class="ot">&lt;-</span> <span class="fu">snp_plinkKINGQC</span>(plink2, <span class="st">&quot;tmp-data/GWAS_data_sorted_QC.bed&quot;</span>,</span>
<span id="cb97-5"><a href="population-structure.html#cb97-5" tabindex="-1"></a>                       <span class="at">thr.king =</span> <span class="dv">2</span><span class="sc">^-</span><span class="fl">4.5</span>, <span class="at">make.bed =</span> <span class="cn">FALSE</span>, <span class="at">ncores =</span> NCORES)</span></code></pre></div>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="population-structure.html#cb98-1" tabindex="-1"></a><span class="fu">hist</span>(<span class="fu">log2</span>(rel<span class="sc">$</span>KINSHIP), <span class="st">&quot;FD&quot;</span>); <span class="fu">abline</span>(<span class="at">v =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">4.5</span>, <span class="sc">-</span><span class="fl">3.5</span>), <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-6-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="infobox info">
<p>When computing relatedness with KING, LD pruning is <em>NOT</em> recommended. However, it may be useful to filter out some variants that are highly associated with population structure, e.g.Â as performed in the UK Biobank <span class="citation">(<a href="#ref-bycroft2018uk">Bycroft et al., 2018</a>)</span>. For example, see <a href="https://github.com/privefl/bigsnpr-extdoc/blob/main/example-code/2-rel-and-pca.R">this code</a>.</p>
</div>
<p>Relatedness should not be a major issue here. Let us now compute PCs.</p>
<div class="infobox info">
<p>All the code that follows could be run on the bigSNP object we made before. Nevertheless, to showcase the <code>bed_*</code> functions here, we will run the following analyses on the bed file directly.</p>
</div>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="population-structure.html#cb99-1" tabindex="-1"></a><span class="co"># Memory-map a bed file directly</span></span>
<span id="cb99-2"><a href="population-structure.html#cb99-2" tabindex="-1"></a>obj.bed <span class="ot">&lt;-</span> <span class="fu">bed</span>(<span class="st">&quot;tmp-data/GWAS_data_sorted_QC.bed&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="population-structure.html#cb100-1" tabindex="-1"></a>obj.svd <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">save_run</span>(</span>
<span id="cb100-2"><a href="population-structure.html#cb100-2" tabindex="-1"></a>  <span class="fu">bed_autoSVD</span>(obj.bed, <span class="at">k =</span> <span class="dv">12</span>, <span class="at">ncores =</span> NCORES),</span>
<span id="cb100-3"><a href="population-structure.html#cb100-3" tabindex="-1"></a>  <span class="at">file =</span> <span class="st">&quot;tmp-data/PCA_GWAS_data.rds&quot;</span>)</span></code></pre></div>
<pre><code>#&gt;    user  system elapsed 
#&gt;   80.42    1.28  392.85 
#&gt; Code finished running at 2025-06-10 13:41:05 CEST</code></pre>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="population-structure.html#cb102-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="population-structure.html#cb103-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd, <span class="at">type =</span> <span class="st">&quot;scores&quot;</span>, <span class="at">scores =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">12</span>, <span class="at">coeff =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-9-2.png" width="70%" style="display: block; margin: auto;" /></p>
<p>There is some population structure (maybe up to 6 PCs). You should also check loadings to make sure there is no LD structure (peaks on loadings):</p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="population-structure.html#cb104-1" tabindex="-1"></a><span class="fu">plot</span>(obj.svd, <span class="at">type =</span> <span class="st">&quot;loadings&quot;</span>, <span class="at">loadings =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">coeff =</span> <span class="fl">0.5</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-10-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>No peaks, but loadings of PC4 and PC5 are a bit odd.</p>
<p>If you expect the individuals to mostly come from one population (e.g.Â in a national biobank), you can simply use a robust distance to identify a homogeneous subset of individuals, then look at the histogram of log-distances to choose a threshold based on visual inspection (here I would probably choose 4.5).</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="population-structure.html#cb105-1" tabindex="-1"></a>PC <span class="ot">&lt;-</span> <span class="fu">predict</span>(obj.svd)</span>
<span id="cb105-2"><a href="population-structure.html#cb105-2" tabindex="-1"></a>ldist <span class="ot">&lt;-</span> <span class="fu">log</span>(bigutilsr<span class="sc">::</span><span class="fu">dist_ogk</span>(PC[, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]))</span>
<span id="cb105-3"><a href="population-structure.html#cb105-3" tabindex="-1"></a><span class="fu">hist</span>(ldist, <span class="st">&quot;FD&quot;</span>)</span></code></pre></div>
<p><img src="popstruct_files/figure-html/unnamed-chunk-11-1.png" width="70%" style="display: block; margin: auto;" /></p>
<div class="sourceCode" id="cb106"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="population-structure.html#cb106-1" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb106-2"><a href="population-structure.html#cb106-2" tabindex="-1"></a><span class="fu">source</span>(<span class="st">&quot;https://raw.githubusercontent.com/privefl/paper4-bedpca/master/code/plot_grid2.R&quot;</span>)</span>
<span id="cb106-3"><a href="population-structure.html#cb106-3" tabindex="-1"></a></span>
<span id="cb106-4"><a href="population-structure.html#cb106-4" tabindex="-1"></a><span class="fu">plot_grid2</span>(<span class="at">plotlist =</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(k) {</span>
<span id="cb106-5"><a href="population-structure.html#cb106-5" tabindex="-1"></a>  k1 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb106-6"><a href="population-structure.html#cb106-6" tabindex="-1"></a>  k2 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k</span>
<span id="cb106-7"><a href="population-structure.html#cb106-7" tabindex="-1"></a>  <span class="fu">qplot</span>(PC[, k1], PC[, k2], <span class="at">color =</span> ldist, <span class="at">size =</span> <span class="fu">I</span>(<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb106-8"><a href="population-structure.html#cb106-8" tabindex="-1"></a>    <span class="fu">scale_color_viridis_c</span>() <span class="sc">+</span></span>
<span id="cb106-9"><a href="population-structure.html#cb106-9" tabindex="-1"></a>    <span class="fu">theme_bigstatsr</span>(<span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb106-10"><a href="population-structure.html#cb106-10" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k1), <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k2), <span class="at">color =</span> <span class="st">&quot;log-distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb106-11"><a href="population-structure.html#cb106-11" tabindex="-1"></a>    <span class="fu">coord_equal</span>()</span>
<span id="cb106-12"><a href="population-structure.html#cb106-12" tabindex="-1"></a>}), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">legend_ratio =</span> <span class="fl">0.2</span>, <span class="at">title_ratio =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>#&gt; Warning: `qplot()` was deprecated in ggplot2 3.4.0.
#&gt; This warning is displayed once every 8 hours.
#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#&gt; generated.</code></pre>
<pre><code>#&gt; Warning in get_plot_component(plot, &quot;guide-box&quot;): Multiple components
#&gt; found; returning the first one. To return all, use `return_all = TRUE`.</code></pre>
<p><img src="popstruct_files/figure-html/unnamed-chunk-12-1.png" width="70%" style="display: block; margin: auto;" /></p>
</details>
</div>
<div id="ancestry-inference" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Ancestry inference<a href="population-structure.html#ancestry-inference" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>It would be nice if we could get a better sense of the ancestry of these individuals.
To achieve this, we can project this data onto the PCA space of many known population groups defined in <span class="citation">PrivÃ© (<a href="#ref-prive2021using">2022</a>)</span>.</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="population-structure.html#cb109-1" tabindex="-1"></a>all_freq <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb109-2"><a href="population-structure.html#cb109-2" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb109-3"><a href="population-structure.html#cb109-3" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019027&quot;</span>,  <span class="co"># for the tutorial (46 MB)</span></span>
<span id="cb109-4"><a href="population-structure.html#cb109-4" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620968&quot;,  # for real analyses (849 MB)</span></span>
<span id="cb109-5"><a href="population-structure.html#cb109-5" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;ref_freqs.csv.gz&quot;</span>))</span>
<span id="cb109-6"><a href="population-structure.html#cb109-6" tabindex="-1"></a></span>
<span id="cb109-7"><a href="population-structure.html#cb109-7" tabindex="-1"></a>projection <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(</span>
<span id="cb109-8"><a href="population-structure.html#cb109-8" tabindex="-1"></a>  runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb109-9"><a href="population-structure.html#cb109-9" tabindex="-1"></a>    <span class="st">&quot;https://figshare.com/ndownloader/files/38019024&quot;</span>,  <span class="co"># for the tutorial (44 MB)</span></span>
<span id="cb109-10"><a href="population-structure.html#cb109-10" tabindex="-1"></a>    <span class="co"># &quot;https://figshare.com/ndownloader/files/31620953&quot;,  # for real analyses (847 MB)</span></span>
<span id="cb109-11"><a href="population-structure.html#cb109-11" tabindex="-1"></a>    <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;projection.csv.gz&quot;</span>))</span>
<span id="cb109-12"><a href="population-structure.html#cb109-12" tabindex="-1"></a></span>
<span id="cb109-13"><a href="population-structure.html#cb109-13" tabindex="-1"></a><span class="co"># coefficients to correct for overfitting of PCA</span></span>
<span id="cb109-14"><a href="population-structure.html#cb109-14" tabindex="-1"></a>correction <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="fl">1.008</span>, <span class="fl">1.021</span>, <span class="fl">1.034</span>, <span class="fl">1.052</span>, <span class="fl">1.074</span>, <span class="fl">1.099</span>,</span>
<span id="cb109-15"><a href="population-structure.html#cb109-15" tabindex="-1"></a>                <span class="fl">1.123</span>, <span class="fl">1.15</span>, <span class="fl">1.195</span>, <span class="fl">1.256</span>, <span class="fl">1.321</span>, <span class="fl">1.382</span>, <span class="fl">1.443</span>)</span></code></pre></div>
<div class="sourceCode" id="cb110"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="population-structure.html#cb110-1" tabindex="-1"></a><span class="co"># match variants between the two datasets</span></span>
<span id="cb110-2"><a href="population-structure.html#cb110-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb110-3"><a href="population-structure.html#cb110-3" tabindex="-1"></a>matched <span class="ot">&lt;-</span> obj.bed<span class="sc">$</span>map <span class="sc">%&gt;%</span></span>
<span id="cb110-4"><a href="population-structure.html#cb110-4" tabindex="-1"></a>  <span class="fu">transmute</span>(<span class="at">chr =</span> chromosome, <span class="at">pos =</span> physical.pos, <span class="at">a1 =</span> allele1, <span class="at">a0 =</span> allele2) <span class="sc">%&gt;%</span></span>
<span id="cb110-5"><a href="population-structure.html#cb110-5" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">beta =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb110-6"><a href="population-structure.html#cb110-6" tabindex="-1"></a>  <span class="fu">snp_match</span>(all_freq[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]) <span class="sc">%&gt;%</span></span>
<span id="cb110-7"><a href="population-structure.html#cb110-7" tabindex="-1"></a>  <span class="fu">print</span>()</span></code></pre></div>
<pre><code>#&gt;   chr     pos a0 a1 beta _NUM_ID_.ss       rsid _NUM_ID_
#&gt; 1   1  752566  G  A   -1           2  rs3094315        1
#&gt; 2   1  785989  T  C   -1           4  rs2980300        2
#&gt; 3   1  798959  G  A    1           5 rs11240777        3
#&gt; 4   1  947034  G  A   -1           6  rs2465126        4
#&gt; 5   1  949608  G  A    1           7     rs1921        5
#&gt; 6   1 1018704  A  G   -1           8  rs9442372        6
#&gt;  [ reached &#39;max&#39; / getOption(&quot;max.print&quot;) -- omitted 301150 rows ]</code></pre>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="population-structure.html#cb112-1" tabindex="-1"></a><span class="co"># further subsetting on missing values</span></span>
<span id="cb112-2"><a href="population-structure.html#cb112-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">bed_counts</span>(obj.bed, <span class="at">ind.col =</span> matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>, <span class="at">ncores =</span> NCORES)</span>
<span id="cb112-3"><a href="population-structure.html#cb112-3" tabindex="-1"></a><span class="co"># hist(nb_na &lt;- counts[4, ])</span></span>
<span id="cb112-4"><a href="population-structure.html#cb112-4" tabindex="-1"></a>ind <span class="ot">&lt;-</span> <span class="fu">which</span>(counts[<span class="dv">4</span>, ] <span class="sc">&lt;</span> (<span class="fu">nrow</span>(obj.bed) <span class="sc">*</span> <span class="fl">0.05</span>))</span>
<span id="cb112-5"><a href="population-structure.html#cb112-5" tabindex="-1"></a></span>
<span id="cb112-6"><a href="population-structure.html#cb112-6" tabindex="-1"></a><span class="co"># project individuals (divided by 2) onto the PC space</span></span>
<span id="cb112-7"><a href="population-structure.html#cb112-7" tabindex="-1"></a>PROJ <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(projection[matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>[ind], <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)])</span>
<span id="cb112-8"><a href="population-structure.html#cb112-8" tabindex="-1"></a>all_proj <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">sweep</span>(PROJ, <span class="dv">2</span>, correction <span class="sc">/</span> <span class="dv">2</span>, <span class="st">&#39;*&#39;</span>), <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb112-9"><a href="population-structure.html#cb112-9" tabindex="-1"></a>  <span class="fu">bed_prodVec</span>(obj.bed, x, <span class="at">ind.col =</span> matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_.ss</span><span class="st">`</span>[ind], <span class="at">ncores =</span> NCORES,</span>
<span id="cb112-10"><a href="population-structure.html#cb112-10" tabindex="-1"></a>              <span class="co"># scaling to get G if beta = 1 and (2 - G) if beta = -1</span></span>
<span id="cb112-11"><a href="population-structure.html#cb112-11" tabindex="-1"></a>              <span class="at">center =</span> <span class="dv">1</span> <span class="sc">-</span> matched<span class="sc">$</span>beta[ind],</span>
<span id="cb112-12"><a href="population-structure.html#cb112-12" tabindex="-1"></a>              <span class="at">scale =</span> matched<span class="sc">$</span>beta[ind])</span>
<span id="cb112-13"><a href="population-structure.html#cb112-13" tabindex="-1"></a>})</span></code></pre></div>
<p>We can then assign individuals to the closest center:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="population-structure.html#cb113-1" tabindex="-1"></a>all_centers <span class="ot">&lt;-</span> <span class="fu">crossprod</span>(<span class="fu">as.matrix</span>(all_freq[matched<span class="sc">$</span><span class="st">`</span><span class="at">_NUM_ID_</span><span class="st">`</span>[ind], <span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]), PROJ)</span>
<span id="cb113-2"><a href="population-structure.html#cb113-2" tabindex="-1"></a>all_sq_dist <span class="ot">&lt;-</span> <span class="fu">apply</span>(all_centers, <span class="dv">1</span>, <span class="cf">function</span>(one_center) {</span>
<span id="cb113-3"><a href="population-structure.html#cb113-3" tabindex="-1"></a>  <span class="fu">rowSums</span>(<span class="fu">sweep</span>(all_proj, <span class="dv">2</span>, one_center, <span class="st">&#39;-&#39;</span>)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb113-4"><a href="population-structure.html#cb113-4" tabindex="-1"></a>})</span>
<span id="cb113-5"><a href="population-structure.html#cb113-5" tabindex="-1"></a></span>
<span id="cb113-6"><a href="population-structure.html#cb113-6" tabindex="-1"></a>THR <span class="ot">&lt;-</span> <span class="fl">0.002</span>  <span class="co"># you can adjust this threshold</span></span>
<span id="cb113-7"><a href="population-structure.html#cb113-7" tabindex="-1"></a>thr_sq_dist <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">dist</span>(all_centers)<span class="sc">^</span><span class="dv">2</span>) <span class="sc">*</span> THR <span class="sc">/</span> <span class="fl">0.16</span></span>
<span id="cb113-8"><a href="population-structure.html#cb113-8" tabindex="-1"></a></span>
<span id="cb113-9"><a href="population-structure.html#cb113-9" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">colnames</span>(all_freq)[<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>)]</span>
<span id="cb113-10"><a href="population-structure.html#cb113-10" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Scandinavia&quot;</span>, <span class="st">&quot;United Kingdom&quot;</span>, <span class="st">&quot;Ireland&quot;</span>)]   <span class="ot">&lt;-</span> <span class="st">&quot;Europe (North West)&quot;</span></span>
<span id="cb113-11"><a href="population-structure.html#cb113-11" tabindex="-1"></a>group[group <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">&quot;Europe (South East)&quot;</span>, <span class="st">&quot;Europe (North East)&quot;</span>)] <span class="ot">&lt;-</span> <span class="st">&quot;Europe (East)&quot;</span></span>
<span id="cb113-12"><a href="population-structure.html#cb113-12" tabindex="-1"></a></span>
<span id="cb113-13"><a href="population-structure.html#cb113-13" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">apply</span>(all_sq_dist, <span class="dv">1</span>, <span class="cf">function</span>(sq_dist) {</span>
<span id="cb113-14"><a href="population-structure.html#cb113-14" tabindex="-1"></a>  ind <span class="ot">&lt;-</span> <span class="fu">which.min</span>(sq_dist)</span>
<span id="cb113-15"><a href="population-structure.html#cb113-15" tabindex="-1"></a>  <span class="cf">if</span> (sq_dist[ind] <span class="sc">&lt;</span> thr_sq_dist) group[ind] <span class="cf">else</span> <span class="cn">NA</span></span>
<span id="cb113-16"><a href="population-structure.html#cb113-16" tabindex="-1"></a>})</span>
<span id="cb113-17"><a href="population-structure.html#cb113-17" tabindex="-1"></a></span>
<span id="cb113-18"><a href="population-structure.html#cb113-18" tabindex="-1"></a><span class="fu">table</span>(cluster, <span class="at">exclude =</span> <span class="cn">NULL</span>)  <span class="co"># 3 NAs -&gt; almost all assigned</span></span></code></pre></div>
<pre><code>#&gt; cluster
#&gt;           Ashkenazi       Europe (East) Europe (North West) 
#&gt;                 110                 148                 872 
#&gt; Europe (South West)               Italy         Middle East 
#&gt;                  45                 219                   4 
#&gt;                &lt;NA&gt; 
#&gt;                   3</code></pre>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="population-structure.html#cb115-1" tabindex="-1"></a><span class="fu">plot_grid2</span>(<span class="at">plotlist =</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="cf">function</span>(k) {</span>
<span id="cb115-2"><a href="population-structure.html#cb115-2" tabindex="-1"></a>  k1 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k <span class="sc">-</span> <span class="dv">1</span></span>
<span id="cb115-3"><a href="population-structure.html#cb115-3" tabindex="-1"></a>  k2 <span class="ot">&lt;-</span> <span class="dv">2</span> <span class="sc">*</span> k</span>
<span id="cb115-4"><a href="population-structure.html#cb115-4" tabindex="-1"></a>  <span class="fu">qplot</span>(PC[, k1], PC[, k2], <span class="at">color =</span> cluster, <span class="at">size =</span> <span class="fu">I</span>(<span class="dv">2</span>)) <span class="sc">+</span></span>
<span id="cb115-5"><a href="population-structure.html#cb115-5" tabindex="-1"></a>    <span class="fu">theme_bigstatsr</span>(<span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb115-6"><a href="population-structure.html#cb115-6" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k1), <span class="at">y =</span> <span class="fu">paste0</span>(<span class="st">&quot;PC&quot;</span>, k2), <span class="at">color =</span> <span class="st">&quot;Assigned</span><span class="sc">\n</span><span class="st">population&quot;</span>) <span class="sc">+</span></span>
<span id="cb115-7"><a href="population-structure.html#cb115-7" tabindex="-1"></a>    <span class="fu">coord_equal</span>()</span>
<span id="cb115-8"><a href="population-structure.html#cb115-8" tabindex="-1"></a>}), <span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">legend_ratio =</span> <span class="fl">0.25</span>, <span class="at">title_ratio =</span> <span class="dv">0</span>)</span></code></pre></div>
<pre><code>#&gt; Warning in get_plot_component(plot, &quot;guide-box&quot;): Multiple components
#&gt; found; returning the first one. To return all, use `return_all = TRUE`.</code></pre>
<p><img src="popstruct_files/figure-html/unnamed-chunk-16-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>These are mostly European individuals.
PC4-PC6 are definitively a bit odd.</p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bycroft2018uk" class="csl-entry">
Bycroft, C., Freeman, C., Petkova, D., Band, G., Elliott, L.T., Sharp, K., et al. (2018). <a href="https://doi.org/10.1038/s41586-018-0579-z">The <span>UK Biobank</span> resource with deep phenotyping and genomic data</a>. <em>Nature</em>, <em>562</em>, 203â209.
</div>
<div id="ref-prive2021using" class="csl-entry">
PrivÃ©, F. (2022). <a href="https://doi.org/10.1093/bioinformatics/btac348"><span class="nocase">Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics</span></a>. <em>Bioinformatics</em>, <em>38</em>, 3477â3480.
</div>
<div id="ref-prive2021high" class="csl-entry">
PrivÃ©, F., Aschard, H., Carmi, S., Folkersen, L., Hoggart, C., OâReilly, P.F., &amp; VilhjÃ¡lmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.ajhg.2021.11.008">Portability of 245 polygenic scores when derived from the <span>UK Biobank</span> and applied to 9 ancestry groups from the same cohort</a>. <em>The American Journal of Human Genetics</em>, <em>109</em>, 12â23.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
PrivÃ©, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781â2787.
</div>
<div id="ref-prive2020efficient" class="csl-entry">
PrivÃ©, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; VilhjÃ¡lmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449â4457.
</div>
<div id="ref-prive2020performing" class="csl-entry">
PrivÃ©, F., Luu, K., VilhjÃ¡lmsson, B.J., &amp; Blum, M.G. (2020). <a href="https://doi.org/10.1093/molbev/msaa053">Performing highly efficient genome scans for local adaptation with <span>R</span> package pcadapt version 4</a>. <em>Molecular Biology and Evolution</em>, <em>37</em>, 2153â2154.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="preprocessing.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genome-wide-association-study-gwas.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
