<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Preprocessing | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 3 Preprocessing | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Preprocessing | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Preprocessing | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-18" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="inputs-and-formats.html"/>
<link rel="next" href="population-structure.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#author"><i class="fa fa-check"></i><b>1.3</b> Author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.3</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.4</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.5</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.5.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.5.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.5.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.5.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.5.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.5.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.5.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.1</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.2</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.2.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.2.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.2.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.3</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Future of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.5</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.5.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.5.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.5.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.5.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.5.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="preprocessing" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> Preprocessing<a href="preprocessing.html#preprocessing" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>In this chapter, I talk about conversion, quality control and imputation. Conversion is also discussed in <a href="inputs-and-formats.html#getting-bigsnp">2.3</a>.</p>
<div id="PLINK" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Conversion and quality control of PLINK files<a href="preprocessing.html#PLINK" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>PLINK is very efficient for conversion and quality control of multiple formats, so I provide some wrappers to PLINK in bigsnpr, for ease of use directly from R:</p>
<ul>
<li><p><code>download_plink()</code> and <code>download_plink2()</code> for downloading the latest stable versions of PLINK 1.9 and 2.0 <span class="citation">(<a href="#ref-chang2015second">Chang et al., 2015</a>)</span>.</p></li>
<li><p><code>snp_plinkQC()</code> for quality control (QC) and conversion to bed/bim/fam.</p></li>
<li><p><code>snp_plinkKINGQC()</code> for QC on relatedness based on KING-robust kinship estimator <span class="citation">(<a href="#ref-manichaikul2010robust">Manichaikul et al., 2010</a>)</span>. Using <code>make.bed = FALSE</code> allows for computing related pairs only, i.e. reporting a data frame without producing new bed/bim/fam files. Note that monozygotic twins or identical samples have a KING coefficient of <span class="math inline">\(0.5\)</span>, not <span class="math inline">\(1\)</span>; <span class="math inline">\(0.25\)</span> = siblings and parents; <span class="math inline">\(2^{-3}\)</span> = second-degree relatives (e.g. grandparents, uncles); <span class="math inline">\(2^{-4}\)</span> = third-degree relatives (e.g. cousins). You can use a threshold of <span class="math inline">\(2^{-4.5} \approx 0.0442\)</span> to remove related individuals (PLINK removes one member of each pair of samples).</p></li>
<li><p><code>snp_plinkIBDQC()</code> for QC based on identity-by-descent (IBD) computed by PLINK using its method-of-moments. The KING one is supposed to be more robust.</p></li>
<li><p><code>snp_plinkRmSamples()</code> for producing new PLINK files after having removed some individuals.</p></li>
<li><p>For any other PLINK function, I prefer calling PLINK directly from R thanks to system calls and R package glue, e.g.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="preprocessing.html#cb61-1" tabindex="-1"></a>plink <span class="ot">&lt;-</span> <span class="fu">download_plink</span>(<span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb61-2"><a href="preprocessing.html#cb61-2" tabindex="-1"></a><span class="fu">system</span>(glue<span class="sc">::</span><span class="fu">glue</span>(</span>
<span id="cb61-3"><a href="preprocessing.html#cb61-3" tabindex="-1"></a>  <span class="st">&quot;{plink} --version&quot;</span></span>
<span id="cb61-4"><a href="preprocessing.html#cb61-4" tabindex="-1"></a>))</span></code></pre></div>
<pre><code>#&gt; PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)</code></pre></li>
</ul>
</div>
<div id="imputation" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Imputation<a href="preprocessing.html#imputation" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Note that most functions from bigstatsr and bigsnpr do NOT handle missing values, except for some of the <code>bed_</code> functions.</p>
<p>Simple imputation (e.g. by the mean) of a ‘double’ FBM can be performed by blocks using e.g. the code from <a href="https://privefl.github.io/bigstatsr/articles/big-apply.html">this vignette</a>.</p>
<p>In bigsnpr, to perform simple imputation of genotyped data, you can use <code>snp_fastImputeSimple()</code>. You can also use the (much) slower <code>snp_fastImpute()</code> that uses XGBoost models to impute genotyped data <span class="citation">(<a href="#ref-prive2017efficient">Privé et al., 2018</a>)</span>.
If you have access to imputed data from large external reference panels, it is even better, and you can read this data as dosages in a bigSNP as discussed in <a href="inputs-and-formats.html#getting-bigsnp">2.3</a>.</p>
</div>
<div id="exo-preprocessing" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Example<a href="preprocessing.html#exo-preprocessing" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>For the exercises, we will use the data provided in <span class="citation">Reed et al. (<a href="#ref-reed2015guide">2015</a>)</span>.</p>
<p>This can be downloaded using</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="preprocessing.html#cb63-1" tabindex="-1"></a>zip <span class="ot">&lt;-</span> runonce<span class="sc">::</span><span class="fu">download_file</span>(</span>
<span id="cb63-2"><a href="preprocessing.html#cb63-2" tabindex="-1"></a>  <span class="st">&quot;https://figshare.com/ndownloader/files/38019072&quot;</span>,</span>
<span id="cb63-3"><a href="preprocessing.html#cb63-3" tabindex="-1"></a>  <span class="at">dir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">fname =</span> <span class="st">&quot;GWAS_data.zip&quot;</span>)</span>
<span id="cb63-4"><a href="preprocessing.html#cb63-4" tabindex="-1"></a><span class="fu">unzip</span>(zip, <span class="at">exdir =</span> <span class="st">&quot;tmp-data&quot;</span>, <span class="at">overwrite =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<p>For some reason, this data is not ordered by chromosome and position; we can use PLINK to get an ordered version of this using</p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="preprocessing.html#cb64-1" tabindex="-1"></a><span class="fu">library</span>(bigsnpr)</span>
<span id="cb64-2"><a href="preprocessing.html#cb64-2" tabindex="-1"></a>plink <span class="ot">&lt;-</span> <span class="fu">download_plink</span>(<span class="st">&quot;tmp-data&quot;</span>)</span>
<span id="cb64-3"><a href="preprocessing.html#cb64-3" tabindex="-1"></a><span class="fu">system</span>(glue<span class="sc">::</span><span class="fu">glue</span>(</span>
<span id="cb64-4"><a href="preprocessing.html#cb64-4" tabindex="-1"></a>  <span class="st">&quot;{plink} --bfile tmp-data/GWAS_data&quot;</span>,</span>
<span id="cb64-5"><a href="preprocessing.html#cb64-5" tabindex="-1"></a>  <span class="st">&quot; --make-bed --out tmp-data/GWAS_data_sorted&quot;</span></span>
<span id="cb64-6"><a href="preprocessing.html#cb64-6" tabindex="-1"></a>))</span></code></pre></div>
<pre><code>#&gt; PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/
#&gt; (C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3
#&gt; Logging to tmp-data/GWAS_data_sorted.log.
#&gt; Options in effect:
#&gt;   --bfile tmp-data/GWAS_data
#&gt;   --make-bed
#&gt;   --out tmp-data/GWAS_data_sorted
#&gt; 
#&gt; 32574 MB RAM detected; reserving 16287 MB for main workspace.
#&gt; 500000 variants loaded from .bim file.
#&gt; 1401 people (937 males, 464 females) loaded from .fam.
#&gt; 933 phenotype values loaded from .fam.
#&gt; Using 1 thread (no multithreaded calculations invoked).
#&gt; Before main variant filters, 1401 founders and 0 nonfounders present.
#&gt; Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99% done.
#&gt; Total genotyping rate is 0.977593.
#&gt; 500000 variants and 1401 people pass filters and QC.
#&gt; Among remaining phenotypes, 463 are cases and 470 are controls.  (468
#&gt; phenotypes are missing.)
#&gt; --make-bed to tmp-data/GWAS_data_sorted.bed + tmp-data/GWAS_data_sorted.bim +
#&gt; tmp-data/GWAS_data_sorted.fam ... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.</code></pre>
<p>As you can see from PLINK output, this data contains 1401 individuals and 500,000 variants, with a small percentage of missing values (2.2%).</p>
<hr />
<div class="infobox exo">
<p>Perform some QC with PLINK then read the QCed data as a bigSNP object in R.</p>
</div>
<div class="infobox caution">
<p>Make sure to look at the documentation of functions before using them.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<p>We can then perform some quality control using</p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="preprocessing.html#cb66-1" tabindex="-1"></a>bedfile2 <span class="ot">&lt;-</span> <span class="fu">snp_plinkQC</span>(plink, <span class="st">&quot;tmp-data/GWAS_data_sorted&quot;</span>)</span></code></pre></div>
<pre><code>#&gt; PLINK v1.9.0-b.7.7 64-bit (22 Oct 2024)            cog-genomics.org/plink/1.9/
#&gt; (C) 2005-2024 Shaun Purcell, Christopher Chang   GNU General Public License v3
#&gt; Logging to tmp-data/GWAS_data_sorted_QC.log.
#&gt; Options in effect:
#&gt;   --bfile tmp-data/GWAS_data_sorted
#&gt;   --geno 0.1
#&gt;   --hwe 1e-50
#&gt;   --maf 0.01
#&gt;   --make-bed
#&gt;   --mind 0.1
#&gt;   --out tmp-data/GWAS_data_sorted_QC
#&gt; 
#&gt; 32574 MB RAM detected; reserving 16287 MB for main workspace.
#&gt; 500000 variants loaded from .bim file.
#&gt; 1401 people (937 males, 464 females) loaded from .fam.
#&gt; 933 phenotype values loaded from .fam.
#&gt; 0 people removed due to missing genotype data (--mind).
#&gt; Using 1 thread (no multithreaded calculations invoked).
#&gt; Before main variant filters, 1401 founders and 0 nonfounders present.
#&gt; Calculating allele frequencies... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99% done.
#&gt; Total genotyping rate is 0.977593.
#&gt; 27390 variants removed due to missing genotype data (--geno).
#&gt; --hwe: 91 variants removed due to Hardy-Weinberg exact test.
#&gt; 67856 variants removed due to minor allele threshold(s)
#&gt; (--maf/--max-maf/--mac/--max-mac).
#&gt; 404663 variants and 1401 people pass filters and QC.
#&gt; Among remaining phenotypes, 463 are cases and 470 are controls.  (468
#&gt; phenotypes are missing.)
#&gt; --make-bed to tmp-data/GWAS_data_sorted_QC.bed +
#&gt; tmp-data/GWAS_data_sorted_QC.bim + tmp-data/GWAS_data_sorted_QC.fam ... 0%1%2%3%4%5%6%7%8%9%10%11%12%13%14%15%16%17%18%19%20%21%22%23%24%25%26%27%28%29%30%31%32%33%34%35%36%37%38%39%40%41%42%43%44%45%46%47%48%49%50%51%52%53%54%55%56%57%58%59%60%61%62%63%64%65%66%67%68%69%70%71%72%73%74%75%76%77%78%79%80%81%82%83%84%85%86%87%88%89%90%91%92%93%94%95%96%97%98%99%done.</code></pre>
<p>404,663 variants are remaining after this quality control.</p>
<div class="infobox question">
<p>How to control the amount of memory and number of threads used by PLINK?</p>
<p>Have a search at <a href="https://www.cog-genomics.org/plink/1.9/" class="uri">https://www.cog-genomics.org/plink/1.9/</a> using the quick index search at the bottom of the sidebar on the left.</p>
</div>
<p>We can then read this data into an R object called <code>bigSNP</code> using</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="preprocessing.html#cb68-1" tabindex="-1"></a>(rds <span class="ot">&lt;-</span> <span class="fu">snp_readBed2</span>(bedfile2, <span class="at">ncores =</span> <span class="fu">nb_cores</span>()))</span></code></pre></div>
<pre><code>#&gt; [1] &quot;C:\\Users\\au639593\\OneDrive - Aarhus universitet\\Desktop\\statgen-course\\tmp-data\\GWAS_data_sorted_QC.rds&quot;</code></pre>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="preprocessing.html#cb70-1" tabindex="-1"></a>obj.bigsnp <span class="ot">&lt;-</span> <span class="fu">snp_attach</span>(rds)</span>
<span id="cb70-2"><a href="preprocessing.html#cb70-2" tabindex="-1"></a><span class="fu">str</span>(obj.bigsnp, <span class="at">max.level =</span> <span class="dv">2</span>)</span></code></pre></div>
<pre><code>#&gt; List of 3
#&gt;  $ genotypes:Reference class &#39;FBM.code256&#39; [package &quot;bigstatsr&quot;] with 16 fields
#&gt;   ..and 26 methods, of which 12 are  possibly relevant:
#&gt;   ..  add_columns, as.FBM, bm, bm.desc, check_dimensions,
#&gt;   ..  check_write_permissions, copy#envRefClass, initialize,
#&gt;   ..  initialize#FBM, save, show#envRefClass, show#FBM
#&gt;  $ fam      :&#39;data.frame&#39;:   1401 obs. of  6 variables:
#&gt;   ..$ family.ID  : int [1:1401] 10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...
#&gt;   ..$ sample.ID  : int [1:1401] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ paternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ maternal.ID: int [1:1401] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ sex        : int [1:1401] 1 2 1 1 1 1 1 2 1 2 ...
#&gt;   ..$ affection  : int [1:1401] 1 1 2 1 2 2 2 1 2 -9 ...
#&gt;  $ map      :&#39;data.frame&#39;:   404663 obs. of  6 variables:
#&gt;   ..$ chromosome  : int [1:404663] 1 1 1 1 1 1 1 1 1 1 ...
#&gt;   ..$ marker.ID   : chr [1:404663] &quot;rs12565286&quot; &quot;rs3094315&quot; &quot;rs2980319&quot; &quot;rs2980300&quot; ...
#&gt;   ..$ genetic.dist: int [1:404663] 0 0 0 0 0 0 0 0 0 0 ...
#&gt;   ..$ physical.pos: int [1:404663] 721290 752566 777122 785989 798959 947034 949608 1018704 1041700 1129672 ...
#&gt;   ..$ allele1     : chr [1:404663] &quot;G&quot; &quot;C&quot; &quot;A&quot; &quot;A&quot; ...
#&gt;   ..$ allele2     : chr [1:404663] &quot;C&quot; &quot;T&quot; &quot;T&quot; &quot;G&quot; ...
#&gt;  - attr(*, &quot;class&quot;)= chr &quot;bigSNP&quot;</code></pre>
</details>
<hr />
<p>We can read and store some extra information on the individuals (e.g. some phenotypes):</p>
<div class="sourceCode" id="cb72"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb72-1"><a href="preprocessing.html#cb72-1" tabindex="-1"></a>clinical <span class="ot">&lt;-</span> bigreadr<span class="sc">::</span><span class="fu">fread2</span>(<span class="st">&quot;tmp-data/GWAS_clinical.csv&quot;</span>)</span>
<span id="cb72-2"><a href="preprocessing.html#cb72-2" tabindex="-1"></a><span class="fu">str</span>(clinical)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    1401 obs. of  7 variables:
#&gt;  $ FamID: int  10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...
#&gt;  $ CAD  : int  1 1 1 1 1 1 1 1 1 0 ...
#&gt;  $ sex  : int  1 2 1 1 1 1 1 2 1 2 ...
#&gt;  $ age  : int  60 50 55 52 58 59 54 66 58 67 ...
#&gt;  $ tg   : int  NA 55 105 314 161 171 147 124 60 160 ...
#&gt;  $ hdl  : int  NA 23 37 54 40 46 69 47 114 40 ...
#&gt;  $ ldl  : int  NA 75 69 108 94 92 109 84 67 112 ...</code></pre>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="preprocessing.html#cb74-1" tabindex="-1"></a><span class="co"># Get the same order as for the genotypes</span></span>
<span id="cb74-2"><a href="preprocessing.html#cb74-2" tabindex="-1"></a><span class="co"># (to match over multiple columns, use `vctrs::vec_match()`)</span></span>
<span id="cb74-3"><a href="preprocessing.html#cb74-3" tabindex="-1"></a>ord <span class="ot">&lt;-</span> <span class="fu">match</span>(obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>family.ID, clinical<span class="sc">$</span>FamID)</span>
<span id="cb74-4"><a href="preprocessing.html#cb74-4" tabindex="-1"></a>pheno <span class="ot">&lt;-</span> clinical[ord, ]</span>
<span id="cb74-5"><a href="preprocessing.html#cb74-5" tabindex="-1"></a><span class="co"># Quick check</span></span>
<span id="cb74-6"><a href="preprocessing.html#cb74-6" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">all.equal</span>(obj.bigsnp<span class="sc">$</span>fam<span class="sc">$</span>sex, pheno<span class="sc">$</span>sex))</span>
<span id="cb74-7"><a href="preprocessing.html#cb74-7" tabindex="-1"></a><span class="co"># Update the $fam component</span></span>
<span id="cb74-8"><a href="preprocessing.html#cb74-8" tabindex="-1"></a>obj.bigsnp<span class="sc">$</span>fam <span class="ot">&lt;-</span> <span class="fu">cbind</span>(obj.bigsnp<span class="sc">$</span>fam, pheno[<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)])</span></code></pre></div>
<div class="infobox exo">
<p>Use <code>dplyr::left_join()</code> on <code>obj.bigsnp$fam[1:6]</code> and <code>clinical</code> to get the same information as we have now in <code>obj.bigsnp$fam</code>. Which columns should you use to join from?</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="preprocessing.html#cb75-1" tabindex="-1"></a><span class="fu">str</span>(obj.bigsnp<span class="sc">$</span>fam)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    1401 obs. of  11 variables:
#&gt;  $ family.ID  : int  10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...
#&gt;  $ sample.ID  : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ paternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ maternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ sex        : int  1 2 1 1 1 1 1 2 1 2 ...
#&gt;  $ affection  : int  1 1 2 1 2 2 2 1 2 -9 ...
#&gt;  $ CAD        : int  1 1 1 1 1 1 1 1 1 0 ...
#&gt;  $ age        : int  60 50 55 52 58 59 54 66 58 67 ...
#&gt;  $ tg         : int  NA 55 105 314 161 171 147 124 60 160 ...
#&gt;  $ hdl        : int  NA 23 37 54 40 46 69 47 114 40 ...
#&gt;  $ ldl        : int  NA 75 69 108 94 92 109 84 67 112 ...</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="preprocessing.html#cb77-1" tabindex="-1"></a>fam2 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(obj.bigsnp<span class="sc">$</span>fam[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>], clinical,</span>
<span id="cb77-2"><a href="preprocessing.html#cb77-2" tabindex="-1"></a>                         <span class="at">by =</span> <span class="fu">c</span>(<span class="st">&quot;family.ID&quot;</span> <span class="ot">=</span> <span class="st">&quot;FamID&quot;</span>, <span class="st">&quot;sex&quot;</span>))</span>
<span id="cb77-3"><a href="preprocessing.html#cb77-3" tabindex="-1"></a><span class="fu">str</span>(fam2)</span></code></pre></div>
<pre><code>#&gt; &#39;data.frame&#39;:    1401 obs. of  11 variables:
#&gt;  $ family.ID  : int  10002 10004 10005 10007 10008 10009 10010 10011 10012 10013 ...
#&gt;  $ sample.ID  : int  1 1 1 1 1 1 1 1 1 1 ...
#&gt;  $ paternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ maternal.ID: int  0 0 0 0 0 0 0 0 0 0 ...
#&gt;  $ sex        : int  1 2 1 1 1 1 1 2 1 2 ...
#&gt;  $ affection  : int  1 1 2 1 2 2 2 1 2 -9 ...
#&gt;  $ CAD        : int  1 1 1 1 1 1 1 1 1 0 ...
#&gt;  $ age        : int  60 50 55 52 58 59 54 66 58 67 ...
#&gt;  $ tg         : int  NA 55 105 314 161 171 147 124 60 160 ...
#&gt;  $ hdl        : int  NA 23 37 54 40 46 69 47 114 40 ...
#&gt;  $ ldl        : int  NA 75 69 108 94 92 109 84 67 112 ...</code></pre>
<div class="infobox caution">
<p>Normally, <code>dplyr::left_join()</code> doesn’t change the order of individuals from the first data frame (as opposed to <code>merge()</code>), but you can end up with more rows if there are duplicates in the second data frame.</p>
</div>
</details>
<hr />
<p>Recall that this data contains some missing values.</p>
<div class="infobox exo">
<p>Look at the number of missing values per variant using <code>big_counts()</code> and impute the data using <code>snp_fastImputeSimple()</code>.</p>
</div>
<details>
<summary>
Click to see solution
</summary>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="preprocessing.html#cb79-1" tabindex="-1"></a>G <span class="ot">&lt;-</span> obj.bigsnp<span class="sc">$</span>genotypes</span>
<span id="cb79-2"><a href="preprocessing.html#cb79-2" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">big_counts</span>(G)  <span class="co"># counts per variant</span></span>
<span id="cb79-3"><a href="preprocessing.html#cb79-3" tabindex="-1"></a>counts[, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;   17   15    0   18    1   78   64   69</code></pre>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="preprocessing.html#cb81-1" tabindex="-1"></a><span class="fu">hist</span>(nbNA <span class="ot">&lt;-</span> counts[<span class="dv">4</span>, ])</span></code></pre></div>
<p><img src="preprocessing_files/figure-html/unnamed-chunk-9-1.png" width="70%" style="display: block; margin: auto;" /></p>
<p>We can e.g. perform a quick imputation by the mean using</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="preprocessing.html#cb82-1" tabindex="-1"></a>G2 <span class="ot">&lt;-</span> <span class="fu">snp_fastImputeSimple</span>(G, <span class="at">method =</span> <span class="st">&quot;mean2&quot;</span>, <span class="at">ncores =</span> <span class="fu">nb_cores</span>())</span>
<span id="cb82-2"><a href="preprocessing.html#cb82-2" tabindex="-1"></a><span class="fu">big_counts</span>(G2, <span class="at">ind.col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;    0    0    0    0    0    0    0    0
#&gt; 0.01    0    0    0    0    0    0    0    0
#&gt; 0.02    0    0    0    0    0    0    0    0
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 196 rows ]</code></pre>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="preprocessing.html#cb84-1" tabindex="-1"></a><span class="fu">big_counts</span>(G, <span class="at">ind.col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>)</span></code></pre></div>
<pre><code>#&gt;      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
#&gt; 0    1247  958 1057  988  831 1201  496  386
#&gt; 1     131  362  316  370  502  115  676  730
#&gt; 2       6   66   28   25   67    7  165  216
#&gt; &lt;NA&gt;   17   15    0   18    1   78   64   69</code></pre>
<div class="infobox caution">
<p><code>G</code> still has missing values, but <code>G2</code> does not. Note that both use the same underlying data (the same binary file <code>.bk</code> on disk), the difference is that they use a different code to decode the underlying data:</p>
</div>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="preprocessing.html#cb86-1" tabindex="-1"></a>G<span class="sc">$</span>code256</span></code></pre></div>
<pre><code>#&gt;  [1]  0  1  2 NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
#&gt; [24] NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA NA
#&gt; [47] NA NA NA NA
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 206 entries ]</code></pre>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="preprocessing.html#cb88-1" tabindex="-1"></a>G2<span class="sc">$</span>code256</span></code></pre></div>
<pre><code>#&gt;  [1] 0.00 1.00 2.00   NA 0.00 1.00 2.00 0.00 0.01 0.02 0.03 0.04 0.05 0.06
#&gt; [15] 0.07 0.08 0.09 0.10 0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18 0.19 0.20
#&gt; [29] 0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.30 0.31 0.32 0.33 0.34
#&gt; [43] 0.35 0.36 0.37 0.38 0.39 0.40 0.41 0.42
#&gt;  [ reached getOption(&quot;max.print&quot;) -- omitted 206 entries ]</code></pre>
</details>
<hr />
<p>To always use <code>G2</code> (with the new <code>code256</code>) and the extended <code>obj.bigsnp$fam</code>, you need to save the new <code>obj.bigsnp</code> using</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="preprocessing.html#cb90-1" tabindex="-1"></a>obj.bigsnp<span class="sc">$</span>genotypes <span class="ot">&lt;-</span> G2</span>
<span id="cb90-2"><a href="preprocessing.html#cb90-2" tabindex="-1"></a><span class="fu">snp_save</span>(obj.bigsnp)</span></code></pre></div>
<p>You can then re-attach this data in another R session later using <code>snp_attach("tmp-data/GWAS_data_sorted_QC.rds")</code>.</p>

</div>
</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-chang2015second" class="csl-entry">
Chang, C.C., Chow, C.C., Tellier, L.C., Vattikuti, S., Purcell, S.M., &amp; Lee, J.J. (2015). <a href="https://doi.org/10.1186/s13742-015-0047-8">Second-generation <span>PLINK</span>: Rising to the challenge of larger and richer datasets</a>. <em>Gigascience</em>, <em>4</em>, s13742–015.
</div>
<div id="ref-manichaikul2010robust" class="csl-entry">
Manichaikul, A., Mychaleckyj, J.C., Rich, S.S., Daly, K., Sale, M., &amp; Chen, W.-M. (2010). <a href="https://doi.org/10.1093/bioinformatics/btq559">Robust relationship inference in genome-wide association studies</a>. <em>Bioinformatics</em>, <em>26</em>, 2867–2873.
</div>
<div id="ref-prive2017efficient" class="csl-entry">
Privé, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781–2787.
</div>
<div id="ref-reed2015guide" class="csl-entry">
Reed, E., Nunez, S., Kulp, D., Qian, J., Reilly, M.P., &amp; Foulkes, A.S. (2015). <a href="https://doi.org/10.1002/sim.6605">A guide to genome-wide association analysis and post-analytic interrogation</a>. <em>Statistics in Medicine</em>, <em>34</em>, 3769–3792.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="inputs-and-formats.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="population-structure.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
