<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>References | Statistical Human Genetics course using R</title>
  <meta name="description" content="References | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="References | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="References | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-06-22" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="some-other-analyses.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#main-author"><i class="fa fa-check"></i><b>1.3</b> Main author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.1</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.2</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.3</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.4</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.5</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.5.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.5.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.5.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.5.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.5.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.5.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.5.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.5.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.6</b> Exercise</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
<li class="chapter" data-level="5.4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses"><i class="fa fa-check"></i><b>5.4</b> Some post GWAS analyses</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.1</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.2</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.2.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.2.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.2.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.3</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Future of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.5</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.5.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.5.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.5.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.5.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.5.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.5.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create an LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="some-other-analyses.html"><a href="some-other-analyses.html"><i class="fa fa-check"></i><b>10</b> Some other analyses</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="references" class="section level1 unnumbered hasAnchor">
<h1>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h1>

<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div class="csl-entry">
Abdellaoui, A., Hottenga, J.-J., Knijff, P. de, Nivard, M.G., Xiao, X., Scheet, P., et al. (2013). <a href="https://doi.org/10.1038/ejhg.2013.48">Population structure, migration, and diversifying selection in the <span>Netherlands</span></a>. <em>European Journal of Human Genetics</em>, <em>21</em>, 1277–1285.
</div>
<div class="csl-entry">
Aschard, H., Vilhjálmsson, B.J., Joshi, A.D., Price, A.L., &amp; Kraft, P. (2015). <a href="https://doi.org/10.1016%2Fj.ajhg.2014.12.021">Adjusting for heritable covariates can bias effect estimates in genome-wide association studies</a>. <em>The American Journal of Human Genetics</em>, <em>96</em>, 329–339.
</div>
<div class="csl-entry">
Bulik-Sullivan, B., Finucane, H.K., Anttila, V., Gusev, A., Day, F.R., Loh, P.-R., et al. (2015). <a href="https://doi.org/10.1038/ng.3406">An atlas of genetic correlations across human diseases and traits</a>. <em>Nature Genetics</em>, <em>47</em>, 1236–1241.
</div>
<div class="csl-entry">
Bulik-Sullivan, B.K., Loh, P.-R., Finucane, H.K., Ripke, S., Yang, J., Psychiatric Genomics Consortium, S.W.G. of the, et al. (2015). <a href="https://doi.org/10.1038/ng.3211"><span>LD</span> score regression distinguishes confounding from polygenicity in genome-wide association studies</a>. <em>Nature Genetics</em>, <em>47</em>, 291–295.
</div>
<div class="csl-entry">
Bycroft, C., Freeman, C., Petkova, D., Band, G., Elliott, L.T., Sharp, K., et al. (2018). <a href="https://doi.org/10.1038/s41586-018-0579-z">The <span>UK Biobank</span> resource with deep phenotyping and genomic data</a>. <em>Nature</em>, <em>562</em>, 203–209.
</div>
<div class="csl-entry">
Chang, C.C., Chow, C.C., Tellier, L.C., Vattikuti, S., Purcell, S.M., &amp; Lee, J.J. (2015). <a href="https://doi.org/10.1186/s13742-015-0047-8">Second-generation <span>PLINK</span>: Rising to the challenge of larger and richer datasets</a>. <em>Gigascience</em>, <em>4</em>, s13742–015.
</div>
<div class="csl-entry">
Chen, W., Wu, Y., Zheng, Z., Qi, T., Visscher, P.M., Zhu, Z., &amp; Yang, J. (2021). <a href="https://doi.org/10.1038/s41467-021-27438-7">Improved analyses of <span>GWAS</span> summary statistics by reducing data heterogeneity and errors</a>. <em>Nature Communications</em>, <em>12</em>, 1–10.
</div>
<div class="csl-entry">
Choi, S.W., &amp; O’Reilly, P.F. (2019). <a href="https://doi.org/10.1093/gigascience/giz082"><span>PRSice-2</span>: Polygenic risk score software for biobank-scale data</a>. <em>Gigascience</em>, <em>8</em>, giz082.
</div>
<div class="csl-entry">
Daetwyler, H.D., Villanueva, B., &amp; Woolliams, J.A. (2008). <a href="https://doi.org/10.1371/journal.pone.0003395">Accuracy of predicting the genetic risk of disease using a genome-wide approach</a>. <em>PloS One</em>, <em>3</em>, e3395.
</div>
<div class="csl-entry">
Day, F.R., Loh, P.-R., Scott, R.A., Ong, K.K., &amp; Perry, J.R. (2016). <a href="https://doi.org/10.1016/j.ajhg.2015.12.019">A robust example of collider bias in a genetic association study</a>. <em>The American Journal of Human Genetics</em>, <em>98</em>, 392–393.
</div>
<div class="csl-entry">
Ding, Y., Hou, K., Xu, Z., Pimplaskar, A., Petter, E., Boulier, K., et al. (2023). <a href="https://doi.org/10.1038/s41586-023-06079-4">Polygenic scoring accuracy varies across the genetic ancestry continuum</a>. <em>Nature</em>, <em>618</em>, 774–781.
</div>
<div class="csl-entry">
Finucane, H.K., Bulik-Sullivan, B., Gusev, A., Trynka, G., Reshef, Y., Loh, P.-R., et al. (2015). <a href="https://doi.org/10.1038/ng.3404">Partitioning heritability by functional annotation using genome-wide association summary statistics</a>. <em>Nature Genetics</em>, <em>47</em>, 1228–1235.
</div>
<div class="csl-entry">
Gamazon, E.R., Wheeler, H.E., Shah, K.P., Mozaffari, S.V., Aquino-Michaels, K., Carroll, R.J., et al. (2015). <a href="https://doi.org/10.1038/ng.3367">A gene-based association method for mapping traits using reference transcriptome data</a>. <em>Nature Genetics</em>, <em>47</em>, 1091–1098.
</div>
<div class="csl-entry">
Gazal, S., Loh, P.-R., Finucane, H.K., Ganna, A., Schoech, A., Sunyaev, S., &amp; Price, A.L. (2018). <a href="https://doi.org/10.1038/s41588-018-0231-8">Functional architecture of low-frequency variants highlights strength of negative selection across coding and non-coding annotations</a>. <em>Nature Genetics</em>, <em>50</em>, 1600–1607.
</div>
<div class="csl-entry">
Ge, T., Chen, C.-Y., Ni, Y., Feng, Y.-C.A., &amp; Smoller, J.W. (2019). <a href="https://doi.org/10.1038/s41467-019-09718-5">Polygenic prediction via <span>Bayesian</span> regression and continuous shrinkage priors</a>. <em>Nature Communications</em>, <em>10</em>, 1776.
</div>
<div class="csl-entry">
Grinde, K.E., Browning, B.L., Reiner, A.P., Thornton, T.A., &amp; Browning, S.R. (2024). <a href="https://doi.org/10.1371/journal.pgen.1011242">Adjusting for principal components can induce collider bias in genome-wide association studies</a>. <em>PLoS Genetics</em>, <em>20</em>, e1011242.
</div>
<div class="csl-entry">
Grotzinger, A.D., Fuente, J. de la, Privé, F., Nivard, M.G., &amp; Tucker-Drob, E.M. (2023). <a href="https://doi.org/10.1016/j.biopsych.2022.05.029">Pervasive downward bias in estimates of liability-scale heritability in genome-wide association study meta-analysis: A simple solution</a>. <em>Biological Psychiatry</em>, <em>93</em>, 29–36.
</div>
<div class="csl-entry">
Grotzinger, A.D., Rhemtulla, M., Vlaming, R. de, Ritchie, S.J., Mallard, T.T., Hill, W.D., et al. (2019). <a href="https://doi.org/10.1038/s41562-019-0566-x">Genomic structural equation modelling provides insights into the multivariate genetic architecture of complex traits</a>. <em>Nature Human Behaviour</em>, <em>3</em>, 513–525.
</div>
<div class="csl-entry">
Gusev, A., Ko, A., Shi, H., Bhatia, G., Chung, W., Penninx, B.W., et al. (2016). <a href="https://doi.org/10.1038/ng.3506">Integrative approaches for large-scale transcriptome-wide association studies</a>. <em>Nature Genetics</em>, <em>48</em>, 245–252.
</div>
<div class="csl-entry">
Hou, K., Ding, Y., Xu, Z., Wu, Y., Bhattacharya, A., Mester, R., et al. (2023). <a href="https://doi.org/10.1038/s41588-023-01338-6">Causal effects on complex traits are similar for common variants across segments of different continental ancestries within admixed individuals</a>. <em>Nature Genetics</em>, <em>55</em>, 549–558.
</div>
<div class="csl-entry">
Hu, S., Ferreira, L.A., Shi, S., Hellenthal, G., Marchini, J., Lawson, D.J., &amp; Myers, S.R. (2025). <a href="https://doi.org/10.1038/s41588-024-02035-8">Fine-scale population structure and widespread conservation of genetic effect sizes between human groups across traits</a>. <em>Nature Genetics</em>, 1–11.
</div>
<div class="csl-entry">
Julienne, H., Laville, V., McCaw, Z.R., He, Z., Guillemot, V., Lasry, C., et al. (2021). <a href="https://doi.org/10.1371/journal.pgen.1009713">Multitrait <span>GWAS</span> to connect disease variants and biological mechanisms</a>. <em>PLoS Genetics</em>, <em>17</em>, e1009713.
</div>
<div class="csl-entry">
Kanai, M., Elzur, R., Zhou, W., Wu, K.-H.H., Rasheed, H., Tsuo, K., et al. (2022). <a href="https://doi.org/10.1016/j.xgen.2022.100210">Meta-analysis fine-mapping is often miscalibrated at single-variant resolution</a>. <em>Cell Genomics</em>, <em>2</em>, 100210.
</div>
<div class="csl-entry">
Kelemen, M., Xu, Y., Jiang, T., Zhao, J.H., Anderson, C.A., Wallace, C., et al. (2025). <a href="https://doi.org/10.1038/s41467-025-60056-1">Performance of deep-learning-based approaches to improve polygenic scores</a>. <em>Nature Communications</em>, <em>16</em>, 1–9.
</div>
<div class="csl-entry">
Levey, D.F., Stein, M.B., Wendt, F.R., Pathak, G.A., Zhou, H., Aslan, M., et al. (2021). <a href="https://doi.org/10.1038/s41593-021-00860-2">Bi-ancestral depression <span>GWAS</span> in the <span>Million Veteran Program</span> and meta-analysis in&gt; 1.2 million individuals highlight new therapeutic directions</a>. <em>Nature Neuroscience</em>, <em>24</em>, 954–963.
</div>
<div class="csl-entry">
Li, A., Liu, S., Bakshi, A., Jiang, L., Chen, W., Zheng, Z., et al. (2023). <a href="https://doi.org/10.1016/j.ajhg.2022.12.006"><span class="nocase">mBAT-combo</span>: A more powerful test to detect gene-trait associations from <span>GWAS</span> data</a>. <em>The American Journal of Human Genetics</em>, <em>110</em>, 30–43.
</div>
<div class="csl-entry">
Lloyd-Jones, L.R., Zeng, J., Sidorenko, J., Yengo, L., Moser, G., Kemper, K.E., et al. (2019). <a href="https://doi.org/10.1038/s41467-019-12653-0">Improved polygenic prediction by <span>Bayesian</span> multiple regression on summary statistics</a>. <em>Nature Communications</em>, <em>10</em>, 5086.
</div>
<div class="csl-entry">
Loh, P.-R., Kichaev, G., Gazal, S., Schoech, A.P., &amp; Price, A.L. (2018). <a href="https://doi.org/10.1038/s41588-018-0144-6">Mixed-model association for biobank-scale datasets</a>. <em>Nature Genetics</em>, <em>50</em>, 906–908.
</div>
<div class="csl-entry">
Mägi, R., Horikoshi, M., Sofer, T., Mahajan, A., Kitajima, H., Franceschini, N., et al. (2017). <a href="https://doi.org/10.1093/hmg/ddx280">Trans-ethnic meta-regression of genome-wide association studies accounting for ancestry increases power for discovery and improves fine-mapping resolution</a>. <em>Human Molecular Genetics</em>, <em>26</em>, 3639–3650.
</div>
<div class="csl-entry">
Mak, T.S.H., Porsch, R.M., Choi, S.W., Zhou, X., &amp; Sham, P.C. (2017). <a href="https://doi.org/10.1002/gepi.22050">Polygenic scores via penalized regression on summary statistics</a>. <em>Genetic Epidemiology</em>, <em>41</em>, 469–480.
</div>
<div class="csl-entry">
Manichaikul, A., Mychaleckyj, J.C., Rich, S.S., Daly, K., Sale, M., &amp; Chen, W.-M. (2010). <a href="https://doi.org/10.1093/bioinformatics/btq559">Robust relationship inference in genome-wide association studies</a>. <em>Bioinformatics</em>, <em>26</em>, 2867–2873.
</div>
<div class="csl-entry">
Mbatchou, J., Barnard, L., Backman, J., Marcketta, A., Kosmicki, J.A., Ziyatdinov, A., et al. (2021). <a href="https://doi.org/10.1038/s41588-021-00870-7">Computationally efficient whole-genome regression for quantitative and binary traits</a>. <em>Nature Genetics</em>, <em>53</em>, 1097–1103.
</div>
<div class="csl-entry">
Murphy, A.E., Schilder, B.M., &amp; Skene, N.G. (2021). <a href="https://doi.org/10.1093/bioinformatics/btab665"><span>MungeSumstats</span>: A <span>Bioconductor</span> package for the standardization and quality control of many <span>GWAS</span> summary statistics</a>. <em>Bioinformatics</em>, <em>37</em>, 4593–4596.
</div>
<div class="csl-entry">
Okbay, A., Wu, Y., Wang, N., Jayashankar, H., Bennett, M., Nehzati, S.M., et al. (2022). <a href="https://doi.org/10.1038/s41588-022-01016-z">Polygenic prediction of educational attainment within and between families from genome-wide association analyses in 3 million individuals</a>. <em>Nature Genetics</em>, <em>54</em>, 437–449.
</div>
<div class="csl-entry">
Price, A.L., Weale, M.E., Patterson, N., Myers, S.R., Need, A.C., Shianna, K.V., et al. (2008). <a href="https://doi.org/10.1016/j.ajhg.2008.06.005">Long-range <span>LD</span> can confound genome scans in admixed populations</a>. <em>The American Journal of Human Genetics</em>, <em>83</em>, 132–135.
</div>
<div class="csl-entry">
Pritchard, J.K., &amp; Przeworski, M. (2001). <a href="https://doi.org/10.1086/321275">Linkage disequilibrium in humans: Models and data</a>. <em>The American Journal of Human Genetics</em>, <em>69</em>, 1–14.
</div>
<div class="csl-entry">
Privé, F. (2021). <a href="https://doi.org/10.1093/bioinformatics/btab519"><span class="nocase">Optimal linkage disequilibrium splitting</span></a>. <em>Bioinformatics</em>, <em>38</em>, 255–256.
</div>
<div class="csl-entry">
Privé, F. (2022). <a href="https://doi.org/10.1093/bioinformatics/btac348"><span class="nocase">Using the UK Biobank as a global reference of worldwide populations: application to measuring ancestry diversity from GWAS summary statistics</span></a>. <em>Bioinformatics</em>, <em>38</em>, 3477–3480.
</div>
<div class="csl-entry">
Privé, F., Albiñana, C., Arbel, J., Pasaniuc, B., &amp; Vilhjálmsson, B.J. (2023). <a href="https://doi.org/10.1016/j.ajhg.2023.10.010">Inferring disease architecture and predictive ability with <span class="nocase">LDpred2-auto</span></a>. <em>The American Journal of Human Genetics</em>, <em>110</em>, 2042–2055.
</div>
<div class="csl-entry">
Privé, F., Arbel, J., Aschard, H., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.xhgg.2022.100136">Identifying and correcting for misspecifications in <span>GWAS</span> summary statistics and polygenic scores</a>. <em>Human Genetics and Genomics Advances</em>, <em>3</em>, 100136.
</div>
<div class="csl-entry">
Privé, F., Arbel, J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa1029"><span class="nocase"><span>LDpred2</span>: better, faster, stronger</span></a>. <em>Bioinformatics</em>, <em>36</em>, 5424–5431.
</div>
<div class="csl-entry">
Privé, F., Aschard, H., &amp; Blum, M.G. (2019). <a href="https://doi.org/10.1534/genetics.119.302019">Efficient implementation of penalized regression for genetic risk prediction</a>. <em>Genetics</em>, <em>212</em>, 65–74.
</div>
<div class="csl-entry">
Privé, F., Aschard, H., Carmi, S., Folkersen, L., Hoggart, C., O’Reilly, P.F., &amp; Vilhjálmsson, B.J. (2022). <a href="https://doi.org/10.1016/j.ajhg.2021.11.008">Portability of 245 polygenic scores when derived from the <span>UK Biobank</span> and applied to 9 ancestry groups from the same cohort</a>. <em>The American Journal of Human Genetics</em>, <em>109</em>, 12–23.
</div>
<div class="csl-entry">
Privé, F., Aschard, H., Ziyatdinov, A., &amp; Blum, M.G.B. (2018). <a href="https://doi.org/10.1093/bioinformatics/bty185">Efficient analysis of large-scale genome-wide data with two <span>R</span> packages: <span class="nocase">bigstatsr</span> and <span class="nocase">bigsnpr</span></a>. <em>Bioinformatics</em>, <em>34</em>, 2781–2787.
</div>
<div class="csl-entry">
Privé, F., Luu, K., Blum, M.G., McGrath, J.J., &amp; Vilhjálmsson, B.J. (2020). <a href="https://doi.org/10.1093/bioinformatics/btaa520">Efficient toolkit implementing best practices for principal component analysis of population genetic data</a>. <em>Bioinformatics</em>, <em>36</em>, 4449–4457.
</div>
<div class="csl-entry">
Privé, F., Luu, K., Vilhjálmsson, B.J., &amp; Blum, M.G. (2020). <a href="https://doi.org/10.1093/molbev/msaa053">Performing highly efficient genome scans for local adaptation with <span>R</span> package pcadapt version 4</a>. <em>Molecular Biology and Evolution</em>, <em>37</em>, 2153–2154.
</div>
<div class="csl-entry">
Privé, F., Vilhjálmsson, B.J., Aschard, H., &amp; Blum, M.G.B. (2019). <a href="https://doi.org/10.1016/j.ajhg.2019.11.001">Making the most of clumping and thresholding for polygenic scores</a>. <em>The American Journal of Human Genetics</em>, <em>105</em>, 1213–1221.
</div>
<div class="csl-entry">
Reed, E., Nunez, S., Kulp, D., Qian, J., Reilly, M.P., &amp; Foulkes, A.S. (2015). <a href="https://doi.org/10.1002/sim.6605">A guide to genome-wide association analysis and post-analytic interrogation</a>. <em>Statistics in Medicine</em>, <em>34</em>, 3769–3792.
</div>
<div class="csl-entry">
Ruth, K.S., Day, F.R., Tyrrell, J., Thompson, D.J., Wood, A.R., Mahajan, A., et al. (2020). <a href="https://doi.org/10.1038/s41591-020-0751-5">Using human genetics to understand the disease impacts of testosterone in men and women</a>. <em>Nature Medicine</em>, <em>26</em>, 252–258.
</div>
<div class="csl-entry">
Speed, D., &amp; Balding, D.J. (2019). <a href="https://doi.org/10.1038/s41588-018-0279-5"><span>SumHer</span> better estimates the <span>SNP</span> heritability of complex traits from summary statistics</a>. <em>Nature Genetics</em>, <em>51</em>, 277–284.
</div>
<div class="csl-entry">
Suzuki, K., Hatzikotoulas, K., Southam, L., Taylor, H.J., Yin, X., Lorenz, K.M., et al. (2023). Multi-ancestry genome-wide study in &gt; 2.5 million individuals reveals heterogeneity in mechanistic pathways of type 2 diabetes and complications. <em>medRxiv</em>. Retrieved from <a href="https://doi.org/10.1101/2023.03.31.23287839">https://doi.org/10.1101/2023.03.31.23287839</a>
</div>
<div class="csl-entry">
Turley, P., Walters, R.K., Maghzian, O., Okbay, A., Lee, J.J., Fontana, M.A., et al. (2018). <a href="https://doi.org/10.1038/s41588-017-0009-4">Multi-trait analysis of genome-wide association summary statistics using <span>MTAG</span></a>. <em>Nature Genetics</em>, <em>50</em>, 229–237.
</div>
<div class="csl-entry">
Wallace, C. (2021). <a href="https://doi.org/10.1371/journal.pgen.1009440">A more accurate method for colocalisation analysis allowing for multiple causal variants</a>. <em>PLoS Genetics</em>, <em>17</em>, e1009440.
</div>
<div class="csl-entry">
Wang, G., Sarkar, A., Carbonetto, P., &amp; Stephens, M. (2020). <a href="https://doi.org/10.1111/rssb.12388">A simple new approach to variable selection in regression, with application to genetic fine mapping</a>. <em>Journal of the Royal Statistical Society Series B: Statistical Methodology</em>, <em>82</em>, 1273–1300.
</div>
<div class="csl-entry">
Wang, X., Walker, A., Revez, J.A., Ni, G., Adams, M.J., McIntosh, A.M., et al. (2023). <a href="https://doi.org/10.1016/j.ajhg.2023.06.006">Polygenic risk prediction: Why and when out-of-sample prediction <span>R2</span> can exceed <span>SNP</span>-based heritability</a>. <em>The American Journal of Human Genetics</em>, <em>110</em>, 1207–1215.
</div>
<div class="csl-entry">
Wang, Y., Guo, J., Ni, G., Yang, J., Visscher, P.M., &amp; Yengo, L. (2020). <a href="https://doi.org/10.1038/s41467-020-17719-y">Theoretical and empirical quantification of the accuracy of polygenic scores in ancestry divergent populations</a>. <em>Nature Communications</em>, <em>11</em>, 3865.
</div>
<div class="csl-entry">
Willer, C.J., Li, Y., &amp; Abecasis, G.R. (2010). <a href="https://doi.org/10.1093/bioinformatics/btq340"><span>METAL</span>: Fast and efficient meta-analysis of genomewide association scans</a>. <em>Bioinformatics</em>, <em>26</em>, 2190–2191.
</div>
<div class="csl-entry">
Yang, J., Ferreira, T., Morris, A.P., Medland, S.E., ANthropometric Traits (GIANT) Consortium, G.I. of, Consortium, D.G.R.A.M. (DIAGRAM), et al. (2012). <a href="https://doi.org/10.1038/ng.2213">Conditional and joint multiple-<span>SNP</span> analysis of <span>GWAS</span> summary statistics identifies additional variants influencing complex traits</a>. <em>Nature Genetics</em>, <em>44</em>, 369–375.
</div>
<div class="csl-entry">
Yang, J., Weedon, M.N., Purcell, S., Lettre, G., Estrada, K., Willer, C.J., et al. (2011). <a href="https://doi.org/10.1038/ejhg.2011.39">Genomic inflation factors under polygenic inheritance</a>. <em>European Journal of Human Genetics</em>, <em>19</em>, 807–812.
</div>
<div class="csl-entry">
Yengo, L., Vedantam, S., Marouli, E., Sidorenko, J., Bartell, E., Sakaue, S., et al. (2022). <a href="https://doi.org/10.1038/s41586-022-05275-y">A saturated map of common genetic variants associated with human height</a>. <em>Nature</em>, <em>610</em>, 704–712.
</div>
<div class="csl-entry">
Zabad, S., Haryan, C.A., Gravel, S., Misra, S., &amp; Li, Y. (2025). Towards whole-genome inference of polygenic scores with fast and memory-efficient algorithms. <em>The American Journal of Human Genetics</em>. Retrieved from <a href="https://doi.org/10.1016/j.ajhg.2025.05.002">https://doi.org/10.1016/j.ajhg.2025.05.002</a>
</div>
<div class="csl-entry">
Zhang, C., Zhang, Y., Zhang, Y., &amp; Zhao, H. (2023). <a href="https://doi.org/10.1093/bib/bbad407">Benchmarking of local genetic correlation estimation methods using summary statistics from genome-wide association studies</a>. <em>Briefings in Bioinformatics</em>, <em>24</em>, bbad407.
</div>
<div class="csl-entry">
Zheng, Z., Liu, S., Sidorenko, J., Wang, Y., Lin, T., Yengo, L., et al. (2024). <a href="https://doi.org/10.1038/s41588-024-01704-y">Leveraging functional genomic annotations and genome coverage to improve polygenic prediction of complex traits within and between ancestries</a>. <em>Nature Genetics</em>, <em>56</em>, 767–777.
</div>
<div class="csl-entry">
Zhou, W., Kanai, M., Wu, K.-H.H., Rasheed, H., Tsuo, K., Hirbo, J.B., et al. (2022). <a href="https://doi.org/10.1016/j.xgen.2022.100192"><span class="nocase">Global Biobank Meta-analysis Initiative</span>: Powering genetic discovery across human disease</a>. <em>Cell Genomics</em>, <em>2</em>, 100192.
</div>
<div class="csl-entry">
Zhu, Z., Zhang, F., Hu, H., Bakshi, A., Robinson, M.R., Powell, J.E., et al. (2016). <a href="https://doi.org/10.1038/ng.3538">Integration of summary data from <span>GWAS</span> and <span class="nocase">eQTL</span> studies predicts complex trait gene targets</a>. <em>Nature Genetics</em>, <em>48</em>, 481–487.
</div>
<div class="csl-entry">
Zou, Y., Carbonetto, P., Wang, G., &amp; Stephens, M. (2022). <a href="https://doi.org/10.1371/journal.pgen.1010299">Fine-mapping from summary data with the <span class="nocase">"Sum of Single Effects"</span> model</a>. <em>PLoS Genetics</em>, <em>18</em>, e1010299.
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="some-other-analyses.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
