<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 10 Some other analyses | Statistical Human Genetics course using R</title>
  <meta name="description" content="Chapter 10 Some other analyses | Statistical Human Genetics course using R" />
  <meta name="generator" content="bookdown 0.42 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 10 Some other analyses | Statistical Human Genetics course using R" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="privefl/statgen-course" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 10 Some other analyses | Statistical Human Genetics course using R" />
  
  
  

<meta name="author" content="Florian Privé &amp; Isabelle McGrath" />


<meta name="date" content="2025-07-04" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="fine-mapping-with-susie.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./index.html">Statistical Human Genetics course using R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Course information</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.1</b> Prerequisites</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-of-smartbiomed-summer-school"><i class="fa fa-check"></i><b>1.2</b> Schedule of SMARTbiomed Summer school</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#main-author"><i class="fa fa-check"></i><b>1.3</b> Main author</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.4</b> Contact</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html"><i class="fa fa-check"></i><b>2</b> Inputs and formats</a>
<ul>
<li class="chapter" data-level="2.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#imputed-data"><i class="fa fa-check"></i><b>2.1</b> Imputed data</a></li>
<li class="chapter" data-level="2.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#formats-of-genetic-data"><i class="fa fa-check"></i><b>2.2</b> Formats of genetic data</a></li>
<li class="chapter" data-level="2.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#main-motivation-for-developing-r-packages-bigstatsr-and-bigsnpr"><i class="fa fa-check"></i><b>2.3</b> Main motivation for developing R packages bigstatsr and bigsnpr</a></li>
<li class="chapter" data-level="2.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-bigsnp-format-from-bigsnpr"><i class="fa fa-check"></i><b>2.4</b> The bigSNP format from bigsnpr</a></li>
<li class="chapter" data-level="2.5" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#getting-bigsnp"><i class="fa fa-check"></i><b>2.5</b> Getting a bigSNP object</a></li>
<li class="chapter" data-level="2.6" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#the-fbm-format-from-bigstatsr"><i class="fa fa-check"></i><b>2.6</b> The FBM format from bigstatsr</a></li>
<li class="chapter" data-level="2.7" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#working-with-an-fbm"><i class="fa fa-check"></i><b>2.7</b> Working with an FBM</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-r-matrices"><i class="fa fa-check"></i><b>2.7.1</b> Similar accessor as R matrices</a></li>
<li class="chapter" data-level="2.7.2" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#split-parapply-combine-strategy"><i class="fa fa-check"></i><b>2.7.2</b> Split-(par)Apply-Combine Strategy</a></li>
<li class="chapter" data-level="2.7.3" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#similar-accessor-as-rcpp-matrices"><i class="fa fa-check"></i><b>2.7.3</b> Similar accessor as Rcpp matrices</a></li>
<li class="chapter" data-level="2.7.4" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#some-summary-functions-are-already-implemented"><i class="fa fa-check"></i><b>2.7.4</b> Some summary functions are already implemented</a></li>
</ul></li>
<li class="chapter" data-level="2.8" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#exercise"><i class="fa fa-check"></i><b>2.8</b> Exercise</a></li>
<li class="chapter" data-level="2.9" data-path="inputs-and-formats.html"><a href="inputs-and-formats.html#matching-genetic-variants-between-datasets"><i class="fa fa-check"></i><b>2.9</b> Matching genetic variants between datasets</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="preprocessing.html"><a href="preprocessing.html"><i class="fa fa-check"></i><b>3</b> Preprocessing</a>
<ul>
<li class="chapter" data-level="3.1" data-path="preprocessing.html"><a href="preprocessing.html#PLINK"><i class="fa fa-check"></i><b>3.1</b> Conversion and quality control of PLINK files</a></li>
<li class="chapter" data-level="3.2" data-path="preprocessing.html"><a href="preprocessing.html#imputation"><i class="fa fa-check"></i><b>3.2</b> Imputation</a></li>
<li class="chapter" data-level="3.3" data-path="preprocessing.html"><a href="preprocessing.html#exo-preprocessing"><i class="fa fa-check"></i><b>3.3</b> Example</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="population-structure.html"><a href="population-structure.html"><i class="fa fa-check"></i><b>4</b> Population structure</a>
<ul>
<li class="chapter" data-level="4.1" data-path="population-structure.html"><a href="population-structure.html#principal-component-analysis-pca"><i class="fa fa-check"></i><b>4.1</b> Principal Component Analysis (PCA)</a></li>
<li class="chapter" data-level="4.2" data-path="population-structure.html"><a href="population-structure.html#the-problem-with-ld"><i class="fa fa-check"></i><b>4.2</b> The problem with LD</a></li>
<li class="chapter" data-level="4.3" data-path="population-structure.html"><a href="population-structure.html#best-practices-for-pca-of-genetic-data"><i class="fa fa-check"></i><b>4.3</b> Best practices for PCA of genetic data</a></li>
<li class="chapter" data-level="4.4" data-path="population-structure.html"><a href="population-structure.html#exo-pca"><i class="fa fa-check"></i><b>4.4</b> Exercise</a></li>
<li class="chapter" data-level="4.5" data-path="population-structure.html"><a href="population-structure.html#ancestry"><i class="fa fa-check"></i><b>4.5</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html"><i class="fa fa-check"></i><b>5</b> Genome-Wide Association Study (GWAS)</a>
<ul>
<li class="chapter" data-level="5.1" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#exo-gwas"><i class="fa fa-check"></i><b>5.1</b> Example</a></li>
<li class="chapter" data-level="5.2" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#regenie"><i class="fa fa-check"></i><b>5.2</b> REGENIE</a></li>
<li class="chapter" data-level="5.3" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#genomic-control-and-ld-score-regression"><i class="fa fa-check"></i><b>5.3</b> Genomic control and LD score regression</a></li>
<li class="chapter" data-level="5.4" data-path="genome-wide-association-study-gwas.html"><a href="genome-wide-association-study-gwas.html#some-post-gwas-analyses"><i class="fa fa-check"></i><b>5.4</b> Some post GWAS analyses</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html"><i class="fa fa-check"></i><b>6</b> GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="6.1" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#format"><i class="fa fa-check"></i><b>6.1</b> Format</a></li>
<li class="chapter" data-level="6.2" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#quality-control"><i class="fa fa-check"></i><b>6.2</b> Quality control</a></li>
<li class="chapter" data-level="6.3" data-path="gwas-summary-statistics.html"><a href="gwas-summary-statistics.html#ssancestry"><i class="fa fa-check"></i><b>6.3</b> Ancestry inference</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html"><i class="fa fa-check"></i><b>7</b> Linkage Disequilibrium (LD)</a>
<ul>
<li class="chapter" data-level="7.1" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#derivation-of-an-ld-matrix"><i class="fa fa-check"></i><b>7.1</b> Derivation of an LD matrix</a></li>
<li class="chapter" data-level="7.2" data-path="linkage-disequilibrium-ld.html"><a href="linkage-disequilibrium-ld.html#subtleties-about-ld-matrices"><i class="fa fa-check"></i><b>7.2</b> Subtleties about LD matrices</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html"><i class="fa fa-check"></i><b>8</b> Polygenic (risk) scores (PGS, or PRS)</a>
<ul>
<li class="chapter" data-level="8.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#one-use-of-prs"><i class="fa fa-check"></i><b>8.1</b> One use of PRS</a></li>
<li class="chapter" data-level="8.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-individual-level-data"><i class="fa fa-check"></i><b>8.2</b> PGS from individual-level data</a></li>
<li class="chapter" data-level="8.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#pgs-from-gwas-summary-statistics"><i class="fa fa-check"></i><b>8.3</b> PGS from GWAS summary statistics</a>
<ul>
<li class="chapter" data-level="8.3.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#clumping-thresholding-ct-or-pt-restricting-predictors"><i class="fa fa-check"></i><b>8.3.1</b> <span style="color:#38761D">Clumping</span> + <span style="color:#1515FF">Thresholding</span> (C+T, or P+T): restricting predictors</a></li>
<li class="chapter" data-level="8.3.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#other-methods"><i class="fa fa-check"></i><b>8.3.2</b> Other methods</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#predictive-ability-of-pgs"><i class="fa fa-check"></i><b>8.4</b> Predictive ability of PGS</a></li>
<li class="chapter" data-level="8.5" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#future-of-pgs"><i class="fa fa-check"></i><b>8.5</b> Future of PGS</a></li>
<li class="chapter" data-level="8.6" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#exo-ldpred2"><i class="fa fa-check"></i><b>8.6</b> Exercise with LDpred2 and lassosum2</a>
<ul>
<li class="chapter" data-level="8.6.1" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#preparing-the-data"><i class="fa fa-check"></i><b>8.6.1</b> Preparing the data</a></li>
<li class="chapter" data-level="8.6.2" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#ldpred2"><i class="fa fa-check"></i><b>8.6.2</b> LDpred2</a></li>
<li class="chapter" data-level="8.6.3" data-path="polygenic-risk-scores-pgs-or-prs.html"><a href="polygenic-risk-scores-pgs-or-prs.html#lassosum2-grid-of-models"><i class="fa fa-check"></i><b>8.6.3</b> lassosum2: grid of models</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html"><i class="fa fa-check"></i><b>9</b> Fine-mapping with SuSiE</a>
<ul>
<li class="chapter" data-level="9.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#objective"><i class="fa fa-check"></i><b>9.1</b> Objective</a>
<ul>
<li class="chapter" data-level="9.1.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#requirements"><i class="fa fa-check"></i><b>9.1.1</b> Requirements:</a></li>
<li class="chapter" data-level="9.1.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-libraries"><i class="fa fa-check"></i><b>9.1.2</b> Load libraries</a></li>
</ul></li>
<li class="chapter" data-level="9.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#simulated-example"><i class="fa fa-check"></i><b>9.2</b> Simulated Example</a>
<ul>
<li class="chapter" data-level="9.2.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#load-data"><i class="fa fa-check"></i><b>9.2.1</b> Load data</a></li>
<li class="chapter" data-level="9.2.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#fit-susie-to-the-data"><i class="fa fa-check"></i><b>9.2.2</b> Fit SuSiE to the data</a></li>
<li class="chapter" data-level="9.2.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#plot-the-susie-results"><i class="fa fa-check"></i><b>9.2.3</b> Plot the SuSiE results</a></li>
<li class="chapter" data-level="9.2.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#lets-take-a-closer-look-at-which-variants-are-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.4</b> Let’s take a closer look at which variants are in the credible sets</a></li>
<li class="chapter" data-level="9.2.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#how-correlated-are-the-variants-in-the-credible-sets"><i class="fa fa-check"></i><b>9.2.5</b> How correlated are the variants in the credible sets?</a></li>
</ul></li>
<li class="chapter" data-level="9.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#real-data"><i class="fa fa-check"></i><b>9.3</b> Real data</a>
<ul>
<li class="chapter" data-level="9.3.1" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#import-gwas-data"><i class="fa fa-check"></i><b>9.3.1</b> Import GWAS data</a></li>
<li class="chapter" data-level="9.3.2" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#find-window-around-loci-of-interest"><i class="fa fa-check"></i><b>9.3.2</b> Find window around loci of interest</a></li>
<li class="chapter" data-level="9.3.3" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#ldplink"><i class="fa fa-check"></i><b>9.3.3</b> Create an LD matrix</a></li>
<li class="chapter" data-level="9.3.4" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#read-in-data-and-fit-susie"><i class="fa fa-check"></i><b>9.3.4</b> Read in data and fit SuSiE</a></li>
<li class="chapter" data-level="9.3.5" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#examine-results"><i class="fa fa-check"></i><b>9.3.5</b> Examine results</a></li>
<li class="chapter" data-level="9.3.6" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#locus-2"><i class="fa fa-check"></i><b>9.3.6</b> Locus 2</a></li>
<li class="chapter" data-level="9.3.7" data-path="fine-mapping-with-susie.html"><a href="fine-mapping-with-susie.html#reflections"><i class="fa fa-check"></i><b>9.3.7</b> Reflections</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="10" data-path="some-other-analyses.html"><a href="some-other-analyses.html"><i class="fa fa-check"></i><b>10</b> Some other analyses</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Human Genetics course using R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="some-other-analyses" class="section level1 hasAnchor" number="10">
<h1><span class="header-section-number">Chapter 10</span> Some other analyses<a href="some-other-analyses.html#some-other-analyses" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Read more in <a href="other-files/Other_Post_GWAS_Analyses_SMARTBiomed%202025.pptx">this PowerPoint presentation</a>.</p>
<p>Here we just mention some other types of analyses that haven’t really been covered in the course, but that are useful to know that they exist. We simply mention a few tools; you can find a list with more <a href="https://github.com/MichelNivard/awesome-complex-trait-genetics">here</a>.</p>
<ul>
<li><p><strong>Estimation of SNP-based heritability</strong>: The proportion of phenotypic variance in a trait that can be explained by genetic variation captured by measured SNPs. A simple way to estimate it is with <a href="https://github.com/bulik/ldsc/wiki/Heritability-and-Genetic-Correlation">LD Score regression (LDSC)</a> using GWAS summary statistics <span class="citation">(<a href="#ref-bulik2015ld">B. K. Bulik-Sullivan et al., 2015</a>)</span>. R packages bigsnpr and GenomicSEM also provide some implementation of LDSC. There exist many other tools to estimate SNP-heritability, such as LDpred2-auto (mentioned before).</p></li>
<li><p><strong>Genetic Correlation</strong>: LDSC can be also used to estimate the genome-wide genetic correlation between two traits using GWAS summary statistics <span class="citation">(<a href="#ref-bulik2015atlas">B. Bulik-Sullivan et al., 2015</a>)</span>, and is also implemented in R package GenomicSEM. Local genetic correlations (e.g. within LD blocks) can also be estimated <span class="citation">(<a href="#ref-darlay2025exploring">Darlay et al., 2025</a>; <a href="#ref-zhang2023benchmarking">Zhang, Zhang, Zhang, &amp; Zhao, 2023</a>)</span>.</p></li>
<li><p><strong>Partitioned heritability</strong>: Heritability partitioned by functional annotation (which may be cell-type or state-specific), which can be estimated using <a href="https://github.com/bulik/ldsc/wiki/Partitioned-Heritability">S-LDSC</a> <span class="citation">(<a href="#ref-finucane2015partitioning">Finucane et al., 2015</a>)</span>, <a href="https://dougspeed.com/heritability-enrichments/">SumHer</a> <span class="citation">(<a href="#ref-speed2019sumher">Speed &amp; Balding, 2019</a>)</span> and <a href="https://github.com/zhilizheng/SBayesRC">SBayesRC</a> <span class="citation">(<a href="#ref-zheng2024leveraging">Zheng et al., 2024</a>)</span>.</p></li>
<li><p><strong>Gene-based tests</strong>: These determine the genes associated with a trait from a GWAS by using proximity, such as mBAT-combo as an <a href="https://github.com/Share-AL-work/mBAT.combo">R package</a> or a <a href="https://yanglab.westlake.edu.cn/software/gcta/#mBAT-combo">command-line software</a> <span class="citation">(<a href="#ref-li2023mbat">Li et al., 2023</a>)</span>.</p></li>
<li><p><strong>Summary statistic imputation</strong>: imputation of GWAS summary statistics help increase the SNP overlap between GWAS summary statistics and LD reference panels.</p></li>
<li><p><strong>Meta-analysis of GWAS</strong>:</p>
<ul>
<li>same trait, different cohorts: <a href="https://github.com/statgen/METAL">METAL</a> <span class="citation">(<a href="#ref-willer2010metal">Willer, Li, &amp; Abecasis, 2010</a>)</span></li>
<li>multiple traits: <a href="https://github.com/JonJala/mtag">MTAG</a> <span class="citation">(<a href="#ref-turley2018multi">Turley et al., 2018</a>)</span></li>
<li>multiple ancestries: <a href="https://genomics.ut.ee/en/tools">MR-MEGA</a> <span class="citation">(<a href="#ref-magi2017trans">Mägi et al., 2017</a>)</span></li>
</ul></li>
<li><p><strong>Colocalisation</strong>: The same variant may be significant for two traits, but not necessarily be causal for both. Tools like <a href="https://chr1swallace.github.io/coloc/articles/a06_SuSiE.html">COLOC + SuSiE</a> directly assess this, yielding probabilities the signal is driven by the same variant, or two different variants <span class="citation">(<a href="#ref-wallace2021more">Wallace, 2021</a>)</span>. <a href="https://yanglab.westlake.edu.cn/software/smr/#SMR&amp;HEIDIanalysis">SMR</a> is a complementary approach designed for testing if a molQTL is also a causal variant for a trait, yielding an association statistic <span class="citation">(<a href="#ref-zhu2016integration">Zhu et al., 2016</a>)</span>.</p></li>
<li><p><strong>Gene expression signature of a trait</strong>: Use GWAS summary stats with eQTL/gene expression data. Early tools include <a href="https://github.com/gusevlab/fusion_twas">FUSION</a> <span class="citation">(<a href="#ref-gusev2016integrative">Gusev et al., 2016</a>)</span> and <a href="https://github.com/hakyimlab/MetaXcan">PrediXcan</a> <span class="citation">(<a href="#ref-gamazon2015gene">Gamazon et al., 2015</a>)</span>, although many more have been developed since.</p></li>
<li><p><strong>Structural Equation Modelling</strong>: <a href="https://github.com/GenomicSEM/GenomicSEM/wiki">GenomicSEM</a> is a tool for structural equation modelling of GWAS summary statistics, allowing you to model the relationships between traits using their genetic architecture <span class="citation">(<a href="#ref-grotzinger2019genomic">Grotzinger et al., 2019</a>)</span>.</p></li>
</ul>

</div>
<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-bulik2015atlas" class="csl-entry">
Bulik-Sullivan, B., Finucane, H.K., Anttila, V., Gusev, A., Day, F.R., Loh, P.-R., et al. (2015). <a href="https://doi.org/10.1038/ng.3406">An atlas of genetic correlations across human diseases and traits</a>. <em>Nature Genetics</em>, <em>47</em>, 1236–1241.
</div>
<div id="ref-bulik2015ld" class="csl-entry">
Bulik-Sullivan, B.K., Loh, P.-R., Finucane, H.K., Ripke, S., Yang, J., Psychiatric Genomics Consortium, S.W.G. of the, et al. (2015). <a href="https://doi.org/10.1038/ng.3211"><span>LD</span> score regression distinguishes confounding from polygenicity in genome-wide association studies</a>. <em>Nature Genetics</em>, <em>47</em>, 291–295.
</div>
<div id="ref-darlay2025exploring" class="csl-entry">
Darlay, R., Shah, R.L., Dodds, R.M., Nair, A.T., Pearson, E.R., Witham, M.D., et al. (2025). <a href="https://doi.org/10.1002/gepi.70012">Exploring similarities and differences between methods that exploit patterns of local genetic correlation to identify shared causal loci through application to genome-wide association studies of multiple long term conditions</a>. <em>Genetic Epidemiology</em>, <em>49</em>, e70012.
</div>
<div id="ref-finucane2015partitioning" class="csl-entry">
Finucane, H.K., Bulik-Sullivan, B., Gusev, A., Trynka, G., Reshef, Y., Loh, P.-R., et al. (2015). <a href="https://doi.org/10.1038/ng.3404">Partitioning heritability by functional annotation using genome-wide association summary statistics</a>. <em>Nature Genetics</em>, <em>47</em>, 1228–1235.
</div>
<div id="ref-gamazon2015gene" class="csl-entry">
Gamazon, E.R., Wheeler, H.E., Shah, K.P., Mozaffari, S.V., Aquino-Michaels, K., Carroll, R.J., et al. (2015). <a href="https://doi.org/10.1038/ng.3367">A gene-based association method for mapping traits using reference transcriptome data</a>. <em>Nature Genetics</em>, <em>47</em>, 1091–1098.
</div>
<div id="ref-grotzinger2019genomic" class="csl-entry">
Grotzinger, A.D., Rhemtulla, M., Vlaming, R. de, Ritchie, S.J., Mallard, T.T., Hill, W.D., et al. (2019). <a href="https://doi.org/10.1038/s41562-019-0566-x">Genomic structural equation modelling provides insights into the multivariate genetic architecture of complex traits</a>. <em>Nature Human Behaviour</em>, <em>3</em>, 513–525.
</div>
<div id="ref-gusev2016integrative" class="csl-entry">
Gusev, A., Ko, A., Shi, H., Bhatia, G., Chung, W., Penninx, B.W., et al. (2016). <a href="https://doi.org/10.1038/ng.3506">Integrative approaches for large-scale transcriptome-wide association studies</a>. <em>Nature Genetics</em>, <em>48</em>, 245–252.
</div>
<div id="ref-li2023mbat" class="csl-entry">
Li, A., Liu, S., Bakshi, A., Jiang, L., Chen, W., Zheng, Z., et al. (2023). <a href="https://doi.org/10.1016/j.ajhg.2022.12.006"><span class="nocase">mBAT-combo</span>: A more powerful test to detect gene-trait associations from <span>GWAS</span> data</a>. <em>The American Journal of Human Genetics</em>, <em>110</em>, 30–43.
</div>
<div id="ref-magi2017trans" class="csl-entry">
Mägi, R., Horikoshi, M., Sofer, T., Mahajan, A., Kitajima, H., Franceschini, N., et al. (2017). <a href="https://doi.org/10.1093/hmg/ddx280">Trans-ethnic meta-regression of genome-wide association studies accounting for ancestry increases power for discovery and improves fine-mapping resolution</a>. <em>Human Molecular Genetics</em>, <em>26</em>, 3639–3650.
</div>
<div id="ref-speed2019sumher" class="csl-entry">
Speed, D., &amp; Balding, D.J. (2019). <a href="https://doi.org/10.1038/s41588-018-0279-5"><span>SumHer</span> better estimates the <span>SNP</span> heritability of complex traits from summary statistics</a>. <em>Nature Genetics</em>, <em>51</em>, 277–284.
</div>
<div id="ref-turley2018multi" class="csl-entry">
Turley, P., Walters, R.K., Maghzian, O., Okbay, A., Lee, J.J., Fontana, M.A., et al. (2018). <a href="https://doi.org/10.1038/s41588-017-0009-4">Multi-trait analysis of genome-wide association summary statistics using <span>MTAG</span></a>. <em>Nature Genetics</em>, <em>50</em>, 229–237.
</div>
<div id="ref-wallace2021more" class="csl-entry">
Wallace, C. (2021). <a href="https://doi.org/10.1371/journal.pgen.1009440">A more accurate method for colocalisation analysis allowing for multiple causal variants</a>. <em>PLoS Genetics</em>, <em>17</em>, e1009440.
</div>
<div id="ref-willer2010metal" class="csl-entry">
Willer, C.J., Li, Y., &amp; Abecasis, G.R. (2010). <a href="https://doi.org/10.1093/bioinformatics/btq340"><span>METAL</span>: Fast and efficient meta-analysis of genomewide association scans</a>. <em>Bioinformatics</em>, <em>26</em>, 2190–2191.
</div>
<div id="ref-zhang2023benchmarking" class="csl-entry">
Zhang, C., Zhang, Y., Zhang, Y., &amp; Zhao, H. (2023). <a href="https://doi.org/10.1093/bib/bbad407">Benchmarking of local genetic correlation estimation methods using summary statistics from genome-wide association studies</a>. <em>Briefings in Bioinformatics</em>, <em>24</em>, bbad407.
</div>
<div id="ref-zheng2024leveraging" class="csl-entry">
Zheng, Z., Liu, S., Sidorenko, J., Wang, Y., Lin, T., Yengo, L., et al. (2024). <a href="https://doi.org/10.1038/s41588-024-01704-y">Leveraging functional genomic annotations and genome coverage to improve polygenic prediction of complex traits within and between ancestries</a>. <em>Nature Genetics</em>, <em>56</em>, 767–777.
</div>
<div id="ref-zhu2016integration" class="csl-entry">
Zhu, Z., Zhang, F., Hu, H., Bakshi, A., Robinson, M.R., Powell, J.E., et al. (2016). <a href="https://doi.org/10.1038/ng.3538">Integration of summary data from <span>GWAS</span> and <span class="nocase">eQTL</span> studies predicts complex trait gene targets</a>. <em>Nature Genetics</em>, <em>48</em>, 481–487.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="fine-mapping-with-susie.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": false,
"twitter": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
